--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -552,6 +_,11 @@
         boolean isBonusGroup = this.shouldSpawnBonusGroup();
 
         for (Raid.RaiderType raiderType : Raid.RaiderType.VALUES) {
+            // Purpur start - exclude Illusioners if the feature is disabled
+            if (raiderType == Raid.RaiderType.ILLUSIONER && !level.purpurConfig.illusionerSpawnInRaids) {
+                continue;
+            }
+            // Purpur end
             int numSpawns = this.getDefaultNumSpawns(raiderType, groupNumber, isBonusGroup)
                 + this.getPotentialBonusSpawns(raiderType, this.random, groupNumber, difficulty, isBonusGroup);
             int ravagersSpawned = 0;
@@ -573,9 +_,17 @@
                     Raider ridingRaider = null;
                     if (groupNumber == this.getNumGroups(Difficulty.NORMAL)) {
                         ridingRaider = EntityTypes.PILLAGER.create(level, EntitySpawnReason.EVENT);
+                        // Purpur start - use the setting to determine if Ravagers with Illusioners should spawn
                     } else if (groupNumber >= this.getNumGroups(Difficulty.HARD)) {
                         if (ravagersSpawned == 0) {
-                            ridingRaider = EntityTypes.EVOKER.create(level, EntitySpawnReason.EVENT);
+                            if (level.purpurConfig.illusionerSpawnInRaids) {
+                                ridingRaider = random.nextBoolean()
+                                    ? EntityTypes.EVOKER.create(level, EntitySpawnReason.EVENT)
+                                    : EntityTypes.ILLUSIONER.create(level, EntitySpawnReason.EVENT);
+                            } else {
+                                ridingRaider = EntityTypes.EVOKER.create(level, EntitySpawnReason.EVENT);
+                            }
+                            // Purpur end
                         } else {
                             ridingRaider = EntityTypes.VINDICATOR.create(level, EntitySpawnReason.EVENT);
                         }
@@ -782,6 +_,7 @@
         int bonusSpawns;
         switch (type) {
             case VINDICATOR:
+            case ILLUSIONER: // Purpur - Ability for illusioners to spawn during raids
             case PILLAGER:
                 if (isEasy) {
                     bonusSpawns = random.nextInt(2);
@@ -867,6 +_,7 @@
         VINDICATOR(EntityTypes.VINDICATOR, new int[]{0, 0, 2, 0, 1, 4, 2, 5}),
         EVOKER(EntityTypes.EVOKER, new int[]{0, 0, 0, 0, 0, 1, 1, 2}),
         PILLAGER(EntityTypes.PILLAGER, new int[]{0, 4, 3, 3, 4, 4, 4, 2}),
+        ILLUSIONER(EntityTypes.ILLUSIONER, new int[]{0, 0, 1, 0, 1, 1, 0, 1}), // Purpur - Add Illusioners to RaiderType
         WITCH(EntityTypes.WITCH, new int[]{0, 0, 0, 0, 3, 0, 0, 1}),
         RAVAGER(EntityTypes.RAVAGER, new int[]{0, 0, 0, 1, 0, 1, 0, 2});
 
