--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -132,6 +_,7 @@
     private Raid.RaidStatus status;
     private int celebrationTicks;
     private Optional<BlockPos> waveSpawnPos = Optional.empty();
+    public boolean includeIllusioners;
 
     public Raid(final BlockPos center, final Difficulty difficulty) {
         this.active = true;
@@ -550,8 +_,12 @@
         this.totalHealth = 0.0F;
         DifficultyInstance difficulty = level.getCurrentDifficultyAt(pos);
         boolean isBonusGroup = this.shouldSpawnBonusGroup();
+        includeIllusioners = level.purpurConfig.illusionerSpawnInRaids;
 
         for (Raid.RaiderType raiderType : Raid.RaiderType.VALUES) {
+            if (raiderType == Raid.RaiderType.ILLUSIONER && !this.includeIllusioners) {
+                continue;
+            }
             int numSpawns = this.getDefaultNumSpawns(raiderType, groupNumber, isBonusGroup)
                 + this.getPotentialBonusSpawns(raiderType, this.random, groupNumber, difficulty, isBonusGroup);
             int ravagersSpawned = 0;
@@ -575,7 +_,13 @@
                         ridingRaider = EntityTypes.PILLAGER.create(level, EntitySpawnReason.EVENT);
                     } else if (groupNumber >= this.getNumGroups(Difficulty.HARD)) {
                         if (ravagersSpawned == 0) {
-                            ridingRaider = EntityTypes.EVOKER.create(level, EntitySpawnReason.EVENT);
+                            if (level.purpurConfig.illusionerSpawnInRaids) {
+                                ridingRaider = random.nextBoolean()
+                                    ? EntityTypes.EVOKER.create(level, EntitySpawnReason.EVENT)
+                                    : EntityTypes.ILLUSIONER.create(level, EntitySpawnReason.EVENT);
+                            } else {
+                                ridingRaider = EntityTypes.EVOKER.create(level, EntitySpawnReason.EVENT);
+                            }
                         } else {
                             ridingRaider = EntityTypes.VINDICATOR.create(level, EntitySpawnReason.EVENT);
                         }
@@ -782,6 +_,7 @@
         int bonusSpawns;
         switch (type) {
             case VINDICATOR:
+            case ILLUSIONER:
             case PILLAGER:
                 if (isEasy) {
                     bonusSpawns = random.nextInt(2);
@@ -867,6 +_,7 @@
         VINDICATOR(EntityTypes.VINDICATOR, new int[]{0, 0, 2, 0, 1, 4, 2, 5}),
         EVOKER(EntityTypes.EVOKER, new int[]{0, 0, 0, 0, 0, 1, 1, 2}),
         PILLAGER(EntityTypes.PILLAGER, new int[]{0, 4, 3, 3, 4, 4, 4, 2}),
+        ILLUSIONER(EntityTypes.ILLUSIONER, new int[]{0, 0, 1, 0, 1, 1, 0, 1}),
         WITCH(EntityTypes.WITCH, new int[]{0, 0, 0, 0, 3, 0, 0, 1}),
         RAVAGER(EntityTypes.RAVAGER, new int[]{0, 0, 0, 1, 0, 1, 0, 2});
 
