--- a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTraderSpawner.java
+++ b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTraderSpawner.java
@@ -134,7 +_,17 @@
             int i1 = pos.getX() + this.random.nextInt(maxDistance * 2) - maxDistance;
             int i2 = pos.getZ() + this.random.nextInt(maxDistance * 2) - maxDistance;
             int height = level.getHeight(Heightmap.Types.WORLD_SURFACE, i1, i2);
-            BlockPos blockPos1 = new BlockPos(i1, height, i2);
+            // Purpur start - Allow toggling special MobSpawners per world - allow traders to spawn below nether roof
+            BlockPos.MutableBlockPos blockPos1 = new BlockPos.MutableBlockPos(i1, height, i2);
+            if (level.dimensionType().hasCeiling()) {
+                do {
+                    blockPos1.relative(net.minecraft.core.Direction.DOWN);
+                } while (!level.getBlockState(blockPos1).isAir());
+                do {
+                    blockPos1.relative(net.minecraft.core.Direction.DOWN);
+                } while (level.getBlockState(blockPos1).isAir() && blockPos1.getY() > 0);
+            }
+            // Purpur end - Allow toggling special MobSpawners per world
             if (placementType.isSpawnPositionOk(level, blockPos1, EntityType.WANDERING_TRADER)) {
                 blockPos = blockPos1;
                 break;
