--- a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTraderSpawner.java
+++ b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTraderSpawner.java
@@ -133,7 +_,17 @@
             int xPosition = referencePosition.getX() + this.random.nextInt(radius * 2) - radius;
             int zPosition = referencePosition.getZ() + this.random.nextInt(radius * 2) - radius;
             int yPosition = level.getHeight(SpawnPlacements.getHeightmapType(EntityType.WANDERING_TRADER), xPosition, zPosition);
-            BlockPos spawnPos = new BlockPos(xPosition, yPosition, zPosition);
+            // Purpur start - Allow toggling special MobSpawners per world - allow traders to spawn below nether roof
+            BlockPos.MutableBlockPos spawnPos = new BlockPos.MutableBlockPos(xPosition, yPosition, zPosition);
+            if (level.dimensionType().hasCeiling()) {
+                do {
+                    spawnPos.relative(net.minecraft.core.Direction.DOWN);
+                } while (!level.getBlockState(spawnPos).isAir());
+                do {
+                    spawnPos.relative(net.minecraft.core.Direction.DOWN);
+                } while (level.getBlockState(spawnPos).isAir() && spawnPos.getY() > 0);
+            }
+            // Purpur end - Allow toggling special MobSpawners per world
             if (wanderingTraderSpawnType.isSpawnPositionOk(level, spawnPos, EntityType.WANDERING_TRADER)) {
                 spawnPosition = spawnPos;
                 break;
