From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index c2dd2da4bfb7d65ad2730ba1ce771e2b488beda4..271e1ff85d778998de465e1099c4fc7219b2d98d 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -154,6 +154,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     private BlockPos homePosition = BlockPos.ZERO;
     private int homeRadius = -1;
     public boolean aware = true; // CraftBukkit
+    public int ticksSinceLastInteraction; // Purpur - Entity lifespan
 
     protected Mob(final EntityType<? extends Mob> type, final Level level) {
         super(type, level);
@@ -308,6 +309,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
                 target = null;
             }
         }
+        if (target instanceof net.minecraft.server.level.ServerPlayer) this.ticksSinceLastInteraction = 0; // Purpur - Entity lifespan
         this.target = target;
         return true;
         // CraftBukkit end
@@ -351,8 +353,28 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
         }
 
         profiler.pop();
+        incrementTicksSinceLastInteraction(); // Purpur - Entity lifespan
     }
 
+    // Purpur start - Entity lifespan
+    private void incrementTicksSinceLastInteraction() {
+        ++this.ticksSinceLastInteraction;
+        if (getRider() != null) {
+            this.ticksSinceLastInteraction = 0;
+            return;
+        }
+        if (this.level().purpurConfig.entityLifeSpan <= 0) {
+            return; // feature disabled
+        }
+        if (!this.removeWhenFarAway(0) || isPersistenceRequired() || requiresCustomPersistence() || hasCustomName()) {
+            return; // mob persistent
+        }
+        if (this.ticksSinceLastInteraction > this.level().purpurConfig.entityLifeSpan) {
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD);
+        }
+    }
+    // Purpur end - Entity lifespan
+
     @Override
     protected void playHurtSound(final DamageSource source) {
         this.resetAmbientSoundTime();
@@ -450,6 +472,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
             output.putBoolean("NoAI", this.isNoAi());
         }
         output.putBoolean("Bukkit.Aware", this.aware); // CraftBukkit
+        output.putInt("Purpur.ticksSinceLastInteraction", this.ticksSinceLastInteraction); // Purpur - Entity lifespan
     }
 
     @Override
@@ -477,6 +500,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
         this.lootTableSeed = input.getLongOr("DeathLootTableSeed", 0L);
         this.setNoAi(input.getBooleanOr("NoAI", false));
         this.aware = input.getBooleanOr("Bukkit.Aware", true); // CraftBukkit
+        this.ticksSinceLastInteraction = input.getIntOr("Purpur.ticksSinceLastInteraction", 0); // Purpur- Entity lifespan
     }
 
     @Override
@@ -1269,7 +1293,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
             );
         }
 
-        this.setLeftHanded(random.nextFloat() < 0.05F);
+        this.setLeftHanded(random.nextFloat() < level.getLevel().purpurConfig.entityLeftHandedChance); // Purpur - Changeable Mob Left Handed Chance
         return groupData;
     }
 
@@ -1626,6 +1650,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
         }
 
         this.postPiercingAttack();
+        if (target instanceof net.minecraft.server.level.ServerPlayer) this.ticksSinceLastInteraction = 0; // Purpur - Entity lifespan
         return wasHurt;
     }
 
