--- a/net/minecraft/world/level/levelgen/structure/templatesystem/loader/TemplatePathFactory.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/loader/TemplatePathFactory.java
@@ -20,13 +_,21 @@
     }
 
     public Path createAndValidatePathToStructure(final Identifier id, final FileToIdConverter converter) {
-        return this.createAndValidatePathToStructure(converter.idToFile(id));
+        // Purpur start - export path override
+        Identifier cleanId = id;
+        if (cleanId.getPath().startsWith("minecraft/")) {
+            cleanId = Identifier.fromNamespaceAndPath(cleanId.getNamespace(), cleanId.getPath().substring("minecraft/".length()));
+        }
+
+        return this.createAndValidatePathToStructure(converter.idToFile(cleanId));
+        // Purpur end
     }
 
     public Path createAndValidatePathToStructure(final Identifier resourceLocation) {
         Path namespacePath = this.sourceDir.resolve(resourceLocation.getNamespace());
         List<String> decomposedPath = FileUtil.decomposePath(resourceLocation.getPath())
             .getOrThrow(msg -> new IdentifierException("Invalid file path '" + resourceLocation + "': " + msg));
+
         if (!decomposedPath.stream().allMatch(FileUtil::isPathPartPortable)) {
             throw new IdentifierException("Resource path '" + resourceLocation + "' is not portable");
         } else {
