From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/entity/animal/rabbit/Rabbit.java b/net/minecraft/world/entity/animal/rabbit/Rabbit.java
index fa503dd420af762fbe8b36f98a86451b24a76428..ab7e618ab8ef872d794f1ff82de881a5bca04bc5 100644
--- a/net/minecraft/world/entity/animal/rabbit/Rabbit.java
+++ b/net/minecraft/world/entity/animal/rabbit/Rabbit.java
@@ -419,10 +419,23 @@ public class Rabbit extends Animal {
         }
 
         this.setVariant(variant);
+
+        // Purpur start - Special mobs naturally spawn
+        if (variant != Variant.EVIL && level.getLevel().purpurConfig.rabbitNaturalToast > 0D && random.nextDouble() <= level.getLevel().purpurConfig.rabbitNaturalToast) {
+            setCustomName(Component.translatable("Toast"));
+        }
+        // Purpur end - Special mobs naturally spawn
+
         return super.finalizeSpawn(level, difficulty, spawnReason, groupData);
     }
 
     private static Rabbit.Variant getRandomRabbitVariant(final LevelAccessor level, final BlockPos pos) {
+        // Purpur start - Special mobs naturally spawn
+        Level world = level.getMinecraftWorld();
+        if (world.purpurConfig.rabbitNaturalKiller > 0D && world.getRandom().nextDouble() <= world.purpurConfig.rabbitNaturalKiller) {
+            return Rabbit.Variant.EVIL;
+        }
+        // Purpur end - Special mobs naturally spawn
         Holder<Biome> biome = level.getBiome(pos);
         int randomVal = level.getRandom().nextInt(100);
         if (biome.is(BiomeTags.SPAWNS_WHITE_RABBITS)) {
