From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
index 78dbdf43ac9095eb3cf0b0653bab2603d5527ff4..993e6aabfd05a74523705b3dc2c4f8f2f88b2de3 100644
--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -190,6 +190,21 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
         }
 
         ItemStack fuel = entity.items.get(1);
+        // Purpur start - Furnace uses lava from underneath
+        boolean usedLavaFromUnderneath = false;
+        if (level.purpurConfig.furnaceUseLavaFromUnderneath && !isLit && fuel.isEmpty() && !entity.items.get(0).isEmpty() && level.getGameTime() % 20 == 0) {
+            BlockPos below = entity.getBlockPos().below();
+            BlockState belowState = level.getBlockStateIfLoaded(below);
+            if (belowState != null && belowState.is(net.minecraft.world.level.block.Blocks.LAVA)) {
+                net.minecraft.world.level.material.FluidState fluidState = belowState.getFluidState();
+                if (fluidState != null && fluidState.isSource()) {
+                    level.setBlock(below, net.minecraft.world.level.block.Blocks.AIR.defaultBlockState(), 3);
+                    fuel = Items.LAVA_BUCKET.getDefaultInstance();
+                    usedLavaFromUnderneath = true;
+                }
+            }
+        }
+        // Purpur end - Furnace uses lava from underneath
         ItemStack ingredient = entity.items.get(0);
         boolean hasIngredient = !ingredient.isEmpty();
         boolean hasFuel = !fuel.isEmpty();
@@ -269,6 +284,8 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
         if (changed) {
             setChanged(level, pos, state);
         }
+
+        if (usedLavaFromUnderneath) entity.items.set(1, ItemStack.EMPTY); // Purpur - Furnace uses lava from underneath
     }
 
     private static void consumeFuel(final NonNullList<ItemStack> items, final ItemStack fuel) {
