From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/block/entity/BeaconBlockEntity.java b/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
index d7a37a3a32c2a305fae9385c2e66adc7ff497346..1e961134cdda4e661951566337d2fe2cf12a9133 100644
--- a/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -142,6 +142,16 @@ public class BeaconBlockEntity extends BlockEntity implements MenuProvider, Name
 
     public double getEffectRange() {
         if (this.effectRange < 0) {
+            // Purpur start - Beacon Activation Range Configurable
+            if (this.level != null) {
+                switch (this.levels) {
+                    case 1: return this.level.purpurConfig.beaconLevelOne;
+                    case 2: return this.level.purpurConfig.beaconLevelTwo;
+                    case 3: return this.level.purpurConfig.beaconLevelThree;
+                    case 4: return this.level.purpurConfig.beaconLevelFour;
+                }
+            }
+            // Purpur end - Beacon Activation Range Configurable
             return this.levels * 10 + 10;
         } else {
             return effectRange;
@@ -170,6 +180,7 @@ public class BeaconBlockEntity extends BlockEntity implements MenuProvider, Name
         int y = pos.getY();
         int z = pos.getZ();
         BlockPos checkPos;
+        boolean isTintedGlass = false; // Purpur - allow beacon effects when covered by tinted glass
         if (entity.lastCheckY < y) {
             checkPos = pos;
             entity.checkingBeamSections = Lists.newArrayList();
@@ -199,11 +210,15 @@ public class BeaconBlockEntity extends BlockEntity implements MenuProvider, Name
                     }
                 }
             } else {
-                if (lastBeamSection == null || state.getLightDampening() >= 15 && !state.is(Blocks.BEDROCK)) {
+                if (level.purpurConfig.beaconAllowEffectsWithTintedGlass && state.getBlock().equals(Blocks.TINTED_GLASS)) {isTintedGlass = true;} // Purpur - allow beacon effects when covered by tinted glass
+                // Purpur start - fix effects being applied when tinted glass is covered
+                if (lastBeamSection == null || state.getLightDampening() >= 15 && !state.is(Blocks.BEDROCK) && !(state.getBlock().equals(Blocks.TINTED_GLASS) && level.purpurConfig.beaconAllowEffectsWithTintedGlass)) {
                     entity.checkingBeamSections.clear();
                     entity.lastCheckY = lastSetBlock;
+                    isTintedGlass = false;
                     break;
                 }
+                // Purpur end - fix effects being applied when tinted glass is covered
 
                 lastBeamSection.increaseHeight();
             }
@@ -214,11 +229,11 @@ public class BeaconBlockEntity extends BlockEntity implements MenuProvider, Name
 
         int previousLevels = entity.levels;
         if (level.getGameTime() % 80L == 0L) {
-            if (!entity.beamSections.isEmpty()) {
+            if (!entity.beamSections.isEmpty() || (level.purpurConfig.beaconAllowEffectsWithTintedGlass && isTintedGlass)) { // Purpur - fix beacon effects persisting with broken base while tinted glass is used
                 entity.levels = updateBase(level, x, y, z);
             }
 
-            if (entity.levels > 0 && !entity.beamSections.isEmpty()) {
+            if (entity.levels > 0 && (!entity.beamSections.isEmpty() || (level.purpurConfig.beaconAllowEffectsWithTintedGlass && isTintedGlass))) { // Purpur - allow beacon effects when covered by tinted glass
                 applyEffects(level, pos, entity.levels, entity.primaryPower, entity.secondaryPower, entity); // Paper - Custom beacon ranges
                 playSound(level, pos, SoundEvents.BEACON_AMBIENT);
             }
