--- a/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -105,6 +_,10 @@
         input.read("PublicBukkitValues", CompoundTag.CODEC)
             .ifPresent(this.persistentDataContainer::putAll);
         // Paper end - read persistent data container
+
+
+        this.persistentLore = input.read("Purpur.persistentLore", net.minecraft.world.item.component.ItemLore.CODEC).orElse(null); // Purpur - Persistent BlockEntity Lore and DisplayName
+
     }
 
     public final void loadWithComponents(ValueInput input) {
@@ -117,6 +_,11 @@
     }
 
     protected void saveAdditional(ValueOutput output) {
+        // Purpur start - Persistent BlockEntity Lore and DisplayName
+        if (this.persistentLore != null) {
+            output.store("Purpur.persistentLore", net.minecraft.world.item.component.ItemLore.CODEC, this.persistentLore);
+        }
+        // Purpur end - Persistent BlockEntity Lore and DisplayName
     }
 
     public final CompoundTag saveWithFullMetadata(HolderLookup.Provider registries) {
@@ -402,4 +_,16 @@
             return this.blockEntity.getNameForReporting() + "@" + this.blockEntity.getBlockPos();
         }
     }
+
+    // Purpur start - Persistent BlockEntity Lore and DisplayName
+    private net.minecraft.world.item.component.@Nullable ItemLore persistentLore = null;
+
+    public void setPersistentLore(net.minecraft.world.item.component.ItemLore lore) {
+        this.persistentLore = lore;
+    }
+
+    public @org.jetbrains.annotations.Nullable net.minecraft.world.item.component.ItemLore getPersistentLore() {
+        return this.persistentLore;
+    }
+    // Purpur end - Persistent BlockEntity Lore and DisplayName
 }
