--- a/net/minecraft/world/level/block/entity/FuelValues.java
+++ b/net/minecraft/world/level/block/entity/FuelValues.java
@@ -17,7 +_,7 @@
 import net.minecraft.world.level.block.Blocks;
 
 public class FuelValues {
-    public final Object2IntSortedMap<Item> values;
+    private final Object2IntSortedMap<Item> values;
 
     private FuelValues(final Object2IntSortedMap<Item> values) {
         this.values = values;
