From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/inventory/AbstractFurnaceMenu.java b/net/minecraft/world/inventory/AbstractFurnaceMenu.java
index 19bfba135158f19a6055af611d4fc2b34be9ca17..bf9acbd00054084b01cdff6ececd16be5778783f 100644
--- a/net/minecraft/world/inventory/AbstractFurnaceMenu.java
+++ b/net/minecraft/world/inventory/AbstractFurnaceMenu.java
@@ -116,7 +116,13 @@ public abstract class AbstractFurnaceMenu extends RecipeBookMenu {
             } else if (slotIndex != 1 && slotIndex != 0) {
                 if (this.canSmelt(stack)) {
                     if (!this.moveItemStackTo(stack, 0, 1, false)) {
-                        return ItemStack.EMPTY;
+                        // Purpur start - Added the ability to add combustible items
+                        if (this.isFuel(stack)) {
+                            if (!this.moveItemStackTo(stack, 1, 2, false)) {
+                                return ItemStack.EMPTY;
+                            }
+                        }
+                        // Purpur end - Added the ability to add combustible items
                     }
                 } else if (this.isFuel(stack)) {
                     if (!this.moveItemStackTo(stack, 1, 2, false)) {
