From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/inventory/AbstractFurnaceMenu.java b/net/minecraft/world/inventory/AbstractFurnaceMenu.java
index 0860bdf9dbc08067872a0b25a48d6031ddccce60..b02f12de2f449208605dd1efc14c2bc402429c35 100644
--- a/net/minecraft/world/inventory/AbstractFurnaceMenu.java
+++ b/net/minecraft/world/inventory/AbstractFurnaceMenu.java
@@ -122,7 +122,13 @@ public abstract class AbstractFurnaceMenu extends RecipeBookMenu {
             } else if (slotIndex != 1 && slotIndex != 0) {
                 if (this.canSmelt(stack)) {
                     if (!this.moveItemStackTo(stack, 0, 1, false)) {
-                        return ItemStack.EMPTY;
+                        // Purpur start - Added the ability to add combustible items
+                        if (this.isFuel(stack)) {
+                            if (!this.moveItemStackTo(stack, 1, 2, false)) {
+                                return ItemStack.EMPTY;
+                            }
+                        }
+                        // Purpur end - Added the ability to add combustible items
                     }
                 } else if (this.isFuel(stack)) {
                     if (!this.moveItemStackTo(stack, 1, 2, false)) {
