--- a/net/minecraft/world/inventory/AbstractFurnaceMenu.java
+++ b/net/minecraft/world/inventory/AbstractFurnaceMenu.java
@@ -122,7 +_,13 @@
             } else if (slotIndex != 1 && slotIndex != 0) {
                 if (this.canSmelt(stack)) {
                     if (!this.moveItemStackTo(stack, 0, 1, false)) {
-                        return ItemStack.EMPTY;
+                        // Purpur start - Added the ability to add combustible items
+                        if (this.isFuel(stack)) {
+                            if (!this.moveItemStackTo(stack, 1, 2, false)) {
+                                return ItemStack.EMPTY;
+                            }
+                        }
+                        // Purpur end - Added the ability to add combustible items
                     }
                 } else if (this.isFuel(stack)) {
                     if (!this.moveItemStackTo(stack, 1, 2, false)) {
