--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -63,6 +_,7 @@
     private final List<ContainerListener> containerListeners = Lists.newArrayList();
     private @Nullable ContainerSynchronizer synchronizer;
     private boolean suppressRemoteUpdates;
+    @Nullable protected ItemStack activeQuickItem = null; // Purpur - Anvil API
     // CraftBukkit start
     public boolean checkReachable = true;
     public abstract org.bukkit.inventory.InventoryView getBukkitView();
