--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftCopperGolem.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftCopperGolem.java
@@ -67,4 +_,17 @@
             default -> throw new IllegalStateException("Unexpected value: " + oxidizing);
         }
     }
+
+    // Purpur start - Summoner API
+    @Override
+    @org.jetbrains.annotations.Nullable
+    public java.util.UUID getSummoner() {
+        return getHandle().getSummoner();
+    }
+
+    @Override
+    public void setSummoner(@org.jetbrains.annotations.Nullable java.util.UUID summoner) {
+        getHandle().setSummoner(summoner);
+    }
+    // Purpur end - Summoner API
 }
