--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPhantom.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPhantom.java
@@ -35,12 +_,12 @@
 
     @Override
     public boolean shouldBurnInDay() {
-        return this.getHandle().shouldBurnInDay;
+        return this.getHandle().shouldBurnInDay();
     }
 
     @Override
     public void setShouldBurnInDay(boolean shouldBurnInDay) {
-        this.getHandle().shouldBurnInDay = shouldBurnInDay;
+        this.getHandle().setShouldBurnInDay(shouldBurnInDay);
     }
 
     @Override
