--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftWolf.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftWolf.java
@@ -146,4 +_,15 @@
             return this.getKey().hashCode();
         }
     }
+    // Purpur start - Configurable chance for wolves to spawn rabid
+    @Override
+    public boolean isRabid() {
+        return getHandle().isRabid();
+    }
+
+    @Override
+    public void setRabid(boolean isRabid) {
+        getHandle().setRabid(isRabid);
+    }
+    // Purpur end - Configurable chance for wolves to spawn rabid
 }
