From 7f0c13483f0e738f728cf05164874cdc61852771 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur paperServer File Patches


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
index d37de00e5b9acd1db9b2f7418aaee517b54518d5..7b425eb594cda1a740da6941a21268bc58a5d171 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
@@ -243,4 +243,11 @@ public class CraftVillager extends CraftAbstractVillager implements Villager {
     public void restock() {
         this.getHandle().restock();
     }
+
+    // Purpur start - Lobotomize stuck villagers
+    @Override
+    public boolean isLobotomized() {
+        return getHandle().isLobotomized();
+    }
+    // Purpur end - Lobotomize stuck villagers
 }
