--- a/net/minecraft/world/entity/ai/sensing/NearestBedSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestBedSensor.java
@@ -54,7 +_,7 @@
                 }
             };
             Set<Pair<Holder<PoiType>, BlockPos>> pois = poiManager.findAllWithType(
-                    e -> e.is(PoiTypes.HOME), cacheTest, body.blockPosition(), AcquirePoi.SCAN_RANGE, PoiManager.Occupancy.ANY
+                    e -> e.is(PoiTypes.HOME), cacheTest, body.blockPosition(), level.purpurConfig.villagerNearestBedSensorSearchRadius, PoiManager.Occupancy.ANY // Purpur - Configurable villager search radius
                 )
                 .collect(Collectors.toSet());
             Path path = AcquirePoi.findPathToPois(body, pois);
