--- a/net/minecraft/world/entity/ai/sensing/NearestBedSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestBedSensor.java
@@ -54,9 +_,9 @@
                 }
             };
             Set<Pair<Holder<PoiType>, BlockPos>> pois = poiManager.findAllWithType(
-                    e -> e.is(PoiTypes.HOME), cacheTest, body.blockPosition(), 48, PoiManager.Occupancy.ANY
+                    e -> e.is(PoiTypes.HOME), cacheTest, body.blockPosition(), level.purpurConfig.villagerNearestBedSensorSearchRadius, PoiManager.Occupancy.ANY
                 )
-                .collect(Collectors.toSet());
+                .collect(Collectors.toSet()); // Purpur - Configurable villager search radius
             Path path = AcquirePoi.findPathToPois(body, pois);
             if (path != null && path.canReach()) {
                 BlockPos targetPos = path.getTarget();
