--- a/net/minecraft/world/entity/animal/golem/CopperGolem.java
+++ b/net/minecraft/world/entity/animal/golem/CopperGolem.java
@@ -83,6 +_,7 @@
     private final AnimationState interactionDropItemAnimationState = new AnimationState();
     private final AnimationState interactionDropNoItemAnimationState = new AnimationState();
     public static final EquipmentSlot EQUIPMENT_SLOT_ANTENNA = EquipmentSlot.SADDLE;
+    @Nullable private UUID summoner; // Purpur - Summoner API
 
     public CopperGolem(EntityType<? extends AbstractGolem> type, Level level) {
         super(type, level);
@@ -96,6 +_,17 @@
         this.getBrain().setMemory(MemoryModuleType.TRANSPORT_ITEMS_COOLDOWN_TICKS, this.getRandom().nextInt(60, 100));
     }
 
+    // Purpur start - Summoner API
+    @Nullable
+    public UUID getSummoner() {
+        return summoner;
+    }
+
+    public void setSummoner(@Nullable UUID summoner) {
+        this.summoner = summoner;
+    }
+    // Purpur end - Summoner API
+
     public static AttributeSupplier.Builder createAttributes() {
         return Mob.createMobAttributes().add(Attributes.MOVEMENT_SPEED, 0.2F).add(Attributes.STEP_HEIGHT, 1.0).add(Attributes.MAX_HEALTH, 12.0);
     }
@@ -171,6 +_,7 @@
         super.addAdditionalSaveData(output);
         output.putLong("next_weather_age", this.nextWeatheringTick);
         output.store("weather_state", WeatheringCopper.WeatherState.CODEC, this.getWeatherState());
+        output.storeNullable("Purpur.Summoner", net.minecraft.core.UUIDUtil.CODEC, getSummoner()); // Purpur - Summoner API
     }
 
     @Override
@@ -178,6 +_,7 @@
         super.readAdditionalSaveData(input);
         this.nextWeatheringTick = input.getLongOr("next_weather_age", -1L);
         this.setWeatherState(input.read("weather_state", WeatheringCopper.WeatherState.CODEC).orElse(WeatheringCopper.WeatherState.UNAFFECTED));
+        this.setSummoner(input.read("Purpur.Summoner", net.minecraft.core.UUIDUtil.CODEC).orElse(null)); // Purpur - Summoner API
     }
 
     @Override
