From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/entity/animal/golem/CopperGolem.java b/net/minecraft/world/entity/animal/golem/CopperGolem.java
index 6e526d65f3c4f982e8483a854d80d23b52bd0fd5..4f412c6c661d5972c61ff77fb02e55baa8020c93 100644
--- a/net/minecraft/world/entity/animal/golem/CopperGolem.java
+++ b/net/minecraft/world/entity/animal/golem/CopperGolem.java
@@ -87,6 +87,7 @@ public class CopperGolem extends AbstractGolem implements ContainerUser, Shearab
     private final AnimationState interactionDropItemAnimationState = new AnimationState();
     private final AnimationState interactionDropNoItemAnimationState = new AnimationState();
     public static final EquipmentSlot EQUIPMENT_SLOT_ANTENNA = EquipmentSlot.SADDLE;
+    @Nullable private UUID summoner; // Purpur - Summoner API
 
     public CopperGolem(final EntityType<? extends AbstractGolem> type, final Level level) {
         super(type, level);
@@ -100,6 +101,17 @@ public class CopperGolem extends AbstractGolem implements ContainerUser, Shearab
         this.getBrain().setMemory(MemoryModuleType.TRANSPORT_ITEMS_COOLDOWN_TICKS, this.getRandom().nextInt(60, 100));
     }
 
+    // Purpur start - Summoner API
+    @Nullable
+    public UUID getSummoner() {
+        return summoner;
+    }
+
+    public void setSummoner(@Nullable UUID summoner) {
+        this.summoner = summoner;
+    }
+    // Purpur end - Summoner API
+
     public static AttributeSupplier.Builder createAttributes() {
         return Mob.createMobAttributes().add(Attributes.MOVEMENT_SPEED, 0.2F).add(Attributes.STEP_HEIGHT, 1.0).add(Attributes.MAX_HEALTH, 12.0);
     }
@@ -170,6 +182,7 @@ public class CopperGolem extends AbstractGolem implements ContainerUser, Shearab
         super.addAdditionalSaveData(output);
         output.putLong("next_weather_age", this.nextWeatheringTick);
         output.store("weather_state", WeatheringCopper.WeatherState.CODEC, this.getWeatherState());
+        output.storeNullable("Purpur.Summoner", net.minecraft.core.UUIDUtil.CODEC, getSummoner()); // Purpur - Summoner API
     }
 
     @Override
@@ -177,6 +190,7 @@ public class CopperGolem extends AbstractGolem implements ContainerUser, Shearab
         super.readAdditionalSaveData(input);
         this.nextWeatheringTick = input.getLongOr("next_weather_age", -1L);
         this.setWeatherState(input.read("weather_state", WeatheringCopper.WeatherState.CODEC).orElse(WeatheringCopper.WeatherState.UNAFFECTED));
+        this.setSummoner(input.read("Purpur.Summoner", net.minecraft.core.UUIDUtil.CODEC).orElse(null)); // Purpur - Summoner API
     }
 
     @Override
