--- a/src/main/java/com/destroystokyo/paper/util/VersionFetcher.java
+++ b/src/main/java/com/destroystokyo/paper/util/VersionFetcher.java
@@ -28,6 +_,12 @@
      */
     Component getVersionMessage(String serverVersion);
 
+    // Purpur start
+    default int distance() {
+        return 0;
+    }
+    // Purpur end
+
     @ApiStatus.Internal
     class DummyVersionFetcher implements VersionFetcher {
 
