--- a/src/main/java/org/bukkit/inventory/RecipeChoice.java
+++ b/src/main/java/org/bukkit/inventory/RecipeChoice.java
@@ -274,6 +_,7 @@
     final class ExactChoice implements RecipeChoice {
 
         private List<ItemStack> choices;
+        private @org.jspecify.annotations.Nullable Predicate<ItemStack> predicate; // Purpur - Add predicate to recipe's ExactChoice ingredient
 
         /**
          * @deprecated Use {@link RecipeChoice#exactChoice(ItemStack, ItemStack...)} instead
@@ -334,6 +_,7 @@
 
         @Override
         public boolean test(ItemStack t) {
+            if (predicate != null) return predicate.test(t); // Purpur - Add predicate to recipe's ExactChoice ingredient
             for (ItemStack match : choices) {
                 if (t.isSimilar(match)) {
                     return true;
@@ -342,6 +_,25 @@
 
             return false;
         }
+
+        // Purpur start - Add predicate to recipe's ExactChoice ingredient
+        /**
+         * @deprecated Use {@link RecipeChoice#predicateChoice(Predicate, ItemStack)} instead
+         */
+        @Deprecated(since = "26.2", forRemoval = true)
+        @org.jspecify.annotations.Nullable
+        public Predicate<ItemStack> getPredicate() {
+            return predicate;
+        }
+
+        /**
+         * @deprecated Use {@link RecipeChoice#predicateChoice(Predicate, ItemStack)} instead
+         */
+        @Deprecated(since = "26.2", forRemoval = true)
+        public void setPredicate(@org.jspecify.annotations.Nullable Predicate<ItemStack> predicate) {
+            this.predicate = predicate;
+        }
+        // Purpur end - Add predicate to recipe's ExactChoice ingredient
 
         @Override
         public int hashCode() {
