--- a/src/main/java/org/bukkit/inventory/RecipeChoice.java
+++ b/src/main/java/org/bukkit/inventory/RecipeChoice.java
@@ -227,6 +_,7 @@
     final class ExactChoice implements RecipeChoice {
 
         private List<ItemStack> choices;
+        private Predicate<ItemStack> predicate; // Purpur - Add predicate to recipe's ExactChoice ingredient
 
         public ExactChoice(ItemStack stack) {
             this(Arrays.asList(stack));
@@ -275,6 +_,7 @@
 
         @Override
         public boolean test(ItemStack t) {
+            if (predicate != null) return predicate.test(t); // Purpur - Add predicate to recipe's ExactChoice ingredient
             for (ItemStack match : choices) {
                 if (t.isSimilar(match)) {
                     return true;
@@ -283,6 +_,17 @@
 
             return false;
         }
+
+        // Purpur start - Add predicate to recipe's ExactChoice ingredient
+        @org.jetbrains.annotations.Nullable
+        public Predicate<ItemStack> getPredicate() {
+            return predicate;
+        }
+
+        public void setPredicate(@org.jetbrains.annotations.Nullable Predicate<ItemStack> predicate) {
+            this.predicate = predicate;
+        }
+        // Purpur end - Add predicate to recipe's ExactChoice ingredient
 
         @Override
         public int hashCode() {
