From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java b/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
index 4397ee2f9006ae9b0e0ff6edf001ec06675a10aa..a8bdf295c83dd0cabc4f85dfbee6d7271dadbe1b 100644
--- a/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
+++ b/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
@@ -53,7 +53,7 @@ public class InteractWithDoor {
                             Node toNode = path.getNextNode();
                             BlockPos fromPos = fromNode.asBlockPos();
                             BlockState fromState = level.getBlockState(fromPos);
-                            if (fromState.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock)) {
+                            if (fromState.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock)&& !DoorBlock.requiresRedstone(body.level(), fromState, fromPos)) { // Purpur - Option to make doors require redstone
                                 DoorBlock fromBlock = (DoorBlock)fromState.getBlock();
                                 if (!fromBlock.isOpen(fromState)) {
                                     // CraftBukkit start - entities opening doors
@@ -70,7 +70,7 @@ public class InteractWithDoor {
 
                             BlockPos toPos = toNode.asBlockPos();
                             BlockState toState = level.getBlockState(toPos);
-                            if (toState.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock)) {
+                            if (toState.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock) && !DoorBlock.requiresRedstone(body.level(), toState, toPos)) { // Purpur - Option to make doors require redstone
                                 DoorBlock door = (DoorBlock)toState.getBlock();
                                 if (!door.isOpen(toState)) {
                                     // CraftBukkit start - entities opening doors
@@ -117,7 +117,7 @@ public class InteractWithDoor {
                     iterator.remove();
                 } else {
                     BlockState state = level.getBlockState(doorPos);
-                    if (!state.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock)) {
+                    if (!state.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock) || DoorBlock.requiresRedstone(body.level(), state, doorPos)) { // Purpur - Option to make doors require redstone
                         iterator.remove();
                     } else {
                         DoorBlock block = (DoorBlock)state.getBlock();
