--- a/net/minecraft/world/entity/ai/behavior/TransportItemsBetweenContainers.java
+++ b/net/minecraft/world/entity/ai/behavior/TransportItemsBetweenContainers.java
@@ -285,7 +_,7 @@
             LevelChunk chunkNow = level.getChunkSource().getChunkNow(chunkPos.x, chunkPos.z);
             if (chunkNow != null) {
                 for (BlockEntity blockEntity : chunkNow.getBlockEntities().values()) {
-                    if (blockEntity instanceof ChestBlockEntity chestBlockEntity) {
+                    if (blockEntity instanceof net.minecraft.world.level.block.entity.BaseContainerBlockEntity chestBlockEntity) { // Purpur - copper golem can place items in barrels or shulkers option
                         double d1 = chestBlockEntity.getBlockPos().distToCenterSqr(mob.position());
                         if (d1 < d) {
                             TransportItemsBetweenContainers.TransportItemTarget transportItemTarget1 = this.isTargetValidToPick(
@@ -369,7 +_,11 @@
     }
 
     private boolean isTargetBlocked(Level level, TransportItemsBetweenContainers.TransportItemTarget target) {
-        return ChestBlock.isChestBlockedAt(level, target.pos);
+        // Purpur start - copper golem can place items in barrels or shulkers option
+        boolean isBarrelBlocked = !level.purpurConfig.copperGolemCanOpenBarrel && target.state.is(net.minecraft.world.level.block.Blocks.BARREL);
+        boolean isShulkerBlocked = !level.purpurConfig.copperGolemCanOpenShulker && target.blockEntity instanceof net.minecraft.world.level.block.entity.ShulkerBoxBlockEntity shulkerBoxBlockEntity && !net.minecraft.world.level.block.ShulkerBoxBlock.canOpen(target.state, level, target.pos, shulkerBoxBlockEntity);
+        return target.state.is(net.minecraft.world.level.block.Blocks.BARREL) ? isBarrelBlocked : isShulkerBlocked || net.minecraft.world.level.block.ChestBlock.isChestBlockedAt(level, target.pos);
+        // Purpur end - copper golem can place items in barrels or shulkers option
     }
 
     private boolean targetHasNotChanged(Level level, TransportItemsBetweenContainers.TransportItemTarget target) {
@@ -446,7 +_,7 @@
     }
 
     private boolean isWantedBlock(PathfinderMob mob, BlockState state) {
-        return isPickingUpItems(mob) ? this.sourceBlockType.test(state) : this.destinationBlockType.test(state);
+        return isPickingUpItems(mob) ? this.sourceBlockType.test(state) : (mob.level().purpurConfig.copperGolemCanOpenBarrel && state.is(net.minecraft.world.level.block.Blocks.BARREL)) || (mob.level().purpurConfig.copperGolemCanOpenShulker && state.is(net.minecraft.tags.BlockTags.SHULKER_BOXES)) || this.destinationBlockType.test(state); // Purpur - copper golem can place items in barrels or shulkers option
     }
 
     private static double getInteractionRange(PathfinderMob mob) {
@@ -525,7 +_,7 @@
         int i = 0;
 
         for (ItemStack itemStack : container) {
-            if (!itemStack.isEmpty()) {
+            if (!itemStack.isEmpty() && (!(container instanceof net.minecraft.world.level.block.entity.ShulkerBoxBlockEntity) || !itemStack.is(net.minecraft.tags.ItemTags.SHULKER_BOXES))) { // Purpur - copper golem can place items in barrels or shulkers option
                 int min = Math.min(itemStack.getCount(), 16);
                 return container.removeItem(i, min);
             }
