--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2968,4 +_,15 @@
     public static Server.Spigot spigot() {
         return server.spigot();
     }
+
+    // Purpur start - Bring back server name
+    /**
+     * Get the name of this server
+     * @return the name of the server
+     */
+    @NotNull
+    public static String getServerName() {
+        return server.getServerName();
+    }
+    // Purpur end - Bring back server name
 }
