--- a/net/minecraft/world/food/FoodProperties.java
+++ b/net/minecraft/world/food/FoodProperties.java
@@ -42,9 +_,11 @@
         level.playSound(null, user.getX(), user.getY(), user.getZ(), consumable.sound().value(), SoundSource.NEUTRAL, 1.0F, random.triangle(1.0F, 0.4F));
         if (user instanceof Player player) {
             player.getFoodData().eat(this, stack, (net.minecraft.server.level.ServerPlayer) player); // CraftBukkit
+            // Purpur start - Burp delay
+            if (player.level().purpurConfig.playerBurpDelay > 0) {
+                player.burpDelay = player.level().purpurConfig.playerBurpDelay;
+                return;
+            }
+            // Purpur end - Burp delay
             level.playSound(
                 null, player.getX(), player.getY(), player.getZ(), SoundEvents.PLAYER_BURP, SoundSource.PLAYERS, 0.5F, Mth.randomBetween(random, 0.9F, 1.0F)
             );
         }
     }
 
