--- a/net/minecraft/world/entity/animal/fox/Fox.java
+++ b/net/minecraft/world/entity/animal/fox/Fox.java
@@ -370,6 +_,11 @@
     }
 
     private void setTargetGoals() {
+        // Purpur start - Tulips change fox type - do not add duplicate goals
+        this.targetSelector.removeGoal(this.landTargetGoal);
+        this.targetSelector.removeGoal(this.turtleEggTargetGoal);
+        this.targetSelector.removeGoal(this.fishTargetGoal);
+        // Purpur end - Tulips change fox type
         if (this.getVariant() == Fox.Variant.RED) {
             this.targetSelector.addGoal(4, this.landTargetGoal);
             this.targetSelector.addGoal(4, this.turtleEggTargetGoal);
@@ -397,6 +_,7 @@
 
     public void setVariant(final Fox.Variant variant) {
         this.entityData.set(DATA_TYPE_ID, variant.getId());
+        this.setTargetGoals(); // Purpur - Tulips change fox type - fix API bug not updating pathfinders on type change
     }
 
     @Override
@@ -715,6 +_,29 @@
     private boolean trusts(final LivingEntity entity) {
         return this.getTrustedEntities().anyMatch(trusted -> trusted.matches(entity));
     }
+
+    // Purpur start - Tulips change fox type
+    @Override
+    public net.minecraft.world.InteractionResult mobInteract(Player player, net.minecraft.world.InteractionHand hand) {
+        if (level().purpurConfig.foxTypeChangesWithTulips) {
+            ItemStack itemstack = player.getItemInHand(hand);
+            if (getVariant() == Variant.RED && itemstack.getItem() == Items.WHITE_TULIP) {
+                setVariant(Variant.SNOW);
+                if (!player.getAbilities().instabuild) {
+                    itemstack.shrink(1);
+                }
+                return net.minecraft.world.InteractionResult.SUCCESS;
+            } else if (getVariant() == Variant.SNOW && itemstack.getItem() == Items.ORANGE_TULIP) {
+                setVariant(Variant.RED);
+                if (!player.getAbilities().instabuild) {
+                    itemstack.shrink(1);
+                }
+                return net.minecraft.world.InteractionResult.SUCCESS;
+            }
+        }
+        return super.mobInteract(player, hand);
+    }
+    // Purpur end - Tulips change fox type
 
     @Override
     // Paper start - Fix MC-153010 (dropCustomDeathLoot is only called when shouldDropLoot = true)
