From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/block/BaseFireBlock.java b/net/minecraft/world/level/block/BaseFireBlock.java
index 6018fda8f5fb1c4a8c8e3b6e0d7e2b9c8cd84487..054c33c7eed5295220828a04673184f2e9a0d245 100644
--- a/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/net/minecraft/world/level/block/BaseFireBlock.java
@@ -215,7 +215,7 @@ public abstract class BaseFireBlock extends Block {
             boolean hasObsidian = false;
 
             for (Direction face : Direction.values()) {
-                if (level.getBlockState(testPos.set(pos).move(face)).is(Blocks.OBSIDIAN)) {
+                if (PortalShape.FRAME.test(level.getBlockState(testPos.set(pos).move(face)), level, testPos)) { // Purpur - Crying obsidian valid for portal frames
                     hasObsidian = true;
                     break;
                 }
