From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/damagesource/CombatTracker.java b/net/minecraft/world/damagesource/CombatTracker.java
index dd9fb54f50ebfeef71dcec0db76618d4916888e1..e78fdda864d951c3ab7c747070c526ea7b1a9cb4 100644
--- a/net/minecraft/world/damagesource/CombatTracker.java
+++ b/net/minecraft/world/damagesource/CombatTracker.java
@@ -75,7 +75,7 @@ public class CombatTracker {
         final Entity attackerEntity, final Component attackerName, final String messageWithItem, final String messageWithoutItem
     ) {
         ItemStack attackerItem = attackerEntity instanceof LivingEntity livingEntity ? livingEntity.getMainHandItem() : ItemStack.EMPTY;
-        return !attackerItem.isEmpty() && attackerItem.has(DataComponents.CUSTOM_NAME)
+        return !attackerItem.isEmpty() && (org.purpurmc.purpur.PurpurConfig.playerDeathsAlwaysShowItem || attackerItem.has(DataComponents.CUSTOM_NAME)) // Purpur - always show item in player death messages
             ? Component.translatable(messageWithItem, this.mob.getDisplayName(), attackerName, attackerItem.getDisplayName())
             : Component.translatable(messageWithoutItem, this.mob.getDisplayName(), attackerName);
     }
@@ -118,6 +118,15 @@ public class CombatTracker {
                 Component link = ComponentUtils.wrapInSquareBrackets(Component.translatable(deathMsg + ".link")).withStyle(INTENTIONAL_GAME_DESIGN_STYLE);
                 return Component.translatable(deathMsg + ".message", this.mob.getDisplayName(), link);
             } else {
+                // Purpur start - Dont run with scissors!
+                if (killingSource.isScissors()) {
+                    return killingSource.getLocalizedDeathMessage(org.purpurmc.purpur.PurpurConfig.deathMsgRunWithScissors, this.mob);
+                // Purpur start - Stonecutter damage
+                } else if (killingSource.isStonecutter()) {
+                    return killingSource.getLocalizedDeathMessage(org.purpurmc.purpur.PurpurConfig.deathMsgStonecutter, this.mob);
+                // Purpur end - Stonecutter damage
+                }
+                // Purpur end - Dont run with scissors!
                 return killingSource.getLocalizedDeathMessage(this.mob);
             }
         }
