--- a/net/minecraft/world/damagesource/CombatRules.java
+++ b/net/minecraft/world/damagesource/CombatRules.java
@@ -17,7 +_,7 @@
         final LivingEntity victim, final float damage, final DamageSource source, final float totalArmor, final float armorToughness
     ) {
         float toughness = 2.0F + armorToughness / 4.0F;
-        float realArmor = Mth.clamp(totalArmor - damage / toughness, totalArmor * 0.2F, 20.0F);
+        float realArmor = Mth.clamp(totalArmor - damage / toughness, totalArmor * 0.2F, org.purpurmc.purpur.PurpurConfig.limitArmor ? 20F : Float.MAX_VALUE); // Purpur - Add attribute clamping and armor limit config
         float armorFraction = realArmor / 25.0F;
         ItemStack weaponItem = source.getWeaponItem();
         float modifiedArmorFraction;
@@ -32,7 +_,7 @@
     }
 
     public static float getDamageAfterMagicAbsorb(final float damage, final float totalMagicArmor) {
-        float realArmor = Mth.clamp(totalMagicArmor, 0.0F, 20.0F);
+        float realArmor = Mth.clamp(totalMagicArmor, 0.0F, org.purpurmc.purpur.PurpurConfig.limitArmor ? 20F : Float.MAX_VALUE); // Purpur - Add attribute clamping and armor limit config
         return damage * (1.0F - realArmor / 25.0F);
     }
 }
