From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/damagesource/DamageSource.java b/net/minecraft/world/damagesource/DamageSource.java
index 5afbe6ed28474555ef6ee15a53206eea4b16117e..2baac450cb405403caed3183a715301848a48772 100644
--- a/net/minecraft/world/damagesource/DamageSource.java
+++ b/net/minecraft/world/damagesource/DamageSource.java
@@ -24,6 +24,8 @@ public class DamageSource {
     private org.bukkit.block.@Nullable Block eventBlockDamager; // Relevant block set. damageSourcePosition is only used for bad respawn point explosion or custom damage
     private org.bukkit.block.@Nullable BlockState fromBlockSnapshot; // Captured block snapshot when the eventBlockDamager is not relevant (e.g. for bad respawn point explosions the block is already removed)
     private boolean critical; // Supports arrows and sweeping damage
+    private boolean scissors = false; // Purpur - Dont run with scissors!
+    private boolean stonecutter = false; // Purpur - Stonecutter damage
 
     public DamageSource knownCause(final org.bukkit.event.entity.EntityDamageEvent.DamageCause cause) {
         final DamageSource damageSource = this.copy();
@@ -35,6 +37,30 @@ public class DamageSource {
         return this.knownCause;
     }
 
+    // Purpur start - Dont run with scissors!
+    public DamageSource scissors() {
+        this.knownCause(org.bukkit.event.entity.EntityDamageEvent.DamageCause.SUICIDE);
+        this.scissors = true;
+        return this;
+    }
+
+    public boolean isScissors() {
+        return this.scissors;
+    }
+    // Purpur end - Dont run with scissors!
+
+    // Purpur start -  - Stonecutter damage
+    public DamageSource stonecutter() {
+        this.knownCause(org.bukkit.event.entity.EntityDamageEvent.DamageCause.CONTACT);
+        this.stonecutter = true;
+        return this;
+    }
+
+    public boolean isStonecutter() {
+        return this.stonecutter;
+    }
+    // Purpur end - Stonecutter damage
+
     @Nullable
     public Entity eventEntityDamager() {
         return this.eventEntityDamager;
@@ -94,6 +120,8 @@ public class DamageSource {
         damageSource.eventBlockDamager = this.eventBlockDamager;
         damageSource.fromBlockSnapshot = this.fromBlockSnapshot;
         damageSource.critical = this.critical;
+        damageSource.scissors = this.isScissors(); // Purpur - Dont run with scissors!
+        damageSource.stonecutter = this.isStonecutter(); // Purpur - Stonecutter damage
         return damageSource;
     }
     // CraftBukkit end
@@ -159,12 +187,21 @@ public class DamageSource {
         } else {
             Component name = this.causingEntity == null ? this.directEntity.getDisplayName() : this.causingEntity.getDisplayName();
             ItemStack held = this.causingEntity instanceof LivingEntity livingEntity ? livingEntity.getMainHandItem() : ItemStack.EMPTY;
-            return !held.isEmpty() && held.has(DataComponents.CUSTOM_NAME)
+            return !held.isEmpty() && (org.purpurmc.purpur.PurpurConfig.playerDeathsAlwaysShowItem || held.has(DataComponents.CUSTOM_NAME)) // Purpur - always show item in player death messages
                 ? Component.translatable(deathMsg + ".item", victim.getDisplayName(), name, held.getDisplayName())
                 : Component.translatable(deathMsg, victim.getDisplayName(), name);
         }
     }
 
+    // Purpur start - Component related conveniences
+    public Component getLocalizedDeathMessage(String str, LivingEntity entity) {
+        net.kyori.adventure.text.Component name = io.papermc.paper.adventure.PaperAdventure.asAdventure(entity.getDisplayName());
+        net.kyori.adventure.text.minimessage.tag.resolver.TagResolver template = net.kyori.adventure.text.minimessage.tag.resolver.Placeholder.component("player", name);
+        net.kyori.adventure.text.Component component = net.kyori.adventure.text.minimessage.MiniMessage.miniMessage().deserialize(str, template);
+        return io.papermc.paper.adventure.PaperAdventure.asVanilla(component);
+    }
+    // Purpur end - Component related conveniences
+
     public String getMsgId() {
         return this.type().msgId();
     }
