--- a/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
@@ -310,7 +_,8 @@
         WORLD_BORDER,
         /**
          * Damage caused when an entity contacts a block such as a Cactus,
-         * Dripstone (Stalagmite) or Berry Bush.
+         * Dripstone (Stalagmite) or Berry Bush. (Stonecutters too if you
+         * have the Stonecutter damage Purpur feature enabled)
          * <p>
          * Damage: variable
          */
