From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/entity/monster/Monster.java b/net/minecraft/world/entity/monster/Monster.java
index 28bab5bc11d98d8836b33d06524549e6e3d15acf..81f1d4faebc38aa07a2c76c2c163584c8e5b624c 100644
--- a/net/minecraft/world/entity/monster/Monster.java
+++ b/net/minecraft/world/entity/monster/Monster.java
@@ -83,6 +83,11 @@ public abstract class Monster extends PathfinderMob implements Enemy {
     }
 
     public static boolean isDarkEnoughToSpawn(final ServerLevelAccessor level, final BlockPos pos, final RandomSource random) {
+        // Purpur start - Config to disable hostile mob spawn on ice
+        if (canSpawnInBlueAndPackedIce(level, pos)) {
+            return false;
+        }
+        // Purpur end - Config to disable hostile mob spawn on ice
         if (level.getBrightness(LightLayer.SKY, pos) > random.nextInt(32)) {
             return false;
         }
@@ -111,6 +116,11 @@ public abstract class Monster extends PathfinderMob implements Enemy {
     public static boolean checkAnyLightMonsterSpawnRules(
         final EntityType<? extends Monster> type, final LevelAccessor level, final EntitySpawnReason spawnReason, final BlockPos pos, final RandomSource random
     ) {
+        // Purpur start - Config to disable hostile mob spawn on ice
+        if (canSpawnInBlueAndPackedIce(level, pos)) {
+            return false;
+        }
+        // Purpur end - Config to disable hostile mob spawn on ice
         return checkMobSpawnRules(type, level, spawnReason, pos, random);
     }
 
@@ -152,4 +162,12 @@ public abstract class Monster extends PathfinderMob implements Enemy {
             return ItemStack.EMPTY;
         }
     }
+
+    // Purpur start - Config to disable hostile mob spawn on ice
+    public static boolean canSpawnInBlueAndPackedIce(LevelAccessor level, BlockPos pos) {
+        net.minecraft.world.level.block.state.BlockState spawnBlock = level.getBlockState(pos.below());
+
+        return (!level.getMinecraftWorld().purpurConfig.mobsSpawnOnPackedIce && spawnBlock.is(net.minecraft.world.level.block.Blocks.PACKED_ICE)) || (!level.getMinecraftWorld().purpurConfig.mobsSpawnOnBlueIce && spawnBlock.is(net.minecraft.world.level.block.Blocks.BLUE_ICE));
+    }
+    // Purpur end - Config to disable hostile mob spawn on ice
 }
