--- a/net/minecraft/world/entity/monster/ZombifiedPiglin.java
+++ b/net/minecraft/world/entity/monster/ZombifiedPiglin.java
@@ -112,6 +_,12 @@
             this.maybeAlertOthers();
         }
 
+        // Purpur start - Toggle for Zombified Piglin death always counting as player kill when angry
+        if (this.isAngry() && this.level().purpurConfig.zombifiedPiglinCountAsPlayerKillWhenAngry) {
+            this.lastHurtByPlayerMemoryTime = this.tickCount;
+        }
+        // Purpur end - Toggle for Zombified Piglin death always counting as player kill when angry
+
         super.customServerAiStep(level);
     }
 
@@ -158,6 +_,12 @@
             this.playFirstAngerSoundIn = FIRST_ANGER_SOUND_DELAY.sample(this.random);
             this.ticksUntilNextAlert = ALERT_INTERVAL.sample(this.random);
         }
+
+        // Purpur start - Toggle for Zombified Piglin death always counting as player kill when angry
+        if (livingEntity instanceof Player player && this.level().purpurConfig.zombifiedPiglinCountAsPlayerKillWhenAngry) {
+            this.setLastHurtByPlayer(player, this.tickCount);
+        }
+        // Purpur end - Toggle for Zombified Piglin death always counting as player kill when angry
 
         return super.setTarget(livingEntity, reason); // CraftBukkit
     }
