From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mickey42302 <bgoodwin423@gmail.com>
Date: Sun, 13 Jul 2025 00:10:53 -0400
Subject: [PATCH] Add permission node for spam bypass


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index a63ade75461b68a780c56bfb5ff7c61f99f76744..5795bee75ea308b59af92e92668f787072ac3608 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -844,7 +844,7 @@ public class ServerGamePacketListenerImpl
     public void handleCustomCommandSuggestions(ServerboundCommandSuggestionPacket packet) {
         // PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level()); // Paper - AsyncTabCompleteEvent; run this async
         // CraftBukkit start
-        if (!this.tabSpamThrottler.isIncrementAndUnderThreshold() && !this.server.getPlayerList().isOp(this.player.getGameProfile()) && !this.server.isSingleplayerOwner(this.player.getGameProfile())) { // Paper - configurable tab spam limits
+        if (!this.tabSpamThrottler.isIncrementAndUnderThreshold() && !this.server.getPlayerList().isOp(this.player.getGameProfile()) && !this.server.isSingleplayerOwner(this.player.getGameProfile()) && !(this.server.getPlayerList().getPlayer(this.player.getUUID())).getBukkitEntity().hasPermission("minecraft.spam.bypass")) { // Paper - configurable tab spam limits
             this.disconnectAsync(Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM); // Paper - Kick event cause // Paper - add proper async disconnect
             return;
         }
@@ -2630,6 +2630,7 @@ public class ServerGamePacketListenerImpl
         if (!this.chatSpamThrottler.isIncrementAndUnderThreshold()
             // CraftBukkit end
             && !this.server.getPlayerList().isOp(this.player.getGameProfile())
+            && !this.server.getPlayerList().getPlayer(this.player.getUUID()).getBukkitEntity().hasPermission("minecraft.spam.bypass")
             && !this.server.isSingleplayerOwner(this.player.getGameProfile())) {
             this.disconnectAsync(Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM); // Paper - kick event cause & add proper async disconnect
         }
