From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mickey42302 <bgoodwin423@gmail.com>
Date: Sun, 13 Jul 2025 02:26:42 -0400
Subject: [PATCH] Add permission nodes for movement bypass


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index a63ade75461b68a780c56bfb5ff7c61f99f76744..ce21dd710b6309e8ad6afbd608b9b988133df1e4 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -609,6 +609,9 @@ public class ServerGamePacketListenerImpl
                 // Paper end - Prevent moving into unloaded chunks
                 if (d7 - d6 > Math.max(100.0, Mth.square(org.spigotmc.SpigotConfig.movedTooQuicklyMultiplier * (float) i * speed)) && !this.isSingleplayerOwner()) {
                     // CraftBukkit end
+                    if (this.player.getBukkitEntity().hasPermission("minecraft.movement.bypass.vehicle")) {
+                        return;
+                    }
                     LOGGER.warn(
                         "{} (vehicle of {}) moved too quickly! {},{},{}", rootVehicle.getName().getString(), this.player.getName().getString(), d3, d4, d5
                     );
@@ -638,6 +641,9 @@ public class ServerGamePacketListenerImpl
                 d7 = d3 * d3 + d4 * d4 + d5 * d5;
                 boolean flag1 = false;
                 if (d7 > org.spigotmc.SpigotConfig.movedWronglyThreshold) { // Spigot
+                    if (this.player.getBukkitEntity().hasPermission("minecraft.movement.bypass.vehicle")) {
+                        return;
+                    }
                     flag1 = true; // Paper - diff on change, this should be moved wrongly
                     LOGGER.warn("{} (vehicle of {}) moved wrongly! {}", rootVehicle.getName().getString(), this.player.getName().getString(), Math.sqrt(d7));
                 }
@@ -1555,6 +1561,9 @@ public class ServerGamePacketListenerImpl
                                         float f2 = isFallFlying ? 300.0F : 100.0F;
                                         if (d7 - d6 > Math.max(f2, Mth.square(org.spigotmc.SpigotConfig.movedTooQuicklyMultiplier * (float) i * speed))) {
                                             // CraftBukkit end
+                                            if (this.player.getBukkitEntity().hasPermission("minecraft.movement.bypass.client")) {
+                                                return;
+                                            }
                                             // Paper start - Add fail move event
                                             io.papermc.paper.event.player.PlayerFailMoveEvent event = fireFailMove(io.papermc.paper.event.player.PlayerFailMoveEvent.FailReason.MOVED_TOO_QUICKLY,
                                                     toX, toY, toZ, toYaw, toPitch, true);
@@ -1631,7 +1640,8 @@ public class ServerGamePacketListenerImpl
                                     && d7 > org.spigotmc.SpigotConfig.movedWronglyThreshold // Spigot
                                     && !this.player.isSleeping()
                                     && !this.player.isCreative()
-                                    && !this.player.isSpectator()) {
+                                    && !this.player.isSpectator()
+                                    && !this.player.getBukkitEntity().hasPermission("minecraft.movement.bypass.client")) {
                                     // Paper start - Add fail move event
                                     io.papermc.paper.event.player.PlayerFailMoveEvent event = fireFailMove(io.papermc.paper.event.player.PlayerFailMoveEvent.FailReason.MOVED_WRONGLY,
                                             toX, toY, toZ, toYaw, toPitch, true);
