From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mickey42302 <bgoodwin423@gmail.com>
Date: Sun, 13 Jul 2025 22:28:28 -0400
Subject: [PATCH] Add bypass permission in vanilla for spawn protection


diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 670553243d26e2faab8a21f099a846d4d1df7927..84ddd5f8632f3621a964470e5879386e48bcb43f 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -543,6 +543,8 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             return false;
         } else if (this.getSpawnProtectionRadius() <= 0) {
             return false;
+        } else if (player instanceof ServerPlayer serverPlayer && serverPlayer.getBukkitEntity().hasPermission("minecraft.spawn-protection.bypass")) {
+            return false;
         } else {
             BlockPos sharedSpawnPos = level.getSharedSpawnPos();
             int abs = Mth.abs(pos.getX() - sharedSpawnPos.getX());
