From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mickey42302 <bgoodwin423@gmail.com>
Date: Mon, 20 Jul 2026 21:30:50 -0400
Subject: [PATCH] Add spectator teleport permission node


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index d731cea2005dff7be6e8e82d602b11355a143ec7..b0814a0189eac6d7f3dc8302323c624ee8e9141d 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2328,6 +2328,7 @@ public class ServerGamePacketListenerImpl
     public void handleTeleportToEntityPacket(final ServerboundTeleportToEntityPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
         if (this.player.isSpectator()) {
+            if (!this.player.getBukkitEntity().hasPermission("purpur.spectator-can-teleport-to-players")) return; // Purpur - Require permission for Spectator to teleport to players
             for (ServerLevel level : this.server.getAllLevels()) {
                 Entity entity = packet.getEntity(level);
                 if (entity != null) {
