From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: granny <granny@purpurmc.org>
Date: Sun, 27 Aug 2023 02:09:51 -0700
Subject: [PATCH] lightning transforms blocks


diff --git a/net/minecraft/world/entity/LightningBolt.java b/net/minecraft/world/entity/LightningBolt.java
index e81eabf901e755e1254fff98cdd1e609cf07aab8..df7166e22bdda74c4e85359ab54d383c54e2a73f 100644
--- a/net/minecraft/world/entity/LightningBolt.java
+++ b/net/minecraft/world/entity/LightningBolt.java
@@ -38,6 +38,9 @@ public class LightningBolt extends Entity {
     public boolean visualOnly;
     private @Nullable ServerPlayer cause;
     private final Set<Entity> hitEntities = Sets.newHashSet();
+    // Purpur start - lightning transforms blocks
+    public List<org.bukkit.craftbukkit.block.CraftBlockState> blocks;
+    // Purpur end - lightning transforms blocks
     private int blocksSetOnFire;
     public boolean isEffect; // Paper - Properly handle lightning effects api
 
@@ -46,6 +49,7 @@ public class LightningBolt extends Entity {
         this.life = 2;
         this.seed = this.random.nextLong();
         this.flashes = this.random.nextInt(3) + 1;
+        this.blocks = new java.util.ArrayList<>(); // Purpur - lightning transforms blocks
     }
 
     public void setVisualOnly(boolean visualOnly) {
@@ -73,6 +77,67 @@ public class LightningBolt extends Entity {
         }
     }
 
+    // Purpur start - lightning transforms blocks
+    public void initBlocks() {
+        this.blocks = this.getTransformedBlocks();
+    }
+
+    private List<org.bukkit.craftbukkit.block.CraftBlockState> getTransformedBlocks() {
+        if (!level().purpurConfig.lightningTransformsBlocks) return new java.util.ArrayList<>();
+
+        BlockPos blockposition = this.getStrikePosition();
+        BlockState iblockdata = this.level().getBlockState(blockposition);
+        BlockPos pos;
+        BlockState iblockdata2;
+
+        if (iblockdata.is(net.minecraft.world.level.block.Blocks.LIGHTNING_ROD)) {
+            pos = blockposition.below();
+            iblockdata2 = this.level().getBlockState(pos);
+        } else {
+            iblockdata2 = iblockdata;
+            pos = blockposition;
+        }
+
+        BlockState targetState = level().purpurConfig.lightningBlockTransformMap.get(iblockdata2.getBlock());
+        if (targetState != null) {
+            level().setBlock(pos, targetState, 2);
+        }
+
+        if (!level().purpurConfig.lightningTransformsNearbyBlocks) {
+            return new java.util.ArrayList<>();
+        }
+
+        java.util.Map<net.minecraft.world.level.block.Block, BlockState> transformMap = level().purpurConfig.lightningBlockTransformMap;
+        org.bukkit.craftbukkit.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.util.BlockStateListPopulator(level());
+        BlockPos.breadthFirstTraversal(
+                pos,
+                level().purpurConfig.lightningNearbyBlocksMaxDepth,
+                level().purpurConfig.lightningNearbyBlocksMaxIterations,
+                (validPos, queueAdder) -> {
+                    for (net.minecraft.core.Direction direction : net.minecraft.core.Direction.values()) {
+                        queueAdder.accept(validPos.relative(direction));
+                    }
+                },
+        blockPos -> {
+                    if (blockPos.equals(pos)) {
+                        return BlockPos.TraversalNodeStatus.ACCEPT;
+                    } else {
+                        BlockState nearbyState = blockList.getBlockState(blockPos);
+                        BlockState nearbyTarget = transformMap.get(nearbyState.getBlock());
+                        if (nearbyTarget == null) {
+                            return BlockPos.TraversalNodeStatus.SKIP;
+                        } else {
+                            blockList.setBlock(blockPos, nearbyTarget, 3);
+                            return BlockPos.TraversalNodeStatus.ACCEPT;
+                        }
+                    }
+                }
+            );
+
+        return blockList.getSnapshotBlocks();
+    }
+    // Purpur end - lightning transforms blocks
+
     @Override
     public void tick() {
         super.tick();
@@ -107,6 +172,7 @@ public class LightningBolt extends Entity {
                 }
 
                 this.powerLightningRod();
+                if (level().purpurConfig.lightningTransformsBlocks) net.minecraft.world.level.block.SpongeBlock.processBlocksBreadthFirstSearch(level(), this.getStrikePosition(), this.blocks, false); // Purpur - lightning transforms blocks
                 clearCopperOnLightningStrike(this.level(), this.getStrikePosition(), this); // Paper - Call EntityChangeBlockEvent
                 this.gameEvent(GameEvent.LIGHTNING_STRIKE);
             }
diff --git a/net/minecraft/world/level/block/SpongeBlock.java b/net/minecraft/world/level/block/SpongeBlock.java
index 0ce317ad5f3f7a08914d04ace2259f7fdb6f906e..888dc057dfb7d619bc13d02dc1f769688d7743a9 100644
--- a/net/minecraft/world/level/block/SpongeBlock.java
+++ b/net/minecraft/world/level/block/SpongeBlock.java
@@ -102,13 +102,20 @@ public class SpongeBlock extends Block {
             );
         // CraftBukkit start
         java.util.List<org.bukkit.craftbukkit.block.CraftBlockState> snapshots = blockList.getSnapshotBlocks(); // Is a clone
+    // Purpur start - lightning transforms blocks
+        return processBlocksBreadthFirstSearch(level, pos, snapshots, true);
+    }
+    public static boolean processBlocksBreadthFirstSearch(net.minecraft.world.level.Level level, net.minecraft.core.BlockPos pos, java.util.List<org.bukkit.craftbukkit.block.CraftBlockState> snapshots, boolean emitSpongeAbscorbEvent) {
+    // Purpur end - lightning transforms blocks
         if (!snapshots.isEmpty()) {
             final org.bukkit.block.Block sponge = org.bukkit.craftbukkit.block.CraftBlock.at(level, pos);
 
+            if (emitSpongeAbscorbEvent) { // Purpur - lightning transforms blocks
             org.bukkit.event.block.SpongeAbsorbEvent event = new org.bukkit.event.block.SpongeAbsorbEvent(sponge, (java.util.List<org.bukkit.block.BlockState>) (java.util.List) snapshots);
             if (!event.callEvent()) {
                 return false;
             }
+            } // Purpur - lightning transforms blocks
 
             for (org.bukkit.craftbukkit.block.CraftBlockState snapshot : snapshots) {
                 BlockPos blockPos = snapshot.getPosition();
