From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mickey42302 <bgoodwin423@gmail.com>
Date: Sat, 12 Jul 2025 12:08:29 -0400
Subject: [PATCH] Permission nodes for more gamemaster blocks


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index a63ade75461b68a780c56bfb5ff7c61f99f76744..f6688fc752b47ef15c3baafcc73a91d48f9bf922 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1128,7 +1128,7 @@ public class ServerGamePacketListenerImpl
     @Override
     public void handleSetStructureBlock(ServerboundSetStructureBlockPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
-        if (this.player.canUseGameMasterBlocks()) {
+        if (this.player.canUseGameMasterBlocks() || this.player.getBukkitEntity().hasPermission("minecraft.structureblock")) {
             BlockPos pos = packet.getPos();
             BlockState blockState = this.player.level().getBlockState(pos);
             if (this.player.level().getBlockEntity(pos) instanceof StructureBlockEntity structureBlockEntity) {
@@ -1181,7 +1181,7 @@ public class ServerGamePacketListenerImpl
     @Override
     public void handleSetTestBlock(ServerboundSetTestBlockPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
-        if (this.player.canUseGameMasterBlocks()) {
+        if (this.player.canUseGameMasterBlocks() || this.player.getBukkitEntity().hasPermission("minecraft.testblock")) {
             BlockPos blockPos = packet.position();
             BlockState blockState = this.player.level().getBlockState(blockPos);
             if (this.player.level().getBlockEntity(blockPos) instanceof TestBlockEntity testBlockEntity) {
@@ -1197,7 +1197,7 @@ public class ServerGamePacketListenerImpl
     public void handleTestInstanceBlockAction(ServerboundTestInstanceBlockActionPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
         BlockPos blockPos = packet.pos();
-        if (this.player.canUseGameMasterBlocks() && this.player.level().getBlockEntity(blockPos) instanceof TestInstanceBlockEntity testInstanceBlockEntity) {
+        if ((this.player.canUseGameMasterBlocks() || this.player.getBukkitEntity().hasPermission("minecraft.testinstanceblock")) && this.player.level().getBlockEntity(blockPos) instanceof TestInstanceBlockEntity testInstanceBlockEntity) {
             if (packet.action() != ServerboundTestInstanceBlockActionPacket.Action.QUERY
                 && packet.action() != ServerboundTestInstanceBlockActionPacket.Action.INIT) {
                 testInstanceBlockEntity.set(packet.data());
@@ -1240,7 +1240,7 @@ public class ServerGamePacketListenerImpl
     @Override
     public void handleSetJigsawBlock(ServerboundSetJigsawBlockPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
-        if (this.player.canUseGameMasterBlocks()) {
+        if (this.player.canUseGameMasterBlocks() || this.player.getBukkitEntity().hasPermission("minecraft.jigsawblock")) {
             BlockPos pos = packet.getPos();
             BlockState blockState = this.player.level().getBlockState(pos);
             if (this.player.level().getBlockEntity(pos) instanceof JigsawBlockEntity jigsawBlockEntity) {
@@ -1260,7 +1260,7 @@ public class ServerGamePacketListenerImpl
     @Override
     public void handleJigsawGenerate(ServerboundJigsawGeneratePacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
-        if (this.player.canUseGameMasterBlocks()) {
+        if (this.player.canUseGameMasterBlocks() || this.player.getBukkitEntity().hasPermission("minecraft.jigsawblock")) {
             BlockPos pos = packet.getPos();
             if (this.player.level().getBlockEntity(pos) instanceof JigsawBlockEntity jigsawBlockEntity) {
                 jigsawBlockEntity.generate(this.player.level(), packet.levels(), packet.keepJigsaws());
diff --git a/net/minecraft/world/level/block/JigsawBlock.java b/net/minecraft/world/level/block/JigsawBlock.java
index 5b5b0ba3d132be18403fb763039ccc723e2b7b46..31e17b20aa2eacf7ae40607a49fa01542ce65d05 100644
--- a/net/minecraft/world/level/block/JigsawBlock.java
+++ b/net/minecraft/world/level/block/JigsawBlock.java
@@ -68,7 +68,7 @@ public class JigsawBlock extends Block implements EntityBlock, GameMasterBlock {
     @Override
     protected InteractionResult useWithoutItem(BlockState state, Level level, BlockPos pos, Player player, BlockHitResult hitResult) {
         BlockEntity blockEntity = level.getBlockEntity(pos);
-        if (blockEntity instanceof JigsawBlockEntity && player.canUseGameMasterBlocks()) {
+        if (blockEntity instanceof JigsawBlockEntity && player.canUseGameMasterBlocks()  || (player.isCreative() && player.getBukkitEntity().hasPermission("minecraft.jigsawblock"))) {
             player.openJigsawBlock((JigsawBlockEntity)blockEntity);
             return InteractionResult.SUCCESS;
         } else {
diff --git a/net/minecraft/world/level/block/LecternBlock.java b/net/minecraft/world/level/block/LecternBlock.java
index 5a9b601b7bf7e80b04ebd8f5c8b7d121031132c7..9fce9dc9e2350f16ed7d191b45b61de893e32dd0 100644
--- a/net/minecraft/world/level/block/LecternBlock.java
+++ b/net/minecraft/world/level/block/LecternBlock.java
@@ -82,7 +82,7 @@ public class LecternBlock extends BaseEntityBlock {
         ItemStack itemInHand = context.getItemInHand();
         Player player = context.getPlayer();
         boolean flag = false;
-        if (!level.isClientSide && player != null && player.canUseGameMasterBlocks()) {
+        if (!level.isClientSide && player != null && player.canUseGameMasterBlocks() || player.getBukkitEntity().hasPermission("minecraft.lecternblock")) {
             CustomData customData = itemInHand.getOrDefault(DataComponents.BLOCK_ENTITY_DATA, CustomData.EMPTY);
             if (customData.contains("Book")) {
                 flag = true;
diff --git a/net/minecraft/world/level/block/LightBlock.java b/net/minecraft/world/level/block/LightBlock.java
index 4e2c0383eda3e81bbc40a724263b49151a342746..902a0ee5f6d82d55a440440b22b60e6941630c2e 100644
--- a/net/minecraft/world/level/block/LightBlock.java
+++ b/net/minecraft/world/level/block/LightBlock.java
@@ -59,7 +59,7 @@ public class LightBlock extends Block implements SimpleWaterloggedBlock {
 
     @Override
     protected InteractionResult useWithoutItem(BlockState state, Level level, BlockPos pos, Player player, BlockHitResult hitResult) {
-        if (!level.isClientSide && player.canUseGameMasterBlocks()) {
+        if (!level.isClientSide && player.canUseGameMasterBlocks() || (player.isCreative() && player.getBukkitEntity().hasPermission("minecraft.lightblock"))) {
             level.setBlock(pos, state.cycle(LEVEL), 2);
             return InteractionResult.SUCCESS_SERVER;
         } else {
diff --git a/net/minecraft/world/level/block/TestBlock.java b/net/minecraft/world/level/block/TestBlock.java
index fdcc98da158af9d93431ce5dac127e1017c1dcfb..ec21cd7c18f0808daf769e557a61544fcf6e5de8 100644
--- a/net/minecraft/world/level/block/TestBlock.java
+++ b/net/minecraft/world/level/block/TestBlock.java
@@ -62,7 +62,7 @@ public class TestBlock extends BaseEntityBlock implements GameMasterBlock {
     @Override
     protected InteractionResult useWithoutItem(BlockState state, Level level, BlockPos pos, Player player, BlockHitResult hitResult) {
         if (level.getBlockEntity(pos) instanceof TestBlockEntity testBlockEntity) {
-            if (!player.canUseGameMasterBlocks()) {
+            if (!player.canUseGameMasterBlocks() && !player.getBukkitEntity().hasPermission("minecraft.testblock")) {
                 return InteractionResult.PASS;
             } else {
                 if (level.isClientSide) {
diff --git a/net/minecraft/world/level/block/TestInstanceBlock.java b/net/minecraft/world/level/block/TestInstanceBlock.java
index 3f586d506090db3cad7ff30c23ef0db9b83e3042..7168d664936d607260dbd42deb034aa3b4690f34 100644
--- a/net/minecraft/world/level/block/TestInstanceBlock.java
+++ b/net/minecraft/world/level/block/TestInstanceBlock.java
@@ -28,7 +28,7 @@ public class TestInstanceBlock extends BaseEntityBlock implements GameMasterBloc
     @Override
     protected InteractionResult useWithoutItem(BlockState state, Level level, BlockPos pos, Player player, BlockHitResult hitResult) {
         if (level.getBlockEntity(pos) instanceof TestInstanceBlockEntity testInstanceBlockEntity) {
-            if (!player.canUseGameMasterBlocks()) {
+            if (!player.canUseGameMasterBlocks() && !player.getBukkitEntity().hasPermission("minecraft.testinstanceblock")) {
                 return InteractionResult.PASS;
             } else {
                 if (player.level().isClientSide) {
diff --git a/net/minecraft/world/level/block/entity/StructureBlockEntity.java b/net/minecraft/world/level/block/entity/StructureBlockEntity.java
index c2c0915dc1458bb5755ee5591872da51c998582d..303baacd5cf5343eea6c71f48631aeaf985deb4a 100644
--- a/net/minecraft/world/level/block/entity/StructureBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/StructureBlockEntity.java
@@ -147,7 +147,7 @@ public class StructureBlockEntity extends BlockEntity implements BoundingBoxRend
     }
 
     public boolean usedBy(Player player) {
-        if (!player.canUseGameMasterBlocks()) {
+        if (!player.canUseGameMasterBlocks() || !(player.isCreative() && player.getBukkitEntity().hasPermission("minecraft.structureblock"))) {
             return false;
         } else {
             if (player.level().isClientSide) {
