From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mickey42302 <bgoodwin423@gmail.com>
Date: Mon, 20 Jul 2026 21:30:50 -0400
Subject: [PATCH] Add spectator teleport permission node


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index 1862df44a3319e459b547cebffb16972674856d6..74340c3ae87846b0610d994331916d130aa4274e 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -107,12 +107,16 @@ public class Main {
                 JvmProfiler.INSTANCE.start(Environment.SERVER);
             }
 
-            // Purpur start - Add toggle for enchant level clamping - load config files early
+            // Purpur - Add toggle for enchant level clamping - load config files early
             org.bukkit.configuration.file.YamlConfiguration purpurConfiguration = io.papermc.paper.configuration.PaperConfigurations.loadLegacyConfigFile((File) options.valueOf("purpur-settings"));
             org.purpurmc.purpur.PurpurConfig.clampEnchantLevels = purpurConfiguration.getBoolean("settings.enchantment.clamp-levels", true);
+
+            // Purpur - Add settings for enabling commands that are unavailable by default.
             org.purpurmc.purpur.PurpurConfig.registerMinecraftDebugCommands = purpurConfiguration.getBoolean("settings.register-minecraft-debug-commands"); // Purpur - register minecraft debug commands
             org.purpurmc.purpur.PurpurConfig.registerMinecraftDisabledCommands = purpurConfiguration.getBoolean("settings.register-minecraft-disabled-commands"); // Purpur - register disabled minecraft commands
-            // Purpur end - Add toggle for enchant level clamping - load config files early
+
+            // Purpur - Add setting which can be used to enable a permission node that controls the ability for spectators to teleport to other players.
+            org.purpurmc.purpur.PurpurConfig.requirePermissionSpectatorTeleportToPlayers = purpurConfiguration.getBoolean("settings.require-permission-spectator-teleport-to-players"); // Purpur - require permission spectator teleport to players
 
             io.papermc.paper.plugin.PluginInitializerManager.load(options); // Paper
             Bootstrap.bootStrap();
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index d731cea2005dff7be6e8e82d602b11355a143ec7..6858bfce9464251025fddc825f32e686df394f9d 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2328,11 +2328,14 @@ public class ServerGamePacketListenerImpl
     public void handleTeleportToEntityPacket(final ServerboundTeleportToEntityPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
         if (this.player.isSpectator()) {
-            for (ServerLevel level : this.server.getAllLevels()) {
-                Entity entity = packet.getEntity(level);
-                if (entity != null) {
-                    this.player.teleportTo(level, entity.getX(), entity.getY(), entity.getZ(), Set.of(), entity.getYRot(), entity.getXRot(), true, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.SPECTATE); // CraftBukkit
-                    return;
+            // Purpur - Check if requiring a permission node for spectators to teleport to players is disabled or if the player has the permission node.
+            if (!org.purpurmc.purpur.PurpurConfig.requirePermissionSpectatorTeleportToPlayers || this.player.getBukkitEntity().hasPermission("purpur.spectator-can-teleport-to-players")) {
+                for (ServerLevel level : this.server.getAllLevels()) {
+                    Entity entity = packet.getEntity(level);
+                    if (entity != null) {
+                        this.player.teleportTo(level, entity.getX(), entity.getY(), entity.getZ(), Set.of(), entity.getYRot(), entity.getXRot(), true, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.SPECTATE); // CraftBukkit
+                        return;
+                    }
                 }
             }
         }
