From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mickey42302 <bgoodwin423@gmail.com>
Date: Mon, 14 Jul 2025 14:54:29 -0400
Subject: [PATCH] Add bypass permission for Elytra movement check


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index a63ade75461b68a780c56bfb5ff7c61f99f76744..c6860babd7f48aa41ef91bc8c483b7f4cafd24ee 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1801,6 +1801,8 @@ public class ServerGamePacketListenerImpl
             return false;
         } else if (this.player.isChangingDimension()) {
             return false;
+        } else if (this.player.getBukkitEntity().hasPermission("minecraft.movement.bypass.elytra")) {
+            return false;
         } else {
             GameRules gameRules = this.player.level().getGameRules();
             return !gameRules.getBoolean(GameRules.RULE_DISABLE_PLAYER_MOVEMENT_CHECK)
