From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mickey42302 <bgoodwin423@gmail.com>
Date: Sat, 2 May 2026 11:33:02 -0400
Subject: [PATCH] Add check for spectatetp setting


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index e7fb8e64321a8001de8024a7c889755361fbb9e1..b5ec2d667fca64219f179b5cd08e8558e50f2dcb 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2236,12 +2236,24 @@ public class ServerGamePacketListenerImpl
     @Override
     public void handleTeleportToEntityPacket(ServerboundTeleportToEntityPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
-        if (this.player.isSpectator() && this.player.getBukkitEntity().hasPermission("minecraft.spectatetp")) {
-            for (ServerLevel serverLevel : this.server.getAllLevels()) {
-                Entity entity = packet.getEntity(serverLevel);
-                if (entity != null) {
-                    this.player.teleportTo(serverLevel, entity.getX(), entity.getY(), entity.getZ(), Set.of(), entity.getYRot(), entity.getXRot(), true, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.SPECTATE); // CraftBukkit
-                    return;
+        if (org.purpurmc.purpur.PurpurConfig.requireSpectatorTeleportPermission) {
+            if (this.player.isSpectator() && this.player.getBukkitEntity().hasPermission("minecraft.spectatetp")) {
+                for (ServerLevel serverLevel : this.server.getAllLevels()) {
+                    Entity entity = packet.getEntity(serverLevel);
+                    if (entity != null) {
+                        this.player.teleportTo(serverLevel, entity.getX(), entity.getY(), entity.getZ(), Set.of(), entity.getYRot(), entity.getXRot(), true, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.SPECTATE); // CraftBukkit
+                        return;
+                    }
+                }
+            }
+        } else {
+            if (this.player.isSpectator()) {
+                for (ServerLevel serverLevel : this.server.getAllLevels()) {
+                    Entity entity = packet.getEntity(serverLevel);
+                    if (entity != null) {
+                        this.player.teleportTo(serverLevel, entity.getX(), entity.getY(), entity.getZ(), Set.of(), entity.getYRot(), entity.getXRot(), true, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.SPECTATE); // CraftBukkit
+                        return;
+                    }
                 }
             }
         }
