From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/entity/animal/Animal.java b/net/minecraft/world/entity/animal/Animal.java
index b7f089835c00cd5d5f98ca83e6ff1b0d61bee909..e18462f9e3353957eb3079112fe7e81986b36491 100644
--- a/net/minecraft/world/entity/animal/Animal.java
+++ b/net/minecraft/world/entity/animal/Animal.java
@@ -146,7 +146,7 @@ public abstract class Animal extends AgeableMob {
         ItemStack itemStack = player.getItemInHand(hand);
         if (this.isFood(itemStack)) {
             int age = this.getAge();
-            if (player instanceof ServerPlayer serverPlayer && age == 0 && this.canFallInLove()) {
+            if (player instanceof ServerPlayer serverPlayer && age == 0 && this.canFallInLove() && (this.level().purpurConfig.animalBreedingCooldownSeconds <= 0 || !this.level().hasBreedingCooldown(player.getUUID(), this.getClass()))) { // Purpur - Add adjustable breeding cooldown to config
                 final ItemStack breedCopy = itemStack.copy(); // Paper - Fix EntityBreedEvent copying
                 this.usePlayerItem(player, hand, itemStack);
                 this.setInLove(serverPlayer, breedCopy); // Paper - Fix EntityBreedEvent copying
@@ -227,10 +227,20 @@ public abstract class Animal extends AgeableMob {
     public void spawnChildFromBreeding(final ServerLevel level, final Animal partner) {
         AgeableMob offspring = this.getBreedOffspring(level, partner);
         if (offspring != null) {
-            offspring.setBaby(true);
-            offspring.snapTo(this.getX(), this.getY(), this.getZ(), 0.0F, 0.0F);
+            //offspring.setBaby(true); // Purpur - Add adjustable breeding cooldown to config - moved down
+            //offspring.snapTo(this.getX(), this.getY(), this.getZ(), 0.0F, 0.0F); // Purpur - Add adjustable breeding cooldown to config - moved down
             // CraftBukkit start - Call EntityBreedEvent
             ServerPlayer breeder = Optional.ofNullable(this.getLoveCause()).or(() -> Optional.ofNullable(partner.getLoveCause())).orElse(null);
+            // Purpur start - Add adjustable breeding cooldown to config
+            if (breeder != null && level.purpurConfig.animalBreedingCooldownSeconds > 0) {
+                if (level.hasBreedingCooldown(breeder.getUUID(), this.getClass())) {
+                    return;
+                }
+                level.addBreedingCooldown(breeder.getUUID(), this.getClass());
+            }
+            offspring.setBaby(true);
+            offspring.snapTo(this.getX(), this.getY(), this.getZ(), 0.0F, 0.0F);
+            // Purpur end - Add adjustable breeding cooldown to config
             int experience = this.getRandom().nextInt(7) + 1;
             org.bukkit.event.entity.EntityBreedEvent entityBreedEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityBreedEvent(offspring, this, partner, breeder, this.breedItem, experience);
             if (entityBreedEvent.isCancelled()) {
