--- a/net/minecraft/world/entity/animal/Animal.java
+++ b/net/minecraft/world/entity/animal/Animal.java
@@ -146,7 +_,7 @@
         ItemStack itemStack = player.getItemInHand(hand);
         if (this.isFood(itemStack)) {
             int age = this.getAge();
-            if (player instanceof ServerPlayer serverPlayer && age == 0 && this.canFallInLove()) {
+            if (player instanceof ServerPlayer serverPlayer && age == 0 && this.canFallInLove() && (this.level().purpurConfig.animalBreedingCooldownSeconds <= 0 || !this.level().hasBreedingCooldown(player.getUUID(), this.getClass()))) { // Purpur - Add adjustable breeding cooldown to config
                 final ItemStack breedCopy = itemStack.copy(); // Paper - Fix EntityBreedEvent copying
                 this.usePlayerItem(player, hand, itemStack);
                 this.setInLove(serverPlayer, breedCopy); // Paper - Fix EntityBreedEvent copying
@@ -227,10 +_,20 @@
     public void spawnChildFromBreeding(final ServerLevel level, final Animal partner) {
         AgeableMob offspring = this.getBreedOffspring(level, partner);
         if (offspring != null) {
-            offspring.setBaby(true);
-            offspring.snapTo(this.getX(), this.getY(), this.getZ(), 0.0F, 0.0F);
+            //offspring.setBaby(true); // Purpur - Add adjustable breeding cooldown to config - moved down
+            //offspring.snapTo(this.getX(), this.getY(), this.getZ(), 0.0F, 0.0F); // Purpur - Add adjustable breeding cooldown to config - moved down
             // CraftBukkit start - Call EntityBreedEvent
             ServerPlayer breeder = Optional.ofNullable(this.getLoveCause()).or(() -> Optional.ofNullable(partner.getLoveCause())).orElse(null);
+            // Purpur start - Add adjustable breeding cooldown to config
+            if (breeder != null && level.purpurConfig.animalBreedingCooldownSeconds > 0) {
+                if (level.hasBreedingCooldown(breeder.getUUID(), this.getClass())) {
+                    return;
+                }
+                level.addBreedingCooldown(breeder.getUUID(), this.getClass());
+            }
+            offspring.setBaby(true);
+            offspring.snapTo(this.getX(), this.getY(), this.getZ(), 0.0F, 0.0F);
+            // Purpur end - Add adjustable breeding cooldown to config
             int experience = this.getRandom().nextInt(7) + 1;
             org.bukkit.event.entity.EntityBreedEvent entityBreedEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityBreedEvent(offspring, this, partner, breeder, this.breedItem, experience);
             if (entityBreedEvent.isCancelled()) {
