--- a/net/minecraft/world/entity/animal/Bee.java
+++ b/net/minecraft/world/entity/animal/Bee.java
@@ -365,7 +_,7 @@
     }
 
     public static boolean isNightOrRaining(Level level) {
-        return level.dimensionType().hasSkyLight() && (level.isNight() || level.isRaining());
+        return level.dimensionType().hasSkyLight() && (level.isNight() && !level.purpurConfig.beeCanWorkAtNight || level.isRaining() && !level.purpurConfig.beeCanWorkInRain); // Purpur - Bee can work when raining or at night
     }
 
     public void setStayOutOfHiveCountdown(int stayOutOfHiveCountdown) {
