--- a/net/minecraft/world/entity/animal/Bee.java
+++ b/net/minecraft/world/entity/animal/Bee.java
@@ -365,7 +_,7 @@
     }
 
     public static boolean isNightOrRaining(Level level) {
-        return level.dimensionType().hasSkyLight() && (level.isNight() || level.isRaining());
+        return level.dimensionType().hasSkyLight() && (level.isNight() && !level.purpurConfig.beeCanWorkAtNight || level.isRaining() && !level.purpurConfig.beeCanWorkInRain); // Purpur - Bee can work when raining or at night
     }
 
     public void setStayOutOfHiveCountdown(int stayOutOfHiveCountdown) {
@@ -398,6 +_,7 @@
             this.hurtServer(level, this.damageSources().drown(), 1.0F);
         }
 
+        if (hasStung && !this.level().purpurConfig.beeDiesAfterSting) setHasStung(false); else // Purpur - Stop bees from dying after stinging
         if (hasStung) {
             this.timeSinceSting++;
             if (this.timeSinceSting % 5 == 0 && this.random.nextInt(Mth.clamp(1200 - this.timeSinceSting, 1, 1200)) == 0) {
