--- a/net/minecraft/world/entity/animal/SnowGolem.java
+++ b/net/minecraft/world/entity/animal/SnowGolem.java
@@ -153,6 +_,14 @@
             }
 
             return InteractionResult.SUCCESS;
+        // Purpur start - Snowman drop and put back pumpkin
+        } else if (level().purpurConfig.snowGolemPutPumpkinBack && !hasPumpkin() && itemInHand.getItem() == Blocks.CARVED_PUMPKIN.asItem()) {
+            setPumpkin(true);
+            if (!player.getAbilities().instabuild) {
+                itemInHand.shrink(1);
+            }
+            return InteractionResult.SUCCESS;
+        // Purpur end - Snowman drop and put back pumpkin
         } else {
             return InteractionResult.PASS;
         }
