From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/item/DyeColor.java b/net/minecraft/world/item/DyeColor.java
index 154b1e60754a7377c855bb86a3c01c1fac0f303e..5c18b33b358316d0619bc180beb6505a9571c3f3 100644
--- a/net/minecraft/world/item/DyeColor.java
+++ b/net/minecraft/world/item/DyeColor.java
@@ -154,4 +154,10 @@ public enum DyeColor implements StringRepresentable {
             }
         }
     }
+
+    // Purpur start - Shulker spawn from bullet options
+    public static DyeColor random(net.minecraft.util.RandomSource random) {
+        return values()[random.nextInt(values().length)];
+    }
+    // Purpur end - Shulker spawn from bullet options
 }
