--- a/net/minecraft/world/item/Items.java
+++ b/net/minecraft/world/item/Items.java
@@ -778,7 +_,7 @@
     public static final Item PURPUR_BLOCK = registerBlock(BlockItemIds.PURPUR_BLOCK, Blocks.PURPUR_BLOCK);
     public static final Item PURPUR_PILLAR = registerBlock(BlockItemIds.PURPUR_PILLAR, Blocks.PURPUR_PILLAR);
     public static final Item PURPUR_STAIRS = registerBlock(BlockItemIds.PURPUR_STAIRS, Blocks.PURPUR_STAIRS);
-    public static final Item SPAWNER = registerBlock(BlockItemIds.SPAWNER, Blocks.SPAWNER);
+    public static final Item SPAWNER = registerBlock(BlockItemIds.SPAWNER, Blocks.SPAWNER, org.purpurmc.purpur.item.SpawnerItem::new, new Item.Properties().rarity(Rarity.EPIC)); // Purpur - Silk touch spawners
     public static final Item CREAKING_HEART = registerBlock(BlockItemIds.CREAKING_HEART, Blocks.CREAKING_HEART);
     public static final Item CHEST = registerBlock(
         BlockItemIds.CHEST,
@@ -2727,7 +_,7 @@
     );
     public static final Item GLOW_BERRIES = registerItem(
         BlockItemIds.GLOW_BERRY_CROP,
-        createBlockItemWithCustomItemName(Blocks.CAVE_VINES),
+        settings -> new org.purpurmc.purpur.item.GlowBerryItem(Blocks.CAVE_VINES, settings.useItemDescriptionPrefix()),  // Purpur - Eating glow berries adds glow effect
         new Item.Properties().food(Foods.GLOW_BERRIES).compostable(ContextIntProviders.COMPOSTABLE_LOW)
     );
     public static final Item CAMPFIRE = registerBlock(
