--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -202,6 +_,7 @@
     public static void renderBiomePreviewMap(final ServerLevel level, final ItemStack mapItemStack) {
         MapItemSavedData data = getSavedData(mapItemStack, level);
         if (data != null) {
+            data.isExplorerMap = true; // Purpur - Explorer Map API
             if (level.dimension() == data.dimension) {
                 int scale = 1 << data.scale;
                 int centerX = data.centerX;
