From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/item/MapItem.java b/net/minecraft/world/item/MapItem.java
index 337b6aa7ee170a1ec9310a72d4d5f25176246c6b..5d9d776bd5ce9109ec473cc31aa9b9e4f6823fbd 100644
--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -202,6 +202,7 @@ public class MapItem extends Item {
     public static void renderBiomePreviewMap(final ServerLevel level, final ItemStack mapItemStack) {
         MapItemSavedData data = getSavedData(mapItemStack, level);
         if (data != null) {
+            data.isExplorerMap = true; // Purpur - Explorer Map API
             if (level.dimension() == data.dimension) {
                 int scale = 1 << data.scale;
                 int centerX = data.centerX;
