From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/item/MapItem.java b/net/minecraft/world/item/MapItem.java
index f4ad2aa954befed7070559446531621732ac85e0..8bad4b3a7b9dfae3f3993ae9ec03df1f725cb711 100644
--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -202,6 +202,7 @@ public class MapItem extends Item {
     public static void renderBiomePreviewMap(final ServerLevel level, final ItemStack mapItemStack) {
         MapItemSavedData data = getSavedData(mapItemStack, level);
         if (data != null) {
+            data.isExplorerMap = true; // Purpur - Explorer Map API
             if (level.dimension() == data.dimension) {
                 int scale = 1 << data.scale;
                 int centerX = data.centerX;
