--- a/net/minecraft/world/item/ShovelItem.java
+++ b/net/minecraft/world/item/ShovelItem.java
@@ -47,9 +_,12 @@
         BlockState newState = FLATTENABLES.get(blockState.getBlock());
         BlockState updatedState = null;
         Runnable afterAction = null; // Paper
-        if (newState != null && level.getBlockState(pos.above()).isAir()) {
-            afterAction = () -> level.playSound(player, pos, SoundEvents.SHOVEL_FLATTEN, SoundSource.BLOCKS, 1.0F, 1.0F); // Paper
-            updatedState = newState;
+        org.purpurmc.purpur.tool.Flattenable flattenable = level.purpurConfig.shovelFlattenables.get(blockState.getBlock()); // Purpur - Tool actionable options
+            if (newState != null && level.getBlockState(pos.above()).isAir()) {
+                // Purpur start - Tool actionable options
+            afterAction = () -> {if (!FLATTENABLES.containsKey(blockState.getBlock())) level.playSound(player, pos, SoundEvents.SHOVEL_FLATTEN, SoundSource.BLOCKS, 1.0F, 1.0F);}; // Paper
+            updatedState = flattenable.into().defaultBlockState();
+                // Purpur end - Tool actionable options
         } else if (blockState.getBlock() instanceof CampfireBlock && blockState.getValue(CampfireBlock.LIT)) {
             afterAction = () -> { // Paper
             if (!level.isClientSide()) {
