From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/item/SpawnEggItem.java b/net/minecraft/world/item/SpawnEggItem.java
index a4b388f7a6084bef8ab805c136d142c8c763fd71..b0b22dbb92b8f2d361e49355ea15c4838d71c887 100644
--- a/net/minecraft/world/item/SpawnEggItem.java
+++ b/net/minecraft/world/item/SpawnEggItem.java
@@ -61,6 +61,23 @@ public class SpawnEggItem extends Item {
                     return InteractionResult.FAIL;
                 } else {
                     if (level.paperConfig().entities.spawning.disableMobSpawnerSpawnEggTransformation) return InteractionResult.FAIL; // Paper - Allow disabling mob spawner spawn egg transformation
+                    // Purpur start - PlayerSetSpawnerTypeWithEggEvent
+                    if (spawnerHolder instanceof net.minecraft.world.level.block.entity.SpawnerBlockEntity) {
+                        org.bukkit.block.Block bukkitBlock = level.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
+                        org.purpurmc.purpur.event.PlayerSetSpawnerTypeWithEggEvent event = new org.purpurmc.purpur.event.PlayerSetSpawnerTypeWithEggEvent((org.bukkit.entity.Player) context.getPlayer().getBukkitEntity(), bukkitBlock, (org.bukkit.block.CreatureSpawner) bukkitBlock.getState(), org.bukkit.entity.EntityType.fromName(type.getName()));
+                        if (!event.callEvent()) {
+                            return InteractionResult.FAIL;
+                        }
+                        type = EntityType.getFromBukkitType(event.getEntityType());
+                    } else if (spawnerHolder instanceof net.minecraft.world.level.block.entity.TrialSpawnerBlockEntity) {
+                        org.bukkit.block.Block bukkitBlock = level.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
+                        org.purpurmc.purpur.event.PlayerSetTrialSpawnerTypeWithEggEvent event = new org.purpurmc.purpur.event.PlayerSetTrialSpawnerTypeWithEggEvent((org.bukkit.entity.Player) context.getPlayer().getBukkitEntity(), bukkitBlock, (org.bukkit.block.TrialSpawner) bukkitBlock.getState(), org.bukkit.entity.EntityType.fromName(type.getName()));
+                        if (!event.callEvent()) {
+                            return InteractionResult.FAIL;
+                        }
+                        type = EntityType.getFromBukkitType(event.getEntityType());
+                    }
+                    // Purpur end - PlayerSetSpawnerTypeWithEggEvent
                     spawnerHolder.setEntityId(type, level.getRandom());
                     level.sendBlockUpdated(pos, blockState, blockState, Block.UPDATE_ALL);
                     level.gameEvent(context.getPlayer(), GameEvent.BLOCK_CHANGE, pos);
