--- a/net/minecraft/world/item/SpawnEggItem.java
+++ b/net/minecraft/world/item/SpawnEggItem.java
@@ -62,6 +_,23 @@
                     return InteractionResult.FAIL;
                 } else {
                     if (level.paperConfig().entities.spawning.disableMobSpawnerSpawnEggTransformation) return InteractionResult.FAIL; // Paper - Allow disabling mob spawner spawn egg transformation
+                    // Purpur start - PlayerSetSpawnerTypeWithEggEvent
+                    if (spawnerHolder instanceof net.minecraft.world.level.block.entity.SpawnerBlockEntity) {
+                        org.bukkit.block.Block bukkitBlock = level.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
+                        org.purpurmc.purpur.event.PlayerSetSpawnerTypeWithEggEvent event = new org.purpurmc.purpur.event.PlayerSetSpawnerTypeWithEggEvent((org.bukkit.entity.Player) context.getPlayer().getBukkitEntity(), bukkitBlock, (org.bukkit.block.CreatureSpawner) bukkitBlock.getState(), org.bukkit.entity.EntityType.fromName(type.getName()));
+                        if (!event.callEvent()) {
+                            return InteractionResult.FAIL;
+                        }
+                        type = EntityType.getFromBukkitType(event.getEntityType());
+                    } else if (spawnerHolder instanceof net.minecraft.world.level.block.entity.TrialSpawnerBlockEntity) {
+                        org.bukkit.block.Block bukkitBlock = level.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
+                        org.purpurmc.purpur.event.PlayerSetTrialSpawnerTypeWithEggEvent event = new org.purpurmc.purpur.event.PlayerSetTrialSpawnerTypeWithEggEvent((org.bukkit.entity.Player) context.getPlayer().getBukkitEntity(), bukkitBlock, (org.bukkit.block.TrialSpawner) bukkitBlock.getState(), org.bukkit.entity.EntityType.fromName(type.getName()));
+                        if (!event.callEvent()) {
+                            return InteractionResult.FAIL;
+                        }
+                        type = EntityType.getFromBukkitType(event.getEntityType());
+                    }
+                    // Purpur end - PlayerSetSpawnerTypeWithEggEvent
                     spawnerHolder.setEntityId(type, level.getRandom());
                     level.sendBlockUpdated(pos, blockState, blockState, Block.UPDATE_ALL);
                     level.gameEvent(context.getPlayer(), GameEvent.BLOCK_CHANGE, pos);
