--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -88,7 +_,7 @@
         Direction.WEST, Direction.EAST, Direction.NORTH, Direction.SOUTH, Direction.DOWN, Direction.UP
     };
     public final boolean hasCollision;
-    public float explosionResistance;
+    protected final float explosionResistance;
     protected final boolean isRandomlyTicking;
     protected final SoundType soundType;
     protected final float friction;
