--- a/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
@@ -78,6 +_,7 @@
     private @Nullable List<Entity> piercedAndKilledEntities;
     public ItemStack pickupItemStack = this.getDefaultPickupItem();
     public @Nullable ItemStack firedFromWeapon = null;
+    public net.minecraft.world.item.enchantment.ItemEnchantments actualEnchantments = net.minecraft.world.item.enchantment.ItemEnchantments.EMPTY; // Purpur - Add an option to fix MC-3304 projectile looting
 
     protected AbstractArrow(EntityType<? extends AbstractArrow> type, Level level) {
         super(type, level);
@@ -593,6 +_,12 @@
     public @Nullable ItemStack getWeaponItem() {
         return this.firedFromWeapon;
     }
+
+    // Purpur start - Add an option to fix MC-3304 projectile looting
+    public void setActualEnchantments(net.minecraft.world.item.enchantment.ItemEnchantments actualEnchantments) {
+        this.actualEnchantments = actualEnchantments;
+    }
+    // Purpur end - Add an option to fix MC-3304 projectile looting
 
     protected SoundEvent getDefaultHitGroundSoundEvent() {
         return SoundEvents.ARROW_HIT;
