--- a/net/minecraft/world/entity/projectile/arrow/ThrownTrident.java
+++ b/net/minecraft/world/entity/projectile/arrow/ThrownTrident.java
@@ -72,7 +_,7 @@
 
         Entity currentOwner = this.getOwner();
         int loyalty = this.entityData.get(ID_LOYALTY);
-        if (loyalty > 0 && (this.dealtDamage || this.isNoPhysics()) && currentOwner != null) {
+        if (loyalty > 0 && (this.dealtDamage || this.isNoPhysics() || (level().purpurConfig.tridentLoyaltyVoidReturnHeight < 0.0D && getY() < level().purpurConfig.tridentLoyaltyVoidReturnHeight)) && currentOwner != null) { // Purpur - Add option to allow loyalty on tridents to work in the void
             if (!this.isAcceptibleReturnOwner()) {
                 if (this.level() instanceof ServerLevel level && this.pickup == AbstractArrow.Pickup.ALLOWED) {
                     this.spawnAtLocation(level, this.getPickupItem(), 0.1F);
