--- a/net/minecraft/world/level/chunk/storage/EntityStorage.java
+++ b/net/minecraft/world/level/chunk/storage/EntityStorage.java
@@ -93,6 +_,7 @@
             try (ProblemReporter.ScopedCollector reporter = new ProblemReporter.ScopedCollector(ChunkAccess.problemPath(pos), LOGGER)) {
                 ListTag entities = new ListTag();
                 chunk.getEntities().forEach(e -> {
+                    if (!e.canSaveToDisk()) return; // Purpur - Add canSaveToDisk to Entity
                     TagValueOutput output = TagValueOutput.createWithContext(reporter.forChild(e.problemPath()), e.registryAccess());
                     if (e.save(output)) {
                         CompoundTag result = output.buildResult();
