From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/chunk/storage/EntityStorage.java b/net/minecraft/world/level/chunk/storage/EntityStorage.java
index 4bd083ecdea207ee5689760fccc9ce8a5a958d57..8878d95e9eee03aa85f6192ea82d5581beae98e4 100644
--- a/net/minecraft/world/level/chunk/storage/EntityStorage.java
+++ b/net/minecraft/world/level/chunk/storage/EntityStorage.java
@@ -94,6 +94,7 @@ public class EntityStorage implements EntityPersistentStorage<Entity> {
                 ListTag entities = new ListTag();
                 final java.util.Map<net.minecraft.world.entity.EntityType<?>, Integer> savedEntityCounts = new java.util.HashMap<>(); // Paper - Entity load/save limit per chunk
                 chunk.getEntities().forEach(e -> {
+                    if (!e.canSaveToDisk()) return; // Purpur - Add canSaveToDisk to Entity
                     // Paper start - Entity load/save limit per chunk
                     final EntityType<?> entityType = e.getType();
                     final int saveLimit = this.level.paperConfig().chunks.entityPerChunkSaveLimit.getOrDefault(entityType, -1);
