--- a/net/minecraft/world/level/chunk/storage/EntityStorage.java
+++ b/net/minecraft/world/level/chunk/storage/EntityStorage.java
@@ -94,6 +_,7 @@
                 ListTag entities = new ListTag();
                 final java.util.Map<net.minecraft.world.entity.EntityType<?>, Integer> savedEntityCounts = new java.util.HashMap<>(); // Paper - Entity load/save limit per chunk
                 chunk.getEntities().forEach(e -> {
+                    if (!e.canSaveToDisk()) return; // Purpur - Add canSaveToDisk to Entity
                     // Paper start - Entity load/save limit per chunk
                     final EntityType<?> entityType = e.getType();
                     final int saveLimit = this.level.paperConfig().chunks.entityPerChunkSaveLimit.getOrDefault(entityType, -1);
