From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: granny <contact@granny.dev>
Date: Mon, 8 Dec 2025 15:44:50 -0800
Subject: [PATCH] Setting to reintroduce end void rings


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 569b46cce9d5207380a9391e969e3a6aa790717b..a0cf70f1ccbb78f058e8ac5fc95bcb5b03300cdc 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -288,6 +288,7 @@ public class GlobalConfiguration extends ConfigurationPart {
         }
     }
 
+    @Setting(org.purpurmc.purpur.configuration.transformation.FarEndTerrainGenerationMigration.MISC_KEY) // Purpur - Migrate Setting to reintroduce end void rings
     public Misc misc;
 
     public class Misc extends ConfigurationPart {
@@ -334,6 +335,7 @@ public class GlobalConfiguration extends ConfigurationPart {
         @Comment("Whether the nether dimension is enabled and will be loaded.")
         public boolean enableNether = true;
         @Comment("Keeps Paper's fix for MC-159283 enabled. Disable to use vanilla End ring terrain.")
+        @Setting(org.purpurmc.purpur.configuration.transformation.FarEndTerrainGenerationMigration.FIX_FAR_END_TERRAIN_GENERATION_KEY) // Purpur - Migrate Setting to reintroduce end void rings
         public boolean fixFarEndTerrainGeneration = true;
     }
 
diff --git a/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java b/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
index 1ab7f28e2ba386c68b9c70617c816b0b55fd796d..1446423550b6637aa955a1fa079f6c5d222093df 100644
--- a/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
+++ b/src/main/java/io/papermc/paper/configuration/PaperConfigurations.java
@@ -293,6 +293,7 @@ public class PaperConfigurations extends Configurations<GlobalConfiguration, Wor
         for (NodePath path : RemovedConfigurations.REMOVED_GLOBAL_PATHS) {
             builder.addAction(path, TransformAction.remove());
         }
+        org.purpurmc.purpur.configuration.transformation.FarEndTerrainGenerationMigration.apply(builder); // Purpur - Migrate Setting to reintroduce end void rings
         builder.build().apply(node);
 
         final ConfigurationTransformation.VersionedBuilder versionedBuilder = Transformations.versionedBuilder();
