From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mickey42302 <bgoodwin423@gmail.com>
Date: Sun, 13 Jul 2025 22:29:59 -0400
Subject: [PATCH] Add bypass permission in Bukkit for spawn protection


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 658385b2887d6debec7fc941c28621da5d263411..d511cd839be7de82302bbd08ab6b195377d120bf 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -272,6 +272,7 @@ public class CraftEventFactory {
         if (spawnSize <= 0) return true;
         if (((CraftServer) Bukkit.getServer()).getHandle().getOps().isEmpty()) return true;
         if (player.isOp()) return true;
+        if (player.hasPermission("minecraft.spawn-protection.bypass")) return true;
 
         BlockPos spawnPos = world.getSharedSpawnPos();
 
