From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/server/players/SleepStatus.java b/net/minecraft/server/players/SleepStatus.java
index 61870e387c93a5f0e62180372438b7011a294cd6..51f69631603b90b194080aeecffafb4d613b2af8 100644
--- a/net/minecraft/server/players/SleepStatus.java
+++ b/net/minecraft/server/players/SleepStatus.java
@@ -15,7 +15,7 @@ public class SleepStatus {
 
     public boolean areEnoughDeepSleeping(final int sleepPercentageNeeded, final List<ServerPlayer> players) {
         // CraftBukkit start
-        int deepSleepers = (int)players.stream().filter(player -> player.isSleepingLongEnough() || player.fauxSleeping).count();
+        int deepSleepers = (int)players.stream().filter(player -> player.isSleepingLongEnough() || player.fauxSleeping || (player.level().purpurConfig.idleTimeoutCountAsSleeping && player.isAfk())).count(); // Purpur - AFK API
         boolean anyDeepSleep = players.stream().anyMatch(Player::isSleepingLongEnough);
         return anyDeepSleep && deepSleepers >= this.sleepersNeeded(sleepPercentageNeeded);
         // CraftBukkit end
@@ -43,7 +43,7 @@ public class SleepStatus {
         for (ServerPlayer player : players) {
             if (!player.isSpectator()) {
                 this.activePlayers++;
-                if (player.isSleeping() || player.fauxSleeping) { // CraftBukkit
+                if (player.isSleeping() || player.fauxSleeping || (player.level().purpurConfig.idleTimeoutCountAsSleeping && player.isAfk())) { // CraftBukkit // Purpur - AFK API
                     this.sleepingPlayers++;
                 }
                 // CraftBukkit start
