--- a/net/minecraft/world/entity/animal/sheep/Sheep.java
+++ b/net/minecraft/world/entity/animal/sheep/Sheep.java
@@ -320,6 +_,11 @@
         ServerLevelAccessor level, DifficultyInstance difficulty, EntitySpawnReason spawnReason, @Nullable SpawnGroupData spawnGroupData
     ) {
         this.setColor(getRandomSheepColor(level, this.blockPosition()));
+        // Purpur start - Special mobs naturally spawn
+        if (level.getLevel().purpurConfig.sheepNaturalJeb > 0D && random.nextDouble() <= level.getLevel().purpurConfig.sheepNaturalJeb) {
+            setCustomName(net.minecraft.network.chat.Component.translatable("jeb_"));
+        }
+        // Purpur end - Special mobs naturally spawn
         return super.finalizeSpawn(level, difficulty, spawnReason, spawnGroupData);
     }
 }
