--- a/net/minecraft/world/entity/monster/warden/Warden.java
+++ b/net/minecraft/world/entity/monster/warden/Warden.java
@@ -169,6 +_,7 @@
 
     @Override
     protected boolean canRide(final Entity vehicle) {
+        if (this.level().purpurConfig.wardenCanRideVehicles) return this.boardingCooldown <= 0; // Purpur - Configs for if Wither/Ender Dragon can ride vehicles
         return false;
     }
 
