--- a/net/minecraft/world/item/trading/MerchantOffer.java
+++ b/net/minecraft/world/item/trading/MerchantOffer.java
@@ -143,8 +_,13 @@
     }
 
     public void updateDemand() {
+        // Purpur start - Configurable minimum demand for trades
+        this.updateDemand(0);
+    }
+    public void updateDemand(int minimumDemand) {
+        // Purpur end - Configurable minimum demand for trades
         this.demand = this.demand + this.uses - (this.maxUses - this.uses);
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().misc.preventNegativeVillagerDemand) this.demand = Math.max(0, this.demand); // Paper - Fix MC-163962
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().misc.preventNegativeVillagerDemand) this.demand = Math.max(minimumDemand, this.demand); // Paper - Fix MC-163962 // Purpur - Configurable minimum demand for trades
     }
 
     public ItemStack assemble() {
