--- a/src/main/java/co/aikar/timings/TimedEventExecutor.java
+++ b/src/main/java/co/aikar/timings/TimedEventExecutor.java
@@ -80,9 +_,9 @@
             executor.execute(listener, event);
             return;
         }
-        try (Timing ignored = timings.startTiming()){
+        //try (Timing ignored = timings.startTiming()){ // Purpur - Remove Timings
             executor.execute(listener, event);
-        }
+        //} // Purpur - Remove Timings
     }
 
     @Override
