--- a/src/main/java/co/aikar/timings/Timing.java
+++ b/src/main/java/co/aikar/timings/Timing.java
@@ -39,6 +_,7 @@
      * @return Timing
      */
     @NotNull
+    @io.papermc.paper.annotation.DoNotUse // Purpur - Remove Timings
     Timing startTiming();
 
     /**
@@ -46,6 +_,7 @@
      *
      * Will automatically be called when this Timing is used with try-with-resources
      */
+    @io.papermc.paper.annotation.DoNotUse // Purpur - Remove Timings
     void stopTiming();
 
     /**
@@ -56,6 +_,7 @@
      * @return Timing
      */
     @NotNull
+    @io.papermc.paper.annotation.DoNotUse // Purpur - Remove Timings
     Timing startTimingIfSync();
 
     /**
@@ -65,12 +_,14 @@
      *
      * But only if we are on the primary thread.
      */
+    @io.papermc.paper.annotation.DoNotUse // Purpur - Remove Timings
     void stopTimingIfSync();
 
     /**
      * @deprecated Doesn't do anything - Removed
      */
     @Deprecated
+    @io.papermc.paper.annotation.DoNotUse // Purpur - Remove Timings
     void abort();
 
     /**
@@ -82,5 +_,6 @@
     TimingHandler getTimingHandler();
 
     @Override
+    @io.papermc.paper.annotation.DoNotUse // Purpur - Remove Timings
     void close();
 }
