--- a/src/main/java/co/aikar/timings/Timings.java
+++ b/src/main/java/co/aikar/timings/Timings.java
@@ -124,7 +_,7 @@
     @NotNull
     public static Timing ofStart(@NotNull Plugin plugin, @NotNull String name, @Nullable Timing groupHandler) {
         Timing timing = of(plugin, name, groupHandler);
-        timing.startTiming();
+        //timing.startTiming(); // Purpur - Remove Timings
         return timing;
     }
 
@@ -146,7 +_,7 @@
      */
     public static void setTimingsEnabled(boolean enabled) {
         if (enabled && !warnedAboutDeprecationOnEnable) {
-            Bukkit.getLogger().severe(PlainTextComponentSerializer.plainText().serialize(deprecationMessage()));
+            //Bukkit.getLogger().severe(PlainTextComponentSerializer.plainText().serialize(deprecationMessage())); // Purpur - Remove Timings
             warnedAboutDeprecationOnEnable = true;
         }
     }
@@ -322,4 +_,3 @@
         return TimingsManager.getHandler(groupName, name, groupHandler);
     }
 }
-
