--- a/net/minecraft/world/item/crafting/Ingredient.java
+++ b/net/minecraft/world/item/crafting/Ingredient.java
@@ -35,6 +_,7 @@
     public final HolderSet<Item> values;
     // CraftBukkit start
     private java.util.@org.jspecify.annotations.Nullable Set<ItemStack> itemStacks; // Paper - Improve exact choice recipe ingredients
+    public Predicate<org.bukkit.inventory.ItemStack> predicate; // Purpur - Add predicate to recipe's ExactChoice ingredient
 
     public boolean isExact() {
         return this.itemStacks != null;
@@ -86,6 +_,11 @@
             return this.itemStacks.contains(input); // Paper - Improve exact choice recipe ingredients (hashing FTW!)
         }
         // CraftBukkit end
+        // Purpur start - Add predicate to recipe's ExactChoice ingredient
+        if (predicate != null) {
+            return predicate.test(input.asBukkitCopy());
+        }
+        // Purpur end - Add predicate to recipe's ExactChoice ingredient
         return input.is(this.values);
     }
 
