From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/item/crafting/Ingredient.java b/net/minecraft/world/item/crafting/Ingredient.java
index 1deeb3fd148c5d59250346147681b7e8f2147574..0a0e1d67578183044783ed646c9dcf93244d9450 100644
--- a/net/minecraft/world/item/crafting/Ingredient.java
+++ b/net/minecraft/world/item/crafting/Ingredient.java
@@ -35,6 +35,7 @@ public final class Ingredient implements Predicate<ItemStack>, StackedContents.I
     public final HolderSet<Item> values;
     // CraftBukkit start
     private java.util.@org.jspecify.annotations.Nullable Set<ItemStack> itemStacks; // Paper - Improve exact choice recipe ingredients
+    public Predicate<org.bukkit.inventory.ItemStack> predicate; // Purpur - Add predicate to recipe's ExactChoice ingredient
 
     public boolean isExact() {
         return this.itemStacks != null;
@@ -86,6 +87,11 @@ public final class Ingredient implements Predicate<ItemStack>, StackedContents.I
             return this.itemStacks.contains(input); // Paper - Improve exact choice recipe ingredients (hashing FTW!)
         }
         // CraftBukkit end
+        // Purpur start - Add predicate to recipe's ExactChoice ingredient
+        if (predicate != null) {
+            return predicate.test(input.asBukkitCopy());
+        }
+        // Purpur end - Add predicate to recipe's ExactChoice ingredient
         return input.is(this.values);
     }
 
