From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index c4396509584011a5bbc171510478e8e62ce44d94..38bd5b119224f8e2c6423d298de27c098beadfdf 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -268,11 +268,11 @@ public class Commands {
             JfrCommand.register(this.dispatcher);
         }
 
-        if (SharedConstants.DEBUG_CHASE_COMMAND) {
+        if (org.purpurmc.purpur.PurpurConfig.registerMinecraftDisabledCommands || SharedConstants.DEBUG_CHASE_COMMAND) { // Purpur - register disabled minecraft commands
             ChaseCommand.register(this.dispatcher);
         }
 
-        if (SharedConstants.DEBUG_DEV_COMMANDS || SharedConstants.IS_RUNNING_IN_IDE) {
+        if (org.purpurmc.purpur.PurpurConfig.registerMinecraftDebugCommands || SharedConstants.DEBUG_DEV_COMMANDS || SharedConstants.IS_RUNNING_IN_IDE) {  // Purpur - register minecraft debug commands
             RaidCommand.register(this.dispatcher, context);
             DebugPathCommand.register(this.dispatcher);
             DebugMobSpawningCommand.register(this.dispatcher);
@@ -300,6 +300,14 @@ public class Commands {
             StopCommand.register(this.dispatcher);
             TransferCommand.register(this.dispatcher);
             WhitelistCommand.register(this.dispatcher);
+            org.purpurmc.purpur.command.CreditsCommand.register(this.dispatcher); // Purpur - Add credits command
+            org.purpurmc.purpur.command.DemoCommand.register(this.dispatcher); // Purpur - Add demo command
+            org.purpurmc.purpur.command.PingCommand.register(this.dispatcher); // Purpur - Add ping command
+            org.purpurmc.purpur.command.UptimeCommand.register(this.dispatcher); // Purpur - Add uptime command
+            org.purpurmc.purpur.command.TPSBarCommand.register(this.dispatcher); // Purpur - Implement TPSBar
+            org.purpurmc.purpur.command.CompassCommand.register(this.dispatcher); // Purpur - Add compass command
+            org.purpurmc.purpur.command.RamBarCommand.register(this.dispatcher); // Purpur - Add rambar command
+            org.purpurmc.purpur.command.RamCommand.register(this.dispatcher); // Purpur - Add ram command
         }
 
         if (commandSelection.includeIntegrated) {
@@ -523,6 +531,7 @@ public class Commands {
     private void runSync(ServerPlayer player, java.util.Collection<String> bukkit, RootCommandNode<CommandSourceStack> root) {
         // Paper end - Perf: Async command map building
         new com.destroystokyo.paper.event.brigadier.AsyncPlayerSendCommandsEvent<CommandSourceStack>(player.getBukkitEntity(), (RootCommandNode) root, true).callEvent(); // Paper - Brigadier API
+        if (org.bukkit.event.player.PlayerCommandSendEvent.getHandlerList().getRegisteredListeners().length > 0) { // Purpur - Skip events if there's no listeners
         org.bukkit.event.player.PlayerCommandSendEvent event = new org.bukkit.event.player.PlayerCommandSendEvent(player.getBukkitEntity(), new java.util.LinkedHashSet<>(bukkit));
         event.getPlayer().getServer().getPluginManager().callEvent(event);
 
@@ -533,6 +542,8 @@ public class Commands {
             }
         }
         // CraftBukkit end
+        } // Purpur - Skip events if there's no listeners
+
         player.connection.send(new ClientboundCommandsPacket(root, COMMAND_NODE_INSPECTOR));
     }
 
