--- a/net/minecraft/core/dispenser/EquipmentDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/EquipmentDispenseItemBehavior.java
@@ -32,7 +_,7 @@
             return false;
         } else {
             LivingEntity livingEntity = entitiesOfClass.getFirst();
-            EquipmentSlot equipmentSlotForItem = livingEntity.getEquipmentSlotForItem(item);
+            EquipmentSlot equipmentSlotForItem = blockSource.level().purpurConfig.dispenserApplyCursedArmor ? livingEntity.getEquipmentSlotForItem(item) : livingEntity.getEquipmentSlotForDispenserItem(item); if (equipmentSlotForItem == null) return false; // Purpur - Dispenser curse of binding protection
             ItemStack itemStack = item.copyWithCount(1); // Paper - shrink below and single item in event
             // CraftBukkit start
             net.minecraft.world.level.Level world = blockSource.level();
