--- a/net/minecraft/core/dispenser/EquipmentDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/EquipmentDispenseItemBehavior.java
@@ -32,7 +_,7 @@
         }
 
         LivingEntity target = entities.getFirst();
-        EquipmentSlot slot = target.getEquipmentSlotForItem(dispensed);
+        EquipmentSlot slot = source.level().purpurConfig.dispenserApplyCursedArmor ? target.getEquipmentSlotForItem(dispensed) : target.getEquipmentSlotForDispenserItem(dispensed); if (slot == null) return false; // Purpur - Dispenser curse of binding protection
         ItemStack equip = dispensed.copyWithCount(1); // Paper - shrink below and single item in event
         // CraftBukkit start
         net.minecraft.world.level.Level world = source.level();
