From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/stats/ServerRecipeBook.java b/net/minecraft/stats/ServerRecipeBook.java
index ac3fbaf20c3ed8ca981c9071f1bf95f2ad169f85..9a278c857c2a6e40c2a3d2a2991ab8c7676da4c6 100644
--- a/net/minecraft/stats/ServerRecipeBook.java
+++ b/net/minecraft/stats/ServerRecipeBook.java
@@ -110,6 +110,7 @@ public class ServerRecipeBook extends RecipeBook {
     ) {
         for (ResourceKey<Recipe<?>> recipe : recipes) {
             if (!validator.test(recipe)) {
+                if (!org.purpurmc.purpur.PurpurConfig.loggerSuppressUnrecognizedRecipeErrors) // Purpur - Logger settings (suppressing pointless logs)
                 LOGGER.error("Tried to load unrecognized recipe: {} removed now.", recipe);
             } else {
                 recipeAddingMethod.accept(recipe);
