--- a/net/minecraft/stats/ServerRecipeBook.java
+++ b/net/minecraft/stats/ServerRecipeBook.java
@@ -138,6 +_,7 @@
             try {
                 ResourceKey<Recipe<?>> resourceKey = ResourceKey.create(Registries.RECIPE, ResourceLocation.parse(string));
                 if (!isRecognized.test(resourceKey)) {
+                    if (!org.purpurmc.purpur.PurpurConfig.loggerSuppressUnrecognizedRecipeErrors) // Purpur - Logger settings (suppressing pointless logs)
                     LOGGER.error("Tried to load unrecognized recipe: {} removed now.", resourceKey);
                 } else {
                     output.accept(resourceKey);
