From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/stats/ServerRecipeBook.java b/net/minecraft/stats/ServerRecipeBook.java
index fdc180f4d68d4462516b24490f4db5e568ce7701..e820e147324e492c9a23fe76c0749c18ea057dc3 100644
--- a/net/minecraft/stats/ServerRecipeBook.java
+++ b/net/minecraft/stats/ServerRecipeBook.java
@@ -110,6 +110,7 @@ public class ServerRecipeBook extends RecipeBook {
     ) {
         for (ResourceKey<Recipe<?>> recipe : recipes) {
             if (!validator.test(recipe)) {
+                if (!org.purpurmc.purpur.PurpurConfig.loggerSuppressUnrecognizedRecipeErrors) // Purpur - Logger settings (suppressing pointless logs)
                 LOGGER.error("Tried to load unrecognized recipe: {} removed now.", recipe);
             } else {
                 recipeAddingMethod.accept(recipe);
