From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/block/ComposterBlock.java b/net/minecraft/world/level/block/ComposterBlock.java
index 69e3969242ad52168e947af209e6fa72f36178ae..ba5ffa84105a6101c4fe8652a112a00ce40cb4d4 100644
--- a/net/minecraft/world/level/block/ComposterBlock.java
+++ b/net/minecraft/world/level/block/ComposterBlock.java
@@ -256,17 +256,27 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
     ) {
         int fillLevel = state.getValue(LEVEL);
         if (fillLevel < 8 && COMPOSTABLES.containsKey(itemStack.getItem())) {
-            if (fillLevel < 7 && !level.isClientSide()) {
-                BlockState newState = addItem(player, state, level, pos, itemStack);
-                // Paper start - handle cancelled events
-                if (newState == null) {
-                    return InteractionResult.PASS;
-                }
-                // Paper end
-                level.levelEvent(LevelEvent.COMPOSTER_FILL, pos, state != newState ? 1 : 0);
-                player.awardStat(Stats.ITEM_USED.get(itemStack.getItem()));
-                itemStack.consume(1, player);
+            // Purpur start - sneak to bulk process composter
+            BlockState newState = process(fillLevel, player, state, level, pos, itemStack);
+            if (newState == null) {
+                return InteractionResult.PASS;
+            }
+            if (level.purpurConfig.composterBulkProcess && player.isShiftKeyDown()) {
+                BlockState oldState;
+                int oldCount, newCount, oldLevel, newLevel;
+                do {
+                    oldState = newState;
+                    oldCount = itemStack.getCount();
+                    oldLevel = oldState.getValue(ComposterBlock.LEVEL);
+                    newState = process(oldLevel, player, oldState, level, pos, itemStack);
+                    if (newState == null) {
+                        return InteractionResult.PASS;
+                    }
+                    newCount = itemStack.getCount();
+                    newLevel = newState.getValue(ComposterBlock.LEVEL);
+                } while (newCount > 0 && (newCount != oldCount || newLevel != oldLevel));
             }
+            // Purpur end - Sneak to bulk process composter
 
             return InteractionResult.SUCCESS;
         } else {
@@ -274,6 +284,25 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
         }
     }
 
+    // Purpur start - sneak to bulk process composter
+    private static @Nullable BlockState process(int fillLevel, Player player, BlockState state, Level level, BlockPos pos, ItemStack itemStack) {
+        if (fillLevel < 7 && !level.isClientSide()) {
+            BlockState newState = ComposterBlock.addItem(player, state, level, pos, itemStack);
+            // Paper start - handle cancelled events
+            if (newState == null) {
+                return null;
+            }
+            // Paper end
+
+            level.levelEvent(LevelEvent.COMPOSTER_FILL, pos, state != newState ? 1 : 0);
+            player.awardStat(Stats.ITEM_USED.get(itemStack.getItem()));
+            itemStack.consume(1, player);
+            return newState;
+        }
+        return state;
+    }
+    // Purpur end - Sneak to bulk process composter
+
     @Override
     protected InteractionResult useWithoutItem(
         final BlockState state, final Level level, final BlockPos pos, final Player player, final BlockHitResult hitResult
