From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/block/LeavesBlock.java b/net/minecraft/world/level/block/LeavesBlock.java
index 23ff958d35aa9199f361aa88cca7fd676f035709..0584467d66198dfb5efc1e68c97133882771f2de 100644
--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -65,6 +65,12 @@ public abstract class LeavesBlock extends Block implements SimpleWaterloggedBloc
 
     @Override
     protected void randomTick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        // Purpur start - Instant leaves decay
+        if (!level.purpurConfig.leavesInstantDecay) checkDecay(state, level, pos, random);
+    }
+
+    private void checkDecay(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        // Purpur end - Instant leaves decay
         if (this.decaying(state)) {
             // CraftBukkit start
             org.bukkit.event.block.LeavesDecayEvent event = new org.bukkit.event.block.LeavesDecayEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, pos));
@@ -86,6 +92,7 @@ public abstract class LeavesBlock extends Block implements SimpleWaterloggedBloc
     @Override
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         level.setBlock(pos, updateDistance(state, level, pos), Block.UPDATE_ALL);
+        if (level.purpurConfig.leavesInstantDecay) checkDecay(state, level, pos, random); // Purpur - Instant leaves decay
     }
 
     @Override
