From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/entity/monster/hoglin/Hoglin.java b/net/minecraft/world/entity/monster/hoglin/Hoglin.java
index 288d03cd7fcf4d3e1536a0d692c31a0343063a8c..e3f59999c9592af213b21afb43bcaddb94fd8279 100644
--- a/net/minecraft/world/entity/monster/hoglin/Hoglin.java
+++ b/net/minecraft/world/entity/monster/hoglin/Hoglin.java
@@ -173,6 +173,11 @@ public class Hoglin extends Animal implements Enemy, HoglinBase {
     public static boolean checkHoglinSpawnRules(
         final EntityType<Hoglin> type, final LevelAccessor level, final EntitySpawnReason spawnReason, final BlockPos pos, final RandomSource random
     ) {
+        // Purpur start - Config to disable hostile mob spawn on ice
+        if (net.minecraft.world.entity.monster.Monster.canSpawnInBlueAndPackedIce(level, pos)) {
+            return false;
+        }
+        // Purpur end - Config to disable hostile mob spawn on ice
         return !level.getBlockState(pos.below()).is(Blocks.NETHER_WART_BLOCK);
     }
 
