--- a/net/minecraft/world/clock/ServerClockManager.java
+++ b/net/minecraft/world/clock/ServerClockManager.java
@@ -122,7 +_,7 @@
         ServerClockManager.ClockInstance instance = this.getInstance(clock);
         action.accept(instance);
         Map<Holder<WorldClock>, ClockNetworkState> updates = Map.of(clock, instance.packNetworkState(this.server));
-        this.server.getPlayerList().broadcastAll(new ClientboundSetTimePacket(this.getGameTime(), updates)); // TODO 26.1 per-player time
+        this.server.getPlayerList().broadcastAll(new ClientboundSetTimePacket(this.getGameTime(), updates)); // TODO 26.1 per-player time // Purpur - TODO: Configurable daylight cycle
         this.setDirty();
 
         for (ServerLevel level : this.server.getAllLevels()) {
@@ -151,7 +_,7 @@
     // Paper end
 
     public ClientboundSetTimePacket createFullSyncPacket() {
-        // TODO - snapshot: 26.1 per-player time
+        // TODO - snapshot: 26.1 per-player time // Purpur - TODO: Configurable daylight cycle
         return new ClientboundSetTimePacket(this.getGameTime(), Util.mapValues(this.clocks, clock -> clock.packNetworkState(this.server)));
     }
 
