From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/clock/ServerClockManager.java b/net/minecraft/world/clock/ServerClockManager.java
index 436ed3d95c21b8590664a125785545e0d07375fd..2cadb4fca189d6060a4eabba6099a52f84f50b1d 100644
--- a/net/minecraft/world/clock/ServerClockManager.java
+++ b/net/minecraft/world/clock/ServerClockManager.java
@@ -70,7 +70,15 @@ public class ServerClockManager extends SavedData implements ClockManager {
     public void tick() {
         boolean advanceTime = this.advanceTime(); // Paper - per-world time
         if (advanceTime) {
-            this.clocks.values().forEach(ServerClockManager.ClockInstance::tick);
+            this.clocks.forEach(((worldClockHolder, clockInstance) -> { // Purpur start - Configurable daylight cycle
+                ServerLevel l = level != null ? level : server.overworld();
+                float incrementTicks = 12000F / (l.isBrightOutside() ? l.purpurConfig.daytimeTicks : l.purpurConfig.nighttimeTicks);
+                float rate = clockInstance.rate;
+                clockInstance.rate *= incrementTicks;
+                clockInstance.tick();
+                clockInstance.rate = rate;
+                this.broadcastUpdates(worldClockHolder, clockInstance);
+            })); // Purpur end - Configurable daylight cycle
             this.setDirty();
         }
     }
