--- a/net/minecraft/world/clock/ServerClockManager.java
+++ b/net/minecraft/world/clock/ServerClockManager.java
@@ -70,7 +_,15 @@
     public void tick() {
         boolean advanceTime = this.advanceTime(); // Paper - per-world time
         if (advanceTime) {
-            this.clocks.values().forEach(ServerClockManager.ClockInstance::tick);
+            this.clocks.forEach(((worldClockHolder, clockInstance) -> { // Purpur start - Configurable daylight cycle
+                ServerLevel l = level != null ? level : server.overworld();
+                int incrementTicks = 12000 / (l.isBrightOutside() ? l.purpurConfig.daytimeTicks : l.purpurConfig.nighttimeTicks);
+                float rate = clockInstance.rate;
+                clockInstance.rate *= incrementTicks;
+                clockInstance.tick();
+                clockInstance.rate = rate;
+                this.broadcastUpdates(worldClockHolder, clockInstance);
+            })); // Purpur end - Configurable daylight cycle
             this.setDirty();
         }
     }
