From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
index 50f12f2581267827c394134cd51a09016a5fa2b0..7ca301068087a6460a7220168ccc0212234ab2ca 100644
--- a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
@@ -116,9 +116,9 @@ public class RangedBowAttackGoal<T extends Monster & RangedAttackMob> extends Go
                 }
 
                 this.mob.lookAt(target, 30.0F, 30.0F);
-            } else {
+            } //else { // Purpur - MC-121706 - Fix mobs not looking up and down when strafing
                 this.mob.getLookControl().setLookAt(target, 30.0F, 30.0F);
-            }
+            //} // Purpur - MC-121706 - Fix mobs not looking up and down when strafing
 
             if (this.mob.isUsingItem()) {
                 if (!hasLineOfSight && this.seeTime < -60) {
