--- a/net/minecraft/world/level/block/BigDripleafBlock.java
+++ b/net/minecraft/world/level/block/BigDripleafBlock.java
@@ -259,7 +_,7 @@
             playTiltSound(level, pos, sound);
         }
 
-        int tickDelay = DELAY_UNTIL_NEXT_TILT_STATE.getInt(tilt);
+        int tickDelay = level.purpurConfig.bigDripleafTiltDelay.getOrDefault(tilt, -1); // Purpur - Big dripleaf tilt delay
         if (tickDelay != -1) {
             level.scheduleTick(pos, this, tickDelay);
         }
