From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/block/BigDripleafBlock.java b/net/minecraft/world/level/block/BigDripleafBlock.java
index adf7399e6589f9d1092be3858ed1ce4b08009925..515a9e98e960ffde78e1b25a26d6f8c2f634def9 100644
--- a/net/minecraft/world/level/block/BigDripleafBlock.java
+++ b/net/minecraft/world/level/block/BigDripleafBlock.java
@@ -259,7 +259,7 @@ public class BigDripleafBlock extends HorizontalDirectionalBlock implements Simp
             playTiltSound(level, pos, sound);
         }
 
-        int tickDelay = DELAY_UNTIL_NEXT_TILT_STATE.getInt(tilt);
+        int tickDelay = level.purpurConfig.bigDripleafTiltDelay.getOrDefault(tilt, -1); // Purpur - Big dripleaf tilt delay
         if (tickDelay != -1) {
             level.scheduleTick(pos, this, tickDelay);
         }
