--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -412,7 +_,15 @@
     }
 
     public void setPlacedBy(Level level, BlockPos pos, BlockState state, @Nullable LivingEntity placer, ItemStack stack) {
-    }
+        this.placer = placer; // Purpur - Store placer on Block when placed
+    }
+
+    // Purpur start - Store placer on Block when placed
+    @Nullable protected LivingEntity placer = null;
+    public void forgetPlacer() {
+        this.placer = null;
+    }
+    // Purpur end - Store placer on Block when placed
 
     public boolean isPossibleToRespawnInThis(BlockState state) {
         return !state.isSolid() && !state.liquid();
