--- a/net/minecraft/world/level/block/CampfireBlock.java
+++ b/net/minecraft/world/level/block/CampfireBlock.java
@@ -136,7 +_,7 @@
         return this.defaultBlockState()
             .setValue(WATERLOGGED, replacedWater)
             .setValue(SIGNAL_FIRE, this.isSmokeSource(level.getBlockState(pos.below())))
-            .setValue(LIT, !replacedWater)
+            .setValue(LIT, level.getMinecraftWorld().purpurConfig.campFireLitWhenPlaced && !replacedWater) // Purpur - Campfire option for lit when placed
             .setValue(FACING, context.getHorizontalDirection());
     }
 
