--- a/net/minecraft/world/level/block/CauldronBlock.java
+++ b/net/minecraft/world/level/block/CauldronBlock.java
@@ -32,8 +_,8 @@
 
     protected static boolean shouldHandlePrecipitation(Level level, Biome.Precipitation precipitation) {
         return precipitation == Biome.Precipitation.RAIN
-            ? level.getRandom().nextFloat() < 0.05F
-            : precipitation == Biome.Precipitation.SNOW && level.getRandom().nextFloat() < 0.1F;
+            ? level.getRandom().nextFloat() < level.purpurConfig.cauldronRainChance // Purpur - Cauldron fill chances
+            : precipitation == Biome.Precipitation.SNOW && level.getRandom().nextFloat() < level.purpurConfig.cauldronPowderSnowChance; // Purpur - Cauldron fill chances
     }
 
     @Override
