--- a/net/minecraft/world/level/block/ChangeOverTimeBlock.java
+++ b/net/minecraft/world/level/block/ChangeOverTimeBlock.java
@@ -51,7 +_,7 @@
         }
 
         float chance = (float)(olderCount + 1) / (olderCount + sameAgeCount + 1);
-        float actualChance = chance * chance * this.getChanceModifier();
+        float actualChance = level.purpurConfig.disableOxidationProximityPenalty ? this.getChanceModifier() : chance * chance * this.getChanceModifier(); // Purpur - option to disable the copper oxidation proximity penalty
         return random.nextFloat() < actualChance ? this.getNext(state) : Optional.empty();
     }
 }
