From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/block/ChangeOverTimeBlock.java b/net/minecraft/world/level/block/ChangeOverTimeBlock.java
index f5bfdef3437cc3fb15ff1170b59598b3dec742eb..94d1fc1983c9eb543c9d16ea82c570e854ed80f0 100644
--- a/net/minecraft/world/level/block/ChangeOverTimeBlock.java
+++ b/net/minecraft/world/level/block/ChangeOverTimeBlock.java
@@ -51,7 +51,7 @@ public interface ChangeOverTimeBlock<T extends Enum<T>> {
         }
 
         float chance = (float)(olderCount + 1) / (olderCount + sameAgeCount + 1);
-        float actualChance = chance * chance * this.getChanceModifier();
+        float actualChance = level.purpurConfig.disableOxidationProximityPenalty ? this.getChanceModifier() : chance * chance * this.getChanceModifier(); // Purpur - option to disable the copper oxidation proximity penalty
         return random.nextFloat() < actualChance ? this.getNext(state) : Optional.empty();
     }
 }
