--- a/net/minecraft/world/level/block/ChestBlock.java
+++ b/net/minecraft/world/level/block/ChestBlock.java
@@ -352,6 +_,7 @@
     }
 
     public static boolean isBlockedChestByBlock(final BlockGetter level, final BlockPos pos) {
+        if (level instanceof Level level1 && level1.purpurConfig.chestOpenWithBlockOnTop) return false; // Purpur - Option for chests to open even with a solid block on top
         BlockPos above = pos.above();
         return level.getBlockState(above).isRedstoneConductor(level, above);
     }
