--- a/net/minecraft/world/level/block/EnderChestBlock.java
+++ b/net/minecraft/world/level/block/EnderChestBlock.java
@@ -78,7 +_,7 @@
         PlayerEnderChestContainer enderChestInventory = player.getEnderChestInventory();
         if (enderChestInventory != null && level.getBlockEntity(pos) instanceof EnderChestBlockEntity enderChestBlockEntity) {
             BlockPos blockPos = pos.above();
-            if (level.getBlockState(blockPos).isRedstoneConductor(level, blockPos)) { // Paper - diff on change; make sure that EnderChest#isBlocked uses the same logic
+            if (level.getBlockState(blockPos).isRedstoneConductor(level, blockPos) && !level.purpurConfig.enderChestOpenWithBlockOnTop) { // Paper - diff on change; make sure that EnderChest#isBlocked uses the same logic // Purpur - Option for ender chests to open even with a solid block on top
                 return InteractionResult.SUCCESS;
             } else {
                 // Paper start - Fix InventoryOpenEvent cancellation - moved up;
