--- a/net/minecraft/world/level/block/FarmBlock.java
+++ b/net/minecraft/world/level/block/FarmBlock.java
@@ -129,6 +_,27 @@
                     return;
                 }
 
+                if (level.purpurConfig.farmlandTramplingDisabled) return; // Purpur - Farmland trampling changes
+                if (level.purpurConfig.farmlandTramplingOnlyPlayers && !(entity instanceof Player)) return; // Purpur - Farmland trampling changes
+                
+                // Purpur start - Ability to re-add farmland mechanics from Alpha
+                if (level.purpurConfig.farmlandAlpha) {
+                    Block block = level.getBlockState(pos.below()).getBlock();
+                    if (block instanceof FenceBlock || block instanceof WallBlock) {
+                        return;
+                    }
+                }
+                // Purpur end - Ability to re-add farmland mechanics from Alpha
+
+                // Purpur start - Farmland trampling changes
+                if (level.purpurConfig.farmlandTramplingFeatherFalling) {
+                    java.util.Iterator<net.minecraft.world.item.ItemStack> armor = ((LivingEntity) entity).getArmorSlots().iterator();
+                    if (armor.hasNext() && net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.FEATHER_FALLING, armor.next()) >= (int) entity.fallDistance) {
+                        return;
+                    }
+                }
+                // Purpur end - Farmland trampling changes
+
                 if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, Blocks.DIRT.defaultBlockState())) {
                     return;
                 }
@@ -174,7 +_,7 @@
             }
         }
 
-        return false;
+        return ((ServerLevel) level).purpurConfig.farmlandGetsMoistFromBelow && level.getFluidState(pos.relative(Direction.DOWN)).is(FluidTags.WATER); // Purpur - Allow soil to moisten from water directly under it
         // Paper end - Perf: remove abstract block iteration
     }
 
