--- a/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
+++ b/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
@@ -38,12 +_,12 @@
 
     @Override
     public BlockState getStateForPlacement(final RandomSource random) {
-        return this.defaultBlockState().setValue(AGE, random.nextInt(25));
+        return this.defaultBlockState().setValue(AGE, getMaxGrowthAge() == 0 ? 0 : random.nextInt(getMaxGrowthAge())); // Purpur - kelp, cave, weeping, and twisting configurable max growth age
     }
 
     @Override
     protected boolean isRandomlyTicking(final BlockState state) {
-        return state.getValue(AGE) < 25;
+        return state.getValue(AGE) < getMaxGrowthAge(); // Purpur - kelp, cave, weeping, and twisting configurable max growth age
     }
 
     @Override
@@ -59,7 +_,7 @@
         } else if (this == Blocks.CAVE_VINES) {
             modifier = level.spigotConfig.caveVinesModifier;
         }
-        if (state.getValue(AGE) < 25 && random.nextDouble() < ((modifier / 100.0D) * this.growPerTickProbability)) { // Spigot - SPIGOT-7159: Better modifier resolution
+        if (state.getValue(AGE) < getMaxGrowthAge() && random.nextDouble() < ((modifier / 100.0D) * this.growPerTickProbability)) { // Spigot - SPIGOT-7159: Better modifier resolution // Purpur - kelp, cave, weeping, and twisting configurable max growth age
             // Spigot end
             BlockPos growthPos = pos.relative(this.growthDirection);
             if (this.canGrowInto(level.getBlockState(growthPos))) {
@@ -79,11 +_,11 @@
     }
 
     public BlockState getMaxAgeState(final BlockState fromState) {
-        return fromState.setValue(AGE, 25);
+        return fromState.setValue(AGE, getMaxGrowthAge()); // Purpur - kelp, cave, weeping, and twisting configurable max growth age
     }
 
     public boolean isMaxAge(final BlockState state) {
-        return state.getValue(AGE) == 25;
+        return state.getValue(AGE) >= getMaxGrowthAge(); // Purpur - kelp, cave, weeping, and twisting configurable max growth age
     }
 
     protected BlockState updateBodyAfterConvertedFromHead(final BlockState headState, final BlockState bodyState) {
@@ -142,13 +_,13 @@
     @Override
     public void performBonemeal(final ServerLevel level, final RandomSource random, final BlockPos pos, final BlockState state) {
         BlockPos forwardPos = pos.relative(this.growthDirection);
-        int nextAge = Math.min(state.getValue(AGE) + 1, 25);
+        int nextAge = Math.min(state.getValue(AGE) + 1, getMaxGrowthAge()); // Purpur - kelp, cave, weeping, and twisting configurable max growth age
         int blocksToGrow = this.getBlocksToGrowWhenBonemealed(random);
 
         for (int i = 0; i < blocksToGrow && this.canGrowInto(level.getBlockState(forwardPos)) && !level.isOutsideBuildHeight(forwardPos); i++) {
             level.setBlockAndUpdate(forwardPos, state.setValue(AGE, nextAge));
             forwardPos = forwardPos.relative(this.growthDirection);
-            nextAge = Math.min(nextAge + 1, 25);
+            nextAge = Math.min(nextAge + 1, getMaxGrowthAge()); // Purpur - kelp, cave, weeping, and twisting configurable max growth age
         }
     }
 
@@ -160,4 +_,6 @@
     protected GrowingPlantHeadBlock getHeadBlock() {
         return this;
     }
+
+    public abstract int getMaxGrowthAge(); // Purpur - kelp, cave, weeping, and twisting configurable max growth age
 }
