--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -65,6 +_,12 @@
 
     @Override
     protected void randomTick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        // Purpur start - Instant leaves decay
+        if (!level.purpurConfig.leavesInstantDecay) checkDecay(state, level, pos, random);
+    }
+
+    protected void checkDecay(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        // Purpur end - Instant leaves decay
         if (this.decaying(state)) {
             // CraftBukkit start
             org.bukkit.event.block.LeavesDecayEvent event = new org.bukkit.event.block.LeavesDecayEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, pos));
@@ -86,6 +_,7 @@
     @Override
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         level.setBlock(pos, updateDistance(state, level, pos), Block.UPDATE_ALL);
+        if (level.purpurConfig.leavesInstantDecay) checkDecay(state, level, pos, random); // Purpur - Instant leaves decay
     }
 
     @Override
