--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -86,6 +_,7 @@
     @Override
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         level.setBlock(pos, updateDistance(state, level, pos), Block.UPDATE_ALL);
+        if (level.purpurConfig.leavesInstantDecay) this.randomTick(state, level, pos, random); // Purpur - Instant leaves decay
     }
 
     @Override
