--- a/net/minecraft/world/level/block/NetherWartBlock.java
+++ b/net/minecraft/world/level/block/NetherWartBlock.java
@@ -16,7 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class NetherWartBlock extends VegetationBlock {
+public class NetherWartBlock extends VegetationBlock implements BonemealableBlock { // Purpur - bonemealable netherwart
     public static final int MAX_AGE = 3;
     public static final IntegerProperty AGE = BlockStateProperties.AGE_3;
     private static final VoxelShape[] SHAPES = Block.boxes(3, age -> Block.column(16.0, 0.0, 5 + age * 3));
@@ -59,4 +_,34 @@
     protected void createBlockStateDefinition(final StateDefinition.Builder<Block, BlockState> builder) {
         builder.add(AGE);
     }
+
+    // Purpur start - Ability for hoe to replant nether warts
+    @Override
+    public void playerDestroy(final ServerLevel level, final net.minecraft.server.level.ServerPlayer player, final BlockPos pos, final BlockState state, final net.minecraft.world.level.block.entity.BlockEntity blockEntity, final ItemStack destroyedWith, boolean includeDrops, boolean dropExp) {
+        if (level.purpurConfig.hoeReplantsNetherWarts && destroyedWith.is(net.minecraft.tags.ItemTags.HOES)) {
+            super.playerDestroyAndReplant(level, player, pos, state, blockEntity, destroyedWith, Items.NETHER_WART);
+        } else {
+            super.playerDestroy(level, player, pos, state, blockEntity, destroyedWith, includeDrops, dropExp);
+        }
+    }
+    // Purpur end - Ability for hoe to replant nether warts
+
+    // Purpur start - bonemealable netherwart
+    @Override
+    public boolean isValidBonemealTarget(final LevelReader level, final BlockPos pos, final BlockState state, final BonemealSource source) {
+        return ((net.minecraft.world.level.Level) level).purpurConfig.netherWartAffectedByBonemeal && state.getValue(NetherWartBlock.AGE) < 3;
+    }
+
+    @Override
+    public boolean isBonemealSuccess(final net.minecraft.world.level.Level level, final RandomSource random, final BlockPos pos, final BlockState state, final BonemealSource source) {
+        return true;
+    }
+
+    @Override
+    public void performBonemeal(final ServerLevel level, final RandomSource random, final BlockPos pos, BlockState state, final BonemealSource source) {
+        int i = Math.min(3, state.getValue(NetherWartBlock.AGE) + 1);
+        state = state.setValue(NetherWartBlock.AGE, i);
+        org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, state, 2); // CraftBukkit
+    }
+    // Purpur end - bonemealable netherwart
 }
