From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/block/ObserverBlock.java b/net/minecraft/world/level/block/ObserverBlock.java
index 6a3413ca875ce78f48cd8d3ce2cfe7798be1a19e..cc60272ff9580dfb57119ad047fb62fdcd6b75cb 100644
--- a/net/minecraft/world/level/block/ObserverBlock.java
+++ b/net/minecraft/world/level/block/ObserverBlock.java
@@ -82,6 +82,7 @@ public class ObserverBlock extends DirectionalBlock {
         final RandomSource random
     ) {
         if (state.getValue(FACING) == directionToNeighbour && !state.getValue(POWERED)) {
+            if (!(level.getWorldBorder().world != null && level.getWorldBorder().world.purpurConfig.disableObserverClocks) || !(neighbourState.getBlock() instanceof ObserverBlock) || neighbourState.getValue(ObserverBlock.FACING).getOpposite() != directionToNeighbour) // Purpur - Add Option for disable observer clocks
             this.startSignal(level, ticks, pos);
         }
 
