--- a/net/minecraft/world/level/block/PointedDripstoneBlock.java
+++ b/net/minecraft/world/level/block/PointedDripstoneBlock.java
@@ -194,20 +_,20 @@
 
     @VisibleForTesting
     public static void maybeTransferFluid(final BlockState state, final ServerLevel level, final BlockPos pos, final float randomValue) {
-        if (!(randomValue > 0.17578125F) || !(randomValue > 0.05859375F)) {
+        if (!(randomValue > level.purpurConfig.cauldronDripstoneWaterFillChance) || !(randomValue > level.purpurConfig.cauldronDripstoneLavaFillChance)) { // Purpur - Cauldron fill chances
             if (isStalactiteStartPos(state, level, pos)) {
                 Optional<PointedDripstoneBlock.FluidInfo> fluidInfo = getFluidAboveStalactite(level, pos, state);
                 if (!fluidInfo.isEmpty()) {
                     Fluid fluid = fluidInfo.get().fluid;
                     float transferProbability;
                     if (fluid == Fluids.WATER) {
-                        transferProbability = 0.17578125F;
+                        transferProbability = level.purpurConfig.cauldronDripstoneWaterFillChance; // Purpur - Cauldron fill chances
                     } else {
                         if (fluid != Fluids.LAVA) {
                             return;
                         }
 
-                        transferProbability = 0.05859375F;
+                        transferProbability = level.purpurConfig.cauldronDripstoneLavaFillChance; // Purpur - Cauldron fill chances
                     }
 
                     if (!(randomValue >= transferProbability)) {
