--- a/net/minecraft/world/level/block/PointedDripstoneBlock.java
+++ b/net/minecraft/world/level/block/PointedDripstoneBlock.java
@@ -194,20 +_,20 @@
 
     @VisibleForTesting
     public static void maybeTransferFluid(BlockState state, ServerLevel level, BlockPos pos, float randChance) {
-        if (!(randChance > 0.17578125F) || !(randChance > 0.05859375F)) {
+        if (!(randChance > level.purpurConfig.cauldronDripstoneWaterFillChance) || !(randChance > level.purpurConfig.cauldronDripstoneLavaFillChance)) { // Purpur - Cauldron fill chances
             if (isStalactiteStartPos(state, level, pos)) {
                 Optional<PointedDripstoneBlock.FluidInfo> fluidAboveStalactite = getFluidAboveStalactite(level, pos, state);
                 if (!fluidAboveStalactite.isEmpty()) {
                     Fluid fluid = fluidAboveStalactite.get().fluid;
                     float f;
                     if (fluid == Fluids.WATER) {
-                        f = 0.17578125F;
+                        f = level.purpurConfig.cauldronDripstoneWaterFillChance; // Purpur - Cauldron fill chances
                     } else {
                         if (fluid != Fluids.LAVA) {
                             return;
                         }
 
-                        f = 0.05859375F;
+                        f = level.purpurConfig.cauldronDripstoneLavaFillChance; // Purpur - Cauldron fill chances
                     }
 
                     if (!(randChance >= f)) {
