--- a/net/minecraft/world/level/block/ShulkerBoxBlock.java
+++ b/net/minecraft/world/level/block/ShulkerBoxBlock.java
@@ -90,7 +_,7 @@
             return true;
         } else {
             AABB aabb = Shulker.getProgressDeltaAabb(1.0F, state.getValue(FACING), 0.0F, 0.5F, pos.getBottomCenter()).deflate(1.0E-6);
-            return level.noCollision(aabb);
+            return level.noCollision(aabb) || level.purpurConfig.shulkerBoxOpenWhenBlocked; // Purpur - Option for shulkerboxes to open even when blocked
         }
     }
 
