From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/block/SnowLayerBlock.java b/net/minecraft/world/level/block/SnowLayerBlock.java
index 6f67b5bdfae6072a61b5e0a730d739601c74b396..9c7cadf1d008843137b299d560beca6daf65900f 100644
--- a/net/minecraft/world/level/block/SnowLayerBlock.java
+++ b/net/minecraft/world/level/block/SnowLayerBlock.java
@@ -76,6 +76,7 @@ public class SnowLayerBlock extends Block {
     @Override
     protected boolean canSurvive(final BlockState state, final LevelReader level, final BlockPos pos) {
         BlockState belowState = level.getBlockState(pos.below());
+        if (belowState.is(Blocks.BLUE_ICE) && !(level.getWorldBorder().world == null || level.getWorldBorder().world.purpurConfig.snowOnBlueIce)) return false; // Purpur - Add config for snow on blue ice
         return !belowState.is(BlockTags.CANNOT_SUPPORT_SNOW_LAYER)
             && (
                 belowState.is(BlockTags.SUPPORT_OVERRIDE_SNOW_LAYER)
