--- a/net/minecraft/world/level/block/SpongeBlock.java
+++ b/net/minecraft/world/level/block/SpongeBlock.java
@@ -53,7 +_,7 @@
 
     private boolean removeWaterBreadthFirstSearch(final Level level, final BlockPos startPos) {
         org.bukkit.craftbukkit.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.util.BlockStateListPopulator(level); // CraftBukkit - Use BlockStateListPopulator
-        BlockPos.breadthFirstTraversal(startPos, 6, 65, (pos, consumer) -> {
+        BlockPos.breadthFirstTraversal(startPos, level.purpurConfig.spongeAbsorptionRadius, level.purpurConfig.spongeAbsorptionArea, (pos, consumer) -> { // Purpur - Configurable sponge absorption
             for (Direction direction : ALL_DIRECTIONS) {
                 consumer.accept(pos.relative(direction));
             }
@@ -65,7 +_,7 @@
                 BlockState blockState = blockList.getBlockState(pos);
                 FluidState fluidState = blockList.getFluidState(pos);
                 // CraftBukkit end
-                if (!fluidState.is(FluidTags.WATER)) {
+                if (!fluidState.is(FluidTags.WATER) && (!level.purpurConfig.spongeAbsorbsLava || !fluidState.is(FluidTags.LAVA)) && (!level.purpurConfig.spongeAbsorbsWaterFromMud || !blockState.is(Blocks.MUD))) { // Purpur - Option for sponges to work on lava and mud
                     return BlockPos.TraversalNodeStatus.SKIP;
                 } else if (blockState.getBlock() instanceof BucketPickup bucketPickup
                     && !bucketPickup.pickupBlock(null, blockList, pos, blockState).isEmpty()) { // CraftBukkit
@@ -73,6 +_,10 @@
                 } else {
                     if (blockState.getBlock() instanceof LiquidBlock) {
                         blockList.setBlock(pos, Blocks.AIR.defaultBlockState(), Block.UPDATE_ALL); // CraftBukkit
+                    // Purpur start - Option for sponges to work on lava and mud
+                    } else if (blockState.is(Blocks.MUD)) {
+                        blockList.setBlock(pos, Blocks.CLAY.defaultBlockState(), Block.UPDATE_ALL);
+                        // Purpur end - Option for sponges to work on lava and mud
                     } else {
                         if (!blockState.is(Blocks.KELP) && !blockState.is(Blocks.KELP_PLANT) && !blockState.is(Blocks.SEAGRASS) && !blockState.is(Blocks.TALL_SEAGRASS)) {
                             return BlockPos.TraversalNodeStatus.SKIP;
