From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/block/SpongeBlock.java b/net/minecraft/world/level/block/SpongeBlock.java
index 604bd261066fbbd98e1b95602187f0b6b3f63010..d7eda43a540e3883a654741e672aba1f05f55cd6 100644
--- a/net/minecraft/world/level/block/SpongeBlock.java
+++ b/net/minecraft/world/level/block/SpongeBlock.java
@@ -53,7 +53,7 @@ public class SpongeBlock extends Block {
 
     private boolean removeWaterBreadthFirstSearch(final Level level, final BlockPos startPos) {
         org.bukkit.craftbukkit.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.util.BlockStateListPopulator(level); // CraftBukkit - Use BlockStateListPopulator
-        BlockPos.breadthFirstTraversal(startPos, 6, 65, (pos, consumer) -> {
+        BlockPos.breadthFirstTraversal(startPos, level.purpurConfig.spongeAbsorptionRadius, level.purpurConfig.spongeAbsorptionArea, (pos, consumer) -> { // Purpur - Configurable sponge absorption
             for (Direction direction : ALL_DIRECTIONS) {
                 consumer.accept(pos.relative(direction));
             }
@@ -65,7 +65,7 @@ public class SpongeBlock extends Block {
                 BlockState blockState = blockList.getBlockState(pos);
                 FluidState fluidState = blockList.getFluidState(pos);
                 // CraftBukkit end
-                if (!fluidState.is(FluidTags.WATER)) {
+                if (!fluidState.is(FluidTags.WATER) && (!level.purpurConfig.spongeAbsorbsLava || !fluidState.is(FluidTags.LAVA)) && (!level.purpurConfig.spongeAbsorbsWaterFromMud || !blockState.is(Blocks.MUD))) { // Purpur - Option for sponges to work on lava and mud
                     return BlockPos.TraversalNodeStatus.SKIP;
                 } else if (blockState.getBlock() instanceof BucketPickup bucketPickup
                     && !bucketPickup.pickupBlock(null, blockList, pos, blockState).isEmpty()) { // CraftBukkit
@@ -73,6 +73,10 @@ public class SpongeBlock extends Block {
                 } else {
                     if (blockState.getBlock() instanceof LiquidBlock) {
                         blockList.setBlock(pos, Blocks.AIR.defaultBlockState(), Block.UPDATE_ALL); // CraftBukkit
+                    // Purpur start - Option for sponges to work on lava and mud
+                    } else if (blockState.is(Blocks.MUD)) {
+                        blockList.setBlock(pos, Blocks.CLAY.defaultBlockState(), Block.UPDATE_ALL);
+                        // Purpur end - Option for sponges to work on lava and mud
                     } else {
                         if (!blockState.is(Blocks.KELP) && !blockState.is(Blocks.KELP_PLANT) && !blockState.is(Blocks.SEAGRASS) && !blockState.is(Blocks.TALL_SEAGRASS)) {
                             return BlockPos.TraversalNodeStatus.SKIP;
