--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -18,7 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class SugarCaneBlock extends Block {
+public class SugarCaneBlock extends Block implements BonemealableBlock { // Purpur - bonemealable sugarcane
     public static final IntegerProperty AGE = BlockStateProperties.AGE_15;
     private static final VoxelShape SHAPE = Block.column(12.0, 0.0, 16.0);
 
@@ -105,4 +_,34 @@
     protected void createBlockStateDefinition(final StateDefinition.Builder<Block, BlockState> builder) {
         builder.add(AGE);
     }
+
+    // Purpur start - bonemealable sugarcane
+    @Override
+    public boolean isValidBonemealTarget(final net.minecraft.world.level.LevelReader level, final net.minecraft.core.BlockPos pos, final net.minecraft.world.level.block.state.BlockState state, final BonemealSource source) {
+        if (!((net.minecraft.world.level.Level) level).purpurConfig.sugarCanAffectedByBonemeal || !level.isEmptyBlock(pos.above())) return false;
+
+        int reedHeight = 0;
+        while (level.getBlockState(pos.below(reedHeight)).is(this)) {
+            reedHeight++;
+        }
+
+        return reedHeight < ((net.minecraft.world.level.Level) level).paperConfig().maxGrowthHeight.reeds;
+    }
+
+    @Override
+    public boolean isBonemealSuccess(final net.minecraft.world.level.Level level, final net.minecraft.util.RandomSource random, final net.minecraft.core.BlockPos pos, final net.minecraft.world.level.block.state.BlockState state, final BonemealSource source) {
+        return true;
+    }
+
+    @Override
+    public void performBonemeal(final net.minecraft.server.level.ServerLevel level, final net.minecraft.util.RandomSource random, final net.minecraft.core.BlockPos pos, final net.minecraft.world.level.block.state.BlockState state, final BonemealSource source) {
+        int reedHeight = 0;
+        while (level.getBlockState(pos.below(reedHeight)).is(this)) {
+            reedHeight++;
+        }
+        for (int i = 0; i <= level.paperConfig().maxGrowthHeight.reeds - reedHeight; i++) {
+            level.setBlockAndUpdate(pos.above(i), state.setValue(SugarCaneBlock.AGE, 0));
+        }
+    }
+    // Purpur end - bonemealable sugarcane
 }
