--- a/net/minecraft/world/level/block/TurtleEggBlock.java
+++ b/net/minecraft/world/level/block/TurtleEggBlock.java
@@ -156,7 +_,7 @@
 
     private boolean shouldUpdateHatchLevel(Level level) {
         float timeOfDay = level.getTimeOfDay(1.0F);
-        return timeOfDay < 0.69 && timeOfDay > 0.65 || level.random.nextInt(500) == 0;
+        return timeOfDay < 0.69 && timeOfDay > 0.65 || level.random.nextInt(level.purpurConfig.turtleEggsRandomTickCrackChance) == 0; // Purpur - Turtle eggs random tick crack chance
     }
 
     @Override
@@ -189,9 +_,32 @@
     }
 
     private boolean canDestroyEgg(ServerLevel level, Entity entity) {
-        return !(entity instanceof Turtle)
-            && !(entity instanceof Bat)
-            && entity instanceof LivingEntity
-            && (entity instanceof Player || level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING));
+        // Purpur start - Add turtle egg block options
+        if (entity instanceof Turtle || entity instanceof Bat) {
+            return false;
+        }
+        if (level.purpurConfig.turtleEggsBreakFromExpOrbs && entity instanceof net.minecraft.world.entity.ExperienceOrb) {
+            return true;
+        }
+        if (level.purpurConfig.turtleEggsBreakFromItems && entity instanceof net.minecraft.world.entity.item.ItemEntity) {
+            return true;
+        }
+        if (level.purpurConfig.turtleEggsBreakFromMinecarts && entity instanceof net.minecraft.world.entity.vehicle.AbstractMinecart) {
+            return true;
+        }
+        if (!(entity instanceof LivingEntity)) {
+            return false;
+        }
+        // Purpur start - Option to disable turtle egg trampling with feather falling
+        if (level.purpurConfig.turtleEggsTramplingFeatherFalling) {
+            net.minecraft.world.item.ItemStack bootsItem = ((net.minecraft.world.entity.LivingEntity) entity).getItemBySlot(net.minecraft.world.entity.EquipmentSlot.FEET);
+
+            return bootsItem != net.minecraft.world.item.ItemStack.EMPTY || net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.FEATHER_FALLING, bootsItem) < (int) entity.fallDistance;
+        }
+        // Purpur end - Option to disable turtle egg trampling with feather falling
+        if (entity instanceof Player) return true;
+
+        return level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
+        // Purpur end - Add turtle egg block options
     }
 }
