From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/entity/monster/piglin/PiglinAi.java b/net/minecraft/world/entity/monster/piglin/PiglinAi.java
index 7d0494dc56bf6dd43e303798c25887e9e0b7e279..1e7232693536eebbdf7cd516667c6f7ede337cfb 100644
--- a/net/minecraft/world/entity/monster/piglin/PiglinAi.java
+++ b/net/minecraft/world/entity/monster/piglin/PiglinAi.java
@@ -660,8 +660,11 @@ public class PiglinAi {
     }
 
     public static boolean isWearingSafeArmor(final LivingEntity livingEntity) {
-        for (EquipmentSlot slot : EquipmentSlotGroup.ARMOR) {
-            if (livingEntity.getItemBySlot(slot).is(ItemTags.PIGLIN_SAFE_ARMOR)) {
+        for (EquipmentSlot equipmentSlot : EquipmentSlotGroup.ARMOR) {
+            // Purpur start - piglins ignore gold-trimmed armor
+            net.minecraft.world.item.ItemStack itemStack = livingEntity.getItemBySlot(equipmentSlot);
+            if (itemStack.is(ItemTags.PIGLIN_SAFE_ARMOR) || (livingEntity.level().purpurConfig.piglinIgnoresArmorWithGoldTrim && isWearingGoldTrim(itemStack))) {
+            // Purpur end - piglins ignore gold-trimmed armor
                 return true;
             }
         }
@@ -669,6 +672,13 @@ public class PiglinAi {
         return false;
     }
 
+    // Purpur start - piglins ignore gold-trimmed armor
+    private static boolean isWearingGoldTrim(net.minecraft.world.item.ItemStack itemstack) {
+        net.minecraft.world.item.equipment.trim.ArmorTrim armorTrim = itemstack.getComponents().get(net.minecraft.core.component.DataComponents.TRIM);
+        return armorTrim != null && armorTrim.material().is(net.minecraft.world.item.equipment.trim.TrimMaterials.GOLD);
+    }
+    // Purpur end - piglins ignore gold-trimmed armor
+
     private static void stopWalking(final Piglin body) {
         body.getBrain().eraseMemory(MemoryModuleType.WALK_TARGET);
         body.getNavigation().stop();
