From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/server/commands/GiveCommand.java b/net/minecraft/server/commands/GiveCommand.java
index 5bd2c78b49539c20351562774313d5ce2b5ee94d..20ff0db2124b0832d150d17704f234b6010ea63a 100644
--- a/net/minecraft/server/commands/GiveCommand.java
+++ b/net/minecraft/server/commands/GiveCommand.java
@@ -62,6 +62,7 @@ public class GiveCommand {
                 remaining -= size;
                 ItemStack copyToDrop = prototypeItemStack.copyWithCount(size);
                 boolean added = player.getInventory().add(copyToDrop);
+                if (org.purpurmc.purpur.PurpurConfig.disableGiveCommandDrops) continue; // Purpur - add config option for toggling give command dropping
                 if (added && copyToDrop.isEmpty()) {
                     ItemEntity drop = player.drop(prototypeItemStack.copy(), false, false, false, null); // Paper - do not fire PlayerDropItemEvent for /give command
                     if (drop != null) {
