From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/gametest/framework/GameTestHelper.java b/net/minecraft/gametest/framework/GameTestHelper.java
index b49efbf597860ac19c3cc308550974b0a0d03314..64a82b414afd093c010878c150592c681665b735 100644
--- a/net/minecraft/gametest/framework/GameTestHelper.java
+++ b/net/minecraft/gametest/framework/GameTestHelper.java
@@ -354,6 +354,8 @@ public class GameTestHelper {
                 return gameType;
             }
 
+            public void setAfk(final boolean afk) {} // Purpur - AFK API
+
             @Override
             public boolean isClientAuthoritative() {
                 return false;
