From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/gametest/framework/TestCommand.java b/net/minecraft/gametest/framework/TestCommand.java
index 5a85809882328436f4f34194435fd046fb1d259b..528408e5c58c4d574eef010fc07ab59add68a0f6 100644
--- a/net/minecraft/gametest/framework/TestCommand.java
+++ b/net/minecraft/gametest/framework/TestCommand.java
@@ -424,7 +424,7 @@ public class TestCommand {
                             )
                     )
             );
-        if (SharedConstants.IS_RUNNING_IN_IDE) {
+        if (org.purpurmc.purpur.PurpurConfig.registerMinecraftDebugCommands || SharedConstants.IS_RUNNING_IN_IDE) { // Purpur - register minecraft debug commands
             testCommand = testCommand.then(
                     Commands.literal("export")
                         .then(
