--- a/net/minecraft/world/entity/animal/horse/Llama.java
+++ b/net/minecraft/world/entity/animal/horse/Llama.java
@@ -72,6 +_,7 @@
     private Llama caravanHead;
     @Nullable
     public Llama caravanTail; // Paper
+    public boolean shouldJoinCaravan = true; // Purpur - Llama API
 
     public Llama(EntityType<? extends Llama> entityType, Level level) {
         super(entityType, level);
@@ -106,6 +_,7 @@
         super.addAdditionalSaveData(compound);
         compound.putInt("Variant", this.getVariant().id);
         compound.putInt("Strength", this.getStrength());
+        compound.putBoolean("Purpur.ShouldJoinCaravan", shouldJoinCaravan); // Purpur - Llama API
     }
 
     @Override
@@ -113,6 +_,7 @@
         this.setStrength(compound.getInt("Strength"));
         super.readAdditionalSaveData(compound);
         this.setVariant(Llama.Variant.byId(compound.getInt("Variant")));
+        if (compound.contains("Purpur.ShouldJoinCaravan")) this.shouldJoinCaravan = compound.getBoolean("Purpur.ShouldJoinCaravan"); // Purpur - Llama API
     }
 
     @Override
@@ -386,6 +_,7 @@
 
     public void leaveCaravan() {
         if (this.caravanHead != null) {
+            new org.purpurmc.purpur.event.entity.LlamaLeaveCaravanEvent((org.bukkit.entity.Llama) getBukkitEntity()).callEvent(); // Purpur - Llama API
             this.caravanHead.caravanTail = null;
         }
 
@@ -393,6 +_,7 @@
     }
 
     public void joinCaravan(Llama caravanHead) {
+        if (!shouldJoinCaravan || !new org.purpurmc.purpur.event.entity.LlamaJoinCaravanEvent((org.bukkit.entity.Llama) getBukkitEntity(), (org.bukkit.entity.Llama) caravanHead.getBukkitEntity()).callEvent()) return; // Purpur - Llama API
         this.caravanHead = caravanHead;
         this.caravanHead.caravanTail = this;
     }
