From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ben Kerllenevich <ben@omega24.dev>
Date: Sat, 12 Jun 2021 12:35:38 -0400
Subject: [PATCH] Build System Changes

todo: merge with rebrand patch

diff --git a/build.gradle.kts b/build.gradle.kts
index 241ee5c98275a4517e040c86355ae6702f8efca1..43ce7fe6b65488d00c69986908c788761ccd74ad 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -40,6 +40,7 @@ dependencies {
     apiAndDocs("net.kyori:adventure-text-serializer-plain")
     api("org.apache.logging.log4j:log4j-api:2.14.1")
     api("org.slf4j:slf4j-api:1.8.0-beta4")
+    api("net.kyori:adventure-text-minimessage:4.1.0-SNAPSHOT") // Purpur
 
     implementation("org.ow2.asm:asm:9.2")
     implementation("org.ow2.asm:asm-commons:9.2")
