From c5475c4147ad18323c8dc56755e7b6f8df714870 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Fri, 3 May 2019 23:21:23 -0500
Subject: [PATCH] cows to mooshroom when fed mushrooms

---
 .../paper/event/entity/EntityTransformedEvent.java          | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/src/main/java/com/destroystokyo/paper/event/entity/EntityTransformedEvent.java b/src/main/java/com/destroystokyo/paper/event/entity/EntityTransformedEvent.java
index 12194f1f..8b2e533d 100644
--- a/src/main/java/com/destroystokyo/paper/event/entity/EntityTransformedEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/entity/EntityTransformedEvent.java
@@ -83,6 +83,12 @@ public class EntityTransformedEvent extends EntityEvent implements Cancellable {
          * When a mooshroom turns to a cow
          */
         SHEARED,
+        // Purpur start
+        /**
+         * When a cow turns to a mooshroom
+         */
+        SHROOMED,
+        // Purpur end
         /**
          * When a pig turns to a zombiepigman
          */
-- 
2.20.1

