From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: granny <granny@purpurmc.org>
Date: Thu, 16 May 2024 19:11:29 -0700
Subject: [PATCH] Rebrand


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index a7c0936a1ee80618f9a504087c88814f738f4693..7196594e07af19a14c320d77df893978525fe386 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -26,6 +26,13 @@ public interface ServerBuildInfo {
     Key BRAND_PUFFERFISH_ID = Key.key("pufferfish", "pufferfish");
     // Purpur end
 
+    // Purpur start
+    /**
+     * The brand id for Purpur.
+     */
+    Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
+    // Purpur end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
