From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: granny <granny@purpurmc.org>
Date: Thu, 16 May 2024 19:11:29 -0700
Subject: [PATCH] Rebrand


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 2655f98322bcab8622e5f84db1085c3b1f5174b7..8c2c5ea14a839c8767bbf120ae5512f65475cbb0 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -25,6 +25,13 @@ public interface ServerBuildInfo {
     Key BRAND_PUFFERFISH_ID = Key.key("pufferfish", "pufferfish");
     // Purpur end
 
+    // Purpur start
+    /**
+     * The brand id for Purpur.
+     */
+    Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
+    // Purpur end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
