From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Sat, 19 Oct 2019 02:35:24 -0500
Subject: [PATCH] SkeletonHorseTrapEvent#getPlayer


diff --git a/src/main/java/com/destroystokyo/paper/event/entity/SkeletonHorseTrapEvent.java b/src/main/java/com/destroystokyo/paper/event/entity/SkeletonHorseTrapEvent.java
index d79dbcd68..1ed3ea8fa 100644
--- a/src/main/java/com/destroystokyo/paper/event/entity/SkeletonHorseTrapEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/entity/SkeletonHorseTrapEvent.java
@@ -1,20 +1,30 @@
 package com.destroystokyo.paper.event.entity;
 
+import org.bukkit.entity.Player;
 import org.bukkit.entity.SkeletonHorse;
 import org.bukkit.event.Cancellable;
 import org.bukkit.event.HandlerList;
 import org.bukkit.event.entity.EntityEvent;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 /**
  * Event called when a player gets close to a skeleton horse and triggers the lightning trap
  */
 public class SkeletonHorseTrapEvent extends EntityEvent implements Cancellable {
     private static final HandlerList handlers = new HandlerList();
+    private final Player player; // Purpur
     private boolean cancelled;
 
     public SkeletonHorseTrapEvent(@NotNull SkeletonHorse horse) {
+        // Purpur start
+        this(horse, null);
+    }
+
+    public SkeletonHorseTrapEvent(@NotNull SkeletonHorse horse, @Nullable Player player) {
+        // Purpur end
         super(horse);
+        this.player = player; // Purpur
     }
 
     @NotNull
@@ -23,6 +33,20 @@ public class SkeletonHorseTrapEvent extends EntityEvent implements Cancellable {
         return (SkeletonHorse) super.getEntity();
     }
 
+    // Purpur start
+    /**
+     * Get the player that triggered this trap event
+     * <p>
+     * Null value if called by a plugin that does not support this feature
+     *
+     * @return Player that triggered event
+     */
+    @Nullable
+    public Player getPlayer() {
+        return player;
+    }
+    // Purpur end
+
     @Override
     public boolean isCancelled() {
         return cancelled;
