From 2a524b7c61b666f93a27174acbf847a1b0b9959c Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Sat, 19 Oct 2019 02:25:05 -0500
Subject: [PATCH] Add playPickupItemAnimation to LivingEntity

---
 .../java/org/bukkit/entity/LivingEntity.java    | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 9c4af8a9b..aaad2da41 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -720,5 +720,22 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      * @param safeFallDistance Safe fall distance
      */
     void setSafeFallDistance(float safeFallDistance);
+
+    /**
+     * Plays pickup item animation
+     *
+     * @param item Item to pickup
+     */
+    default void playPickupItemAnimation(@NotNull Item item) {
+        playPickupItemAnimation(item, item.getItemStack().getAmount());
+    }
+
+    /**
+     * Plays pickup item animation
+     *
+     * @param item Item to pickup
+     * @param quantity Quantity of item
+     */
+    void playPickupItemAnimation(@NotNull Item item, int quantity);
     // Purpur end
 }
-- 
2.24.0

