From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Mon, 30 Nov 2020 06:02:54 -0600
Subject: [PATCH] Left handed API


diff --git a/src/main/java/org/bukkit/entity/Mob.java b/src/main/java/org/bukkit/entity/Mob.java
index d726453c041a980576312b6bee96a07837f37974..9d4eae5cf0bf8d01954db85b431bcdca8490ee8f 100644
--- a/src/main/java/org/bukkit/entity/Mob.java
+++ b/src/main/java/org/bukkit/entity/Mob.java
@@ -64,4 +64,20 @@ public interface Mob extends LivingEntity, Lootable {
      * @return whether the mob is aware
      */
     public boolean isAware();
+
+    // Purpur start
+    /**
+     * Check if Mob is left-handed
+     *
+     * @return True if left-handed
+     */
+    public boolean isLeftHanded();
+
+    /**
+     * Set if Mob is left-handed
+     *
+     * @param leftHanded True if left-handed
+     */
+    public void setLeftHanded(boolean leftHanded);
+    // Purpur end
 }
