From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Tue, 24 Nov 2020 04:30:34 -0600
Subject: [PATCH] Add critical hit check to EntityDamagedByEntityEvent


diff --git a/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java b/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java
index 869bad740..7ef65376a 100644
--- a/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java
@@ -10,6 +10,7 @@ import org.jetbrains.annotations.NotNull;
  */
 public class EntityDamageByEntityEvent extends EntityDamageEvent {
     private final Entity damager;
+    private boolean isCritical; // Purpur
 
     public EntityDamageByEntityEvent(@NotNull final Entity damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, final double damage) {
         super(damagee, cause, damage);
@@ -30,4 +31,25 @@ public class EntityDamageByEntityEvent extends EntityDamageEvent {
     public Entity getDamager() {
         return damager;
     }
+
+    // Purpur start
+
+    /**
+     * Whether this damage was done by a critical hit
+     *
+     * @return True if critical hit
+     */
+    public boolean isCritical() {
+        return this.isCritical;
+    }
+
+    /**
+     * Set whether this damage was done by a critical hit
+     *
+     * @param isCritical True if critical hit
+     */
+    public void setCritical(boolean isCritical) {
+        this.isCritical = isCritical;
+    }
+    // Purpur end
 }
