From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Sun, 23 Aug 2020 16:29:01 -0500
Subject: [PATCH] Fix params in javadocs


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 36bb9d37e..b4e55c173 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -175,7 +175,7 @@ public final class Bukkit {
     /**
      * Set the maximum amount of players which can login to this server.
      *
-     * @param the amount of players this server allows
+     * @param maxPlayers the amount of players this server allows
      */
     public static void setMaxPlayers(int maxPlayers) {
         server.setMaxPlayers(maxPlayers);
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 5b04a7699..70bcd6c3a 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -148,7 +148,7 @@ public interface Server extends PluginMessageRecipient {
     /**
      * Set the maximum amount of players which can login to this server.
      *
-     * @param the amount of players this server allows
+     * @param maxPlayers the amount of players this server allows
      */
     public void setMaxPlayers(int maxPlayers);
     // Paper end
