From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Encode42 <me@encode42.dev>
Date: Tue, 8 Dec 2020 17:15:15 -0500
Subject: [PATCH] Wolf naturally spawn rabid


diff --git a/src/main/java/org/bukkit/entity/Wolf.java b/src/main/java/org/bukkit/entity/Wolf.java
index 0e5decadf..26e3d6516 100644
--- a/src/main/java/org/bukkit/entity/Wolf.java
+++ b/src/main/java/org/bukkit/entity/Wolf.java
@@ -39,4 +39,21 @@ public interface Wolf extends Tameable, Sittable {
      * @param color the color to apply
      */
     public void setCollarColor(@NotNull DyeColor color);
+
+    // Purpur start
+
+    /**
+     * Check if this wolf is rabid
+     *
+     * @return True if rabid
+     */
+    public boolean isRabid();
+
+    /**
+     * Set if this wolf is rabid
+     *
+     * @param isRabid True if rabid
+     */
+    public void setRabid(boolean isRabid);
+    // Purpur end
 }
