From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Encode42 <me@encode42.dev>
Date: Tue, 11 May 2021 15:18:25 -0400
Subject: [PATCH] Allow Creepers and Spiders to burn in day


diff --git a/src/main/java/org/bukkit/entity/Creeper.java b/src/main/java/org/bukkit/entity/Creeper.java
index e144f618702122bf28ebedc5cb8ce0f6ef27c107..790fee13dff90c41662c75d20ce3683946d162ac 100644
--- a/src/main/java/org/bukkit/entity/Creeper.java
+++ b/src/main/java/org/bukkit/entity/Creeper.java
@@ -103,4 +103,20 @@ public interface Creeper extends Monster {
      */
     public boolean isIgnited();
     // Paper end
+
+    // Purpur start
+    /**
+     * Check if this zombie will burn in the sunlight
+     *
+     * @return True if zombie will burn in sunlight
+     */
+    boolean shouldBurnInDay();
+
+    /**
+     * Set if this zombie should burn in the sunlight
+     *
+     * @param shouldBurnInDay True to burn in sunlight
+     */
+    void setShouldBurnInDay(boolean shouldBurnInDay);
+    // Purpur end
 }
diff --git a/src/main/java/org/bukkit/entity/Spider.java b/src/main/java/org/bukkit/entity/Spider.java
index f9ee8cc75a6eae3c1fad58b57c70f64b71977c59..5a3a8e676d56bd4d89befef8b7e697da4c229f50 100644
--- a/src/main/java/org/bukkit/entity/Spider.java
+++ b/src/main/java/org/bukkit/entity/Spider.java
@@ -3,4 +3,20 @@ package org.bukkit.entity;
 /**
  * Represents a Spider.
  */
-public interface Spider extends Monster {}
+public interface Spider extends Monster {
+    // Purpur start
+    /**
+     * Check if this zombie will burn in the sunlight
+     *
+     * @return True if zombie will burn in sunlight
+     */
+    boolean shouldBurnInDay();
+
+    /**
+     * Set if this zombie should burn in the sunlight
+     *
+     * @param shouldBurnInDay True to burn in sunlight
+     */
+    void setShouldBurnInDay(boolean shouldBurnInDay);
+    // Purpur end
+}
