From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ben Kerllenevich <me@notom3ga.me>
Date: Tue, 25 May 2021 16:30:30 -0400
Subject: [PATCH] API for any mob to burn daylight


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 42811d18ff304082f74f45794344891208599c04..ebe10b5612835d22dfcf8b30b0f028022e47216e 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -872,5 +872,19 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      * @param slot Equipment slot to play break animation for
      */
     void broadcastItemBreak(@NotNull org.bukkit.inventory.EquipmentSlot slot);
+
+    /**
+     * If this mob will burn in the sunlight
+     *
+     * @return True if mob will burn in sunlight
+     */
+    boolean shouldBurnInDay();
+
+    /**
+     * Set if this mob should burn in the sunlight
+     *
+     * @param shouldBurnInDay True to burn in sunlight
+     */
+    void setShouldBurnInDay(boolean shouldBurnInDay);
     // Purpur end
 }
diff --git a/src/main/java/org/bukkit/entity/Phantom.java b/src/main/java/org/bukkit/entity/Phantom.java
index a40b045f08b85e22e75459b547e7e7c0b95103ed..5523d0a48bae3287767f507d6702511cadecb931 100644
--- a/src/main/java/org/bukkit/entity/Phantom.java
+++ b/src/main/java/org/bukkit/entity/Phantom.java
@@ -31,14 +31,18 @@ public interface Phantom extends Flying {
      * Check if this phantom will burn in the sunlight
      *
      * @return True if phantom will burn in sunlight
+     * @deprecated Use {@link LivingEntity#shouldBurnInDay()}
      */
+    @Deprecated // Purpur
     public boolean shouldBurnInDay();

     /**
      * Set if this phantom should burn in the sunlight
      *
      * @param shouldBurnInDay True to burn in sunlight
+     * @deprecated Use {@link LivingEntity#setShouldBurnInDay(boolean)}
      */
+    @Deprecated // Purpur
     public void setShouldBurnInDay(boolean shouldBurnInDay);
     // Paper end
 }
diff --git a/src/main/java/org/bukkit/entity/Skeleton.java b/src/main/java/org/bukkit/entity/Skeleton.java
index 581abc69290ca14b8e64f50fdf5a49c14be13940..6e71ce7094d9175513e333dd8ee0e31f53a34a14 100644
--- a/src/main/java/org/bukkit/entity/Skeleton.java
+++ b/src/main/java/org/bukkit/entity/Skeleton.java
@@ -53,14 +53,18 @@ public interface Skeleton extends Monster, RangedEntity { // Paper
      * Check if this skeleton will burn in the sunlight
      *
      * @return True if skeleton will burn in sunlight
+     * @deprecated Use {@link LivingEntity#shouldBurnInDay()}
      */
+    @Deprecated // Purpur
     boolean shouldBurnInDay();

     /**
      * Set if this skeleton should burn in the sunlight
      *
      * @param shouldBurnInDay True to burn in sunlight
+     * @deprecated Use {@link LivingEntity#setShouldBurnInDay(boolean)}
      */
+    @Deprecated // Purpur
     void setShouldBurnInDay(boolean shouldBurnInDay);
     // Paper end
 }
diff --git a/src/main/java/org/bukkit/entity/Zombie.java b/src/main/java/org/bukkit/entity/Zombie.java
index a39fbc9fa0903be8ed8e89f3ef39f93c02dfc90b..07109aca15376fee2ec72776291041e793b84c36 100644
--- a/src/main/java/org/bukkit/entity/Zombie.java
+++ b/src/main/java/org/bukkit/entity/Zombie.java
@@ -131,14 +131,18 @@ public interface Zombie extends Monster, Ageable {
      * Check if this zombie will burn in the sunlight
      *
      * @return True if zombie will burn in sunlight
+     * @deprecated Use {@link LivingEntity#shouldBurnInDay()}
      */
+    @Deprecated // Purpur
     boolean shouldBurnInDay();

     /**
      * Set if this zombie should burn in the sunlight
      *
      * @param shouldBurnInDay True to burn in sunlight
+     * @deprecated Use {@link LivingEntity#setShouldBurnInDay(boolean)}
      */
+    @Deprecated // Purpur
     void setShouldBurnInDay(boolean shouldBurnInDay);

     /**
