From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Mon, 5 Jul 2021 00:58:48 -0500
Subject: [PATCH] API to geta plugin from class name


diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
index 0e25119564dfa9cb12f3c5dc5f653d7f2c147a9d..ce9680f93886499519bca3934239bc24a972c745 100644
--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
@@ -923,4 +923,18 @@ public final class SimplePluginManager implements PluginManager {
     }
     // Paper end
 
+    // Purpur start
+    @Nullable
+    public Class<?> getClass(@NotNull String name) {
+        Class<?> clazz;
+        for (Plugin plugin : this.plugins) {
+            clazz = ((org.bukkit.plugin.java.JavaPluginLoader) plugin.getPluginLoader()).getClass(name);
+            if (clazz != null) {
+                return clazz;
+            }
+        }
+        return null;
+    }
+    // Purpur end
+
 }
diff --git a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
index 04fa3991f6ce4e9dad804f28fc6c947695857089..b0277f39e69945d12e18c393233223855b8a8a4f 100644
--- a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
+++ b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
@@ -426,4 +426,12 @@ public abstract class JavaPlugin extends PluginBase {
         }
         return plugin;
     }
+
+    // Purpur start
+    @Nullable
+    public static JavaPlugin getProvidingPlugin(@NotNull String name) {
+        Class<?> clazz = ((org.bukkit.plugin.SimplePluginManager) org.bukkit.Bukkit.getPluginManager()).getClass(name);
+        return clazz == null ? null : getProvidingPlugin(clazz);
+    }
+    // Purpur end
 }
diff --git a/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java b/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java
index b622cedeeee017f042bcf92485d81832030a8030..b7e415004bd648fe6a9aaec6bb9e66eb5cda16e1 100644
--- a/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java
@@ -435,4 +435,18 @@ public final class JavaPluginLoader implements PluginLoader {
             }
         }
     }
+
+    // Purpur start
+    @Nullable
+    public Class<?> getClass(@NotNull String name) {
+        Class<?> clazz;
+        for (PluginClassLoader loader : this.loaders) {
+            clazz = loader.getClass(name);
+            if (clazz != null) {
+                return clazz;
+            }
+        }
+        return null;
+    }
+    // Purpur end
 }
diff --git a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
index 81292899918c4dc880661ee628384cb840a6244f..cf75fa2eb8721d67aa15dbddf0f1da98bfe82733 100644
--- a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
@@ -242,4 +242,11 @@ public final class PluginClassLoader extends URLClassLoader { // Spigot
                    '}';
     }
     // Paper end
+
+    // Purpur start
+    @Nullable
+    public Class<?> getClass(@NotNull String name) {
+        return this.classes.get(name);
+    }
+    // Purpur end
 }
