From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: granny <granny@pl3x.net>
Date: Thu, 30 Jun 2022 02:32:52 -0700
Subject: [PATCH] Spark Profiler


diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
index aed2209b66f11a0d03473cf19437f3da0e9e573a..e46c3cb737d644d167b4426b29bb4ade0c24e74e 100644
--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
@@ -164,6 +164,12 @@ public final class SimplePluginManager implements PluginManager {
                 server.getLogger().log(Level.SEVERE, "Could not load '" + file.getPath() + "' in folder '" + file.getParentFile().getPath() + "'", ex); // Paper
                 continue;
             }
+            // Purpur start
+            if (!Boolean.getBoolean("Purpur.IReallyDontWantSpark") && plugins.containsKey(description.getName()) && description.getName().equalsIgnoreCase("spark")) {
+                server.getLogger().log(Level.INFO, "Purpur: Using user-provided spark plugin instead of our own.");
+                continue;
+            }
+            // Purpur end
 
             File replacedFile = plugins.put(description.getName(), file);
             if (replacedFile != null) {
