From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MelnCat <melncatuwu@gmail.com>
Date: Fri, 23 Sep 2022 18:35:28 -0700
Subject: [PATCH] Add death screen API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index a003254281a07505cdd929fa9ee443749e48d5eb..dc437885404ae147a06cac653e519a4674a9a951 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3186,5 +3186,21 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * Clears all debug block highlights
      */
     void clearBlockHighlights();
+
+    /**
+     * Sends a player the death screen with a specified death message.
+     *
+     * @param message The death message to show the player
+     */
+    void sendDeathScreen(@NotNull net.kyori.adventure.text.Component message);
+
+    /**
+     * Sends a player the death screen with a specified death message,
+     * along with the entity that caused the death.
+     *
+     * @param message The death message to show the player
+     * @param killer The entity that killed the player
+     */
+    void sendDeathScreen(@NotNull net.kyori.adventure.text.Component message, @Nullable Entity killer);
     // Purpur end
 }
