From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ben Kerllenevich <ben@omega24.dev>
Date: Mon, 18 Jul 2022 21:51:49 -0400
Subject: [PATCH] Mod name API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 2436ece75493dc05352dfa0027b98aaeaab0d5f7..213a754ec2948dba861e55d83ed0aad0f92515db 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2767,5 +2767,12 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @return True if fall damage is enabled when {@link #getAllowFlight()} is true
      */
     public boolean hasFlyingFallDamage();
+
+    /**
+     * Allows you to set the mod name that is used on the client
+     *
+     * @param component New mod name
+     */
+    public void setServerModName(@NotNull Component component);
     // Purpur end
 }
