From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: granny <granny@pl3x.net>
Date: Fri, 26 Aug 2022 21:45:01 -0700
Subject: [PATCH] add Player#showCreditScreen


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 5e06858d68f725508e0055de3f5d8c0f6a56601b..8ab346562c23ff2c14727408e61e75e324885e67 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2375,6 +2375,14 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public void showDemoScreen();
 
+    /**
+     * Shows the credits to the player, this screen is normally only seen after
+     * going into the end portal for the first time.
+     * <br>
+     * Servers can modify the text on this screen using a resource pack.
+     */
+    public void showCreditScreen();
+
     /**
      * Gets whether the player has the "Allow Server Listings" setting enabled.
      *
