From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: granny <granny@purpurmc.org>
Date: Fri, 26 Aug 2022 21:45:01 -0700
Subject: [PATCH] add Player#showCreditScreen


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 6782a41eb845c69e9749e340d9fe8d6402caecc2..faa2954f68cf5c675e732bd1c6b456321ccb2ec2 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2411,6 +2411,14 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public void showDemoScreen();
 
+    /**
+     * Shows the credits to the player, this screen is normally only seen after
+     * going into the end portal for the first time.
+     * <br>
+     * Servers can modify the text on this screen using a resource pack.
+     */
+    public void showCreditScreen();
+
     /**
      * Gets whether the player has the "Allow Server Listings" setting enabled.
      *
