From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MelnCat <melncatuwu@gmail.com>
Date: Fri, 23 Sep 2022 18:35:28 -0700
Subject: [PATCH] Add death screen API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index f50c61a8c375af03f3f0f5469e376e3f9c19f03e..872a1c0f4ac33ee6739b3d73ee99670da05e33c6 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2948,5 +2948,21 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * Clears all debug block highlights
      */
     void clearBlockHighlights();
+
+    /**
+     * Sends a player the death screen with a specified death message.
+     *
+     * @param message The death message to show the player
+     */
+    void sendDeathScreen(@NotNull Component message);
+
+    /**
+     * Sends a player the death screen with a specified death message,
+     * along with the entity that caused the death.
+     *
+     * @param message The death message to show the player
+     * @param killer The entity that killed the player
+     */
+    void sendDeathScreen(@NotNull Component message, @Nullable Entity killer);
     // Purpur end
 }
