From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MelnCat <melncatuwu@gmail.com>
Date: Fri, 23 Sep 2022 18:35:28 -0700
Subject: [PATCH] Add death screen API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index caf710fbfa91c2e2bba51ad395cf14cec91feb21..93e18658bb7a83a35f812811c2c87630a4b7eb0b 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3043,5 +3043,21 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * Clears all debug block highlights
      */
     void clearBlockHighlights();
+
+    /**
+     * Sends a player the death screen with a specified death message.
+     *
+     * @param message The death message to show the player
+     */
+    void sendDeathScreen(@NotNull Component message);
+
+    /**
+     * Sends a player the death screen with a specified death message,
+     * along with the entity that caused the death.
+     *
+     * @param message The death message to show the player
+     * @param killer The entity that killed the player
+     */
+    void sendDeathScreen(@NotNull Component message, @Nullable Entity killer);
     // Purpur end
 }
