From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: granny <granny@purpurmc.org>
Date: Wed, 4 Oct 2023 00:27:09 -0700
Subject: [PATCH] temporarily comment out javadoc.io links in javadoc


diff --git a/build.gradle.kts b/build.gradle.kts
index f22c0403729b3554ff769b06f7455516b15fd974..9ee4cc8afc99dd4da9f76bb7fb6740f5e715c1cc 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -117,12 +117,12 @@ tasks.withType<Javadoc> {
     options.isDocFilesSubDirs = true
     options.links(
         "https://guava.dev/releases/32.1.2-jre/api/docs/",
-        "https://javadoc.io/doc/org.yaml/snakeyaml/2.2/",
-        "https://javadoc.io/doc/org.jetbrains/annotations/$annotationsVersion/", // Paper - we don't want Java 5 annotations
+        // "https://javadoc.io/doc/org.yaml/snakeyaml/2.2/", // Purpur
+        // "https://javadoc.io/doc/org.jetbrains/annotations/$annotationsVersion/", // Paper - we don't want Java 5 annotations // Purpur
         // "https://javadoc.io/doc/net.md-5/bungeecord-chat/$bungeeCordChatVersion/", // Paper - don't link to bungee chat
         // Paper start - add missing javadoc links
-        "https://javadoc.io/doc/org.joml/joml/1.10.5/index.html",
-        "https://www.javadoc.io/doc/com.google.code.gson/gson/2.10.1",
+        // "https://javadoc.io/doc/org.joml/joml/1.10.5/index.html", // Purpur
+        // "https://www.javadoc.io/doc/com.google.code.gson/gson/2.10.1", // Purpur
         // Paper end
         // Paper start
         "https://jd.advntr.dev/api/$adventureVersion/",
@@ -131,10 +131,10 @@ tasks.withType<Javadoc> {
         "https://jd.advntr.dev/text-serializer-legacy/$adventureVersion/",
         "https://jd.advntr.dev/text-serializer-plain/$adventureVersion/",
         "https://jd.advntr.dev/text-logger-slf4j/$adventureVersion/",
-        "https://javadoc.io/doc/org.slf4j/slf4j-api/$slf4jVersion/",
-        "https://javadoc.io/doc/org.apache.logging.log4j/log4j-api/$log4jVersion/",
+        // "https://javadoc.io/doc/org.slf4j/slf4j-api/$slf4jVersion/", // Purpur
+        // "https://javadoc.io/doc/org.apache.logging.log4j/log4j-api/$log4jVersion/", // Purpur
         // Paper end
-        "https://javadoc.io/doc/org.apache.maven.resolver/maven-resolver-api/1.7.3", // Paper
+        // "https://javadoc.io/doc/org.apache.maven.resolver/maven-resolver-api/1.7.3", // Paper // Purpur
     )
     options.tags("apiNote:a:API Note:")
 
