From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Racci <90304606+DaRacci@users.noreply.github.com>
Date: Fri, 4 Feb 2022 16:09:47 +1100
Subject: [PATCH] Fire Immunity API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 28be71aa677ed8cb9a9d06b3aba7af40f0992f68..3959765b110b6fed91348c11ff85e302a885776c 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -988,5 +988,18 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      * @return True if in daylight
      */
     boolean isInDaylight();
+
+    /**
+     * Checks if the entity is fire immune
+     *
+     * @return True if fire immune
+     */
+    boolean isImmuneToFire();
+
+    /**
+     * Sets if the entity is fire immune
+     * Set this to null to restore the entity type default
+     */
+    void setImmuneToFire(@Nullable Boolean fireImmune);
     // Purpur end
 }
diff --git a/src/main/java/org/bukkit/entity/Item.java b/src/main/java/org/bukkit/entity/Item.java
index fd213c026fdfb3030d0c37f651a88857ca038cba..35e7684cf8d9d5eea4b88ada24ff252ba225bd62 100644
--- a/src/main/java/org/bukkit/entity/Item.java
+++ b/src/main/java/org/bukkit/entity/Item.java
@@ -1,6 +1,7 @@
 package org.bukkit.entity;
 
 import java.util.UUID;
+
 import org.bukkit.inventory.ItemStack;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
@@ -187,7 +188,9 @@ public interface Item extends Entity {
      * Set whether or not this item is immune to fire
      *
      * @param immuneToFire True to make immune to fire
+     @deprecated use {@link #setImmuneToFire(Boolean)} instead
      */
+    @Deprecated
     void setImmuneToFire(boolean immuneToFire);
 
     /**
