From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RhythmicSys <mc.adhdmc@gmail.com>
Date: Sat, 14 Jan 2023 18:51:19 -0800
Subject: [PATCH] Option to disable mining slowdown while flying


diff --git a/src/main/java/org/bukkit/util/permissions/PurpurPermissions.java b/src/main/java/org/bukkit/util/permissions/PurpurPermissions.java
index baec4c87d7ea4d54934ca22fd1eb7b46dd69061b..17152324029d29d7feb6d266d872964147df72d7 100644
--- a/src/main/java/org/bukkit/util/permissions/PurpurPermissions.java
+++ b/src/main/java/org/bukkit/util/permissions/PurpurPermissions.java
@@ -81,6 +81,10 @@ public final class PurpurPermissions {
 
         DefaultPermissions.registerPermission(PREFIX + "portal.instant", "Allows the user to bypass portal wait time", PermissionDefault.FALSE, purpur);
 
+        Permission flight = DefaultPermissions.registerPermission(PREFIX + "flight", "Adjusted creative flight behavior", PermissionDefault.FALSE, purpur);
+        DefaultPermissions.registerPermission(PREFIX + "flight.mining-slowdown-bypass", "Allows the user to bypass the mining slowdown that usually happens when you fly",PermissionDefault.FALSE, purpur);
+        flight.recalculatePermissibles();
+
         purpur.recalculatePermissibles();
         return purpur;
     }
