--- a/net/minecraft/commands/arguments/selector/EntitySelector.java
+++ b/net/minecraft/commands/arguments/selector/EntitySelector.java
@@ -206,26 +_,27 @@
         this.checkPermissions(sender);
         if (this.playerName != null) {
             ServerPlayer result = sender.getServer().getPlayerList().getPlayerByName(this.playerName);
-            return result == null ? List.of() : List.of(result);
+            return result == null || !canSee(sender, result) ? List.of() : List.of(result); // Purpur - Hide hidden players from entity selector
         } else if (this.entityUUID != null) {
             ServerPlayer result = sender.getServer().getPlayerList().getPlayer(this.entityUUID);
-            return result == null ? List.of() : List.of(result);
+            return result == null || !canSee(sender, result) ? List.of() : List.of(result); // Purpur - Hide hidden players from entity selector
         } else {
             Vec3 pos = this.position.apply(sender.getPosition());
             AABB absoluteAabb = this.getAbsoluteAabb(pos);
             Predicate<Entity> predicate = this.getPredicate(pos, absoluteAabb, null);
             if (this.currentEntity) {
-                return sender.getEntity() instanceof ServerPlayer player && predicate.test(player) ? List.of(player) : List.of();
+                return sender.getEntity() instanceof ServerPlayer player && predicate.test(player) && canSee(sender, player) ? List.of(player) : List.of(); // Purpur - Hide hidden players from entity selector
             } else {
                 int limit = this.getResultLimit();
                 List<ServerPlayer> result;
                 if (this.isWorldLimited()) {
                     result = sender.getLevel().getPlayers(predicate, limit);
+                    result.removeIf(entityplayer3 -> !canSee(sender, entityplayer3)); // Purpur - Hide hidden players from entity selector
                 } else {
                     result = new ObjectArrayList<>();
 
                     for (ServerPlayer player : sender.getServer().getPlayerList().getPlayers()) {
-                        if (predicate.test(player)) {
+                        if (predicate.test(player) && canSee(sender, player)) { // Purpur - Hide hidden players from entity selector
                             result.add(player);
                             if (result.size() >= limit) {
                                 return result;
@@ -283,4 +_,10 @@
     public static Component joinNames(final List<? extends Entity> entities) {
         return ComponentUtils.formatList(entities, Entity::getDisplayName);
     }
+
+    // Purpur start - Hide hidden players from entity selector
+    private boolean canSee(CommandSourceStack sender, ServerPlayer target) {
+        return !org.purpurmc.purpur.PurpurConfig.hideHiddenPlayersFromEntitySelector || !(sender.getEntity() instanceof ServerPlayer player) || player.getBukkitEntity().canSee(target.getBukkitEntity());
+    }
+    // Purpur end - Hide hidden players from entity selector
 }
