From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ben Kerllenevich <me@notom3ga.me>
Date: Sun, 13 Jun 2021 06:52:56 -0400
Subject: [PATCH] Build System Changes

todo: merge with rebrand patch

diff --git a/build.gradle.kts b/build.gradle.kts
index f9d93a70ac754150b6e545c70b3071632be8f294..79b0248caa2bcce0d3574a06877bd7e2213b443c 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -20,8 +20,8 @@ repositories {
 }
 
 dependencies {
-    implementation(project(":Paper-API"))
-    implementation(project(":Paper-MojangAPI"))
+    implementation(project(":Purpur-API")) // Purpur
+    implementation("com.destroystokyo.paper:paper-mojangapi:1.16.5-R0.1-SNAPSHOT") // Purpur
     // Paper start
     implementation("org.jline:jline-terminal-jansi:3.12.1")
     implementation("net.minecrell:terminalconsoleappender:1.2.0")
@@ -52,6 +52,7 @@ dependencies {
 
     implementation("co.aikar:cleaner:1.0-SNAPSHOT") // Paper
     implementation("io.netty:netty-all:4.1.65.Final") // Paper
+    implementation("cat.inspiracio:rhino-js-engine:1.7.7.1") // Purpur
 
     testImplementation("io.github.classgraph:classgraph:4.8.47") // Paper - mob goal test
     testImplementation("junit:junit:4.13.1")
@@ -66,7 +67,7 @@ tasks.jar {
         attributes(mapOf(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
             "Implementation-Title" to "CraftBukkit",
-            "Implementation-Version" to "git-Paper-\"$gitHash\"",
+            "Implementation-Version" to "git-Purpur-\"$gitHash\"", // Purpur
             "Implementation-Vendor" to SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'").format(Date()), // Paper
             "Specification-Title" to "Bukkit",
             "Specification-Version" to project.version,
@@ -119,7 +120,7 @@ tasks.test {
 
 fun TaskContainer.registerRunTask(name: String, block: JavaExec.() -> Unit): TaskProvider<JavaExec> =
     register<JavaExec>(name) {
-        group = "Paper"
+        group = "paperweight" // Purpur
         workingDir = rootProject.layout.projectDirectory.dir(
             providers.gradleProperty("runWorkDir").forUseAtConfigurationTime().orElse("run")
         ).get().asFile
