From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ben Kerllenevich <me@notom3ga.me>
Date: Sat, 12 Jun 2021 12:36:01 -0400
Subject: [PATCH] Build System Changes

todo: merge with rebrand patch

diff --git a/build.gradle.kts b/build.gradle.kts
index ac607a3be84bc02dedb95cab9a5eace40e33b511..6a576c6482b937d866cd6ed75d468e3a18d00786 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -20,7 +20,7 @@ repositories {
 }
 
 dependencies {
-    implementation(project(":Paper-API"))
+    implementation(project(":Purpur-API")) // Purpur
     // Paper start
     implementation("org.jline:jline-terminal-jansi:3.12.1")
     implementation("net.minecrell:terminalconsoleappender:1.2.0")
@@ -50,6 +50,8 @@ dependencies {
     implementation("co.aikar:cleaner:1.0-SNAPSHOT") // Paper
     implementation("io.netty:netty-all:4.1.65.Final") // Paper
 
+    implementation("cat.inspiracio", "rhino-js-engine", "1.7.7.1") // Purpur
+
     testImplementation("junit:junit:4.13.1")
     testImplementation("org.hamcrest:hamcrest-library:1.3")
 }
@@ -62,7 +64,7 @@ tasks.jar {
         attributes(mapOf(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
             "Implementation-Title" to "CraftBukkit",
-            "Implementation-Version" to "git-Paper-\"$gitHash\"",
+            "Implementation-Version" to "git-Purpur-\"$gitHash\"", // Purpur
             "Implementation-Vendor" to SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'").format(Date()), // Paper
             "Specification-Title" to "Bukkit",
             "Specification-Version" to project.version,
@@ -99,7 +101,7 @@ tasks.test {
 
 fun TaskContainer.registerRunTask(name: String, block: JavaExec.() -> Unit): TaskProvider<JavaExec> =
     register<JavaExec>(name) {
-        group = "Paper"
+        group = "paperweight" // Purpur
         workingDir = rootProject.projectDir.resolve("run")
         doFirst {
             if (!workingDir.exists()) workingDir.mkdir()
@@ -115,10 +117,12 @@ tasks.registerRunTask("runShadow") {
     classpath(tasks.shadowJar.flatMap { it.archiveFile })
 }
 
+/* // Purpur Start - reobfJar not found
 tasks.registerRunTask("runReobf") {
     description = "Spin up a test server from the reobfJar output jar"
     classpath(tasks.reobfJar.flatMap { it.outputJar })
 }
+ */ // Purpur End
 
 tasks.registerRunTask("runDev") {
     description = "Spin up a non-shaded non-remapped test server"
