From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Sun, 26 May 2019 15:19:14 -0500
Subject: [PATCH] Bring back server name


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
index 1fab9b9c7d41a0d2a551096c2c15f741a887fa2d..f33309f4c1ad92960d0634f3f5b8105c284f26a2 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -20,6 +20,7 @@ public class DedicatedServerProperties extends PropertyManager<DedicatedServerPr
     public final boolean onlineMode = this.getBoolean("online-mode", true);
     public final boolean preventProxyConnections = this.getBoolean("prevent-proxy-connections", false);
     public final String serverIp = this.getString("server-ip", "");
+    public final String serverName = this.getString("server-name", "Unknown Server"); // Purpur
     public final boolean spawnAnimals = this.getBoolean("spawn-animals", true);
     public final boolean spawnNpcs = this.getBoolean("spawn-npcs", true);
     public final boolean pvp = this.getBoolean("pvp", true);
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index c827314d7aa2a33a48d0747e194245ddf6fd30a6..58fb4e51e2f707a371ffa285486c1bd837280918 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2554,4 +2554,11 @@ public final class CraftServer implements Server {
     }
 
     // Paper end
+
+    // Purpur start
+    @Override
+    public String getServerName() {
+        return getProperties().serverName;
+    }
+    // Purpur end
 }
