From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Sat, 8 Feb 2020 23:30:17 -0600
Subject: [PATCH] World#getPlacementBlockData


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 0c721a5fd..a9dfd40f1 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2620,5 +2620,10 @@ public class CraftWorld implements World {
     public net.pl3x.purpur.MoonPhase getMoonPhase() {
         return net.pl3x.purpur.MoonPhase.getPhase(getFullTime() / 24000L);
     }
+
+    public BlockData getPlacementBlockData(BlockData blockdata, Location location) {
+        IBlockData validData = net.minecraft.server.Block.getValidBlockForPosition(((CraftBlockData) blockdata).getState(), getHandle(), net.minecraft.server.MCUtil.toBlockPosition(location));
+        return CraftBlockData.fromData(validData == null ? Blocks.AIR.getBlockData() : validData);
+    }
     // Purpur end
 }
