From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Sat, 19 Oct 2019 02:35:35 -0500
Subject: [PATCH] SkeletonHorseTrapEvent#getPlayer


diff --git a/src/main/java/net/minecraft/server/PathfinderGoalHorseTrap.java b/src/main/java/net/minecraft/server/PathfinderGoalHorseTrap.java
index 6a9af67d0..c10a37fe1 100644
--- a/src/main/java/net/minecraft/server/PathfinderGoalHorseTrap.java
+++ b/src/main/java/net/minecraft/server/PathfinderGoalHorseTrap.java
@@ -2,7 +2,8 @@ package net.minecraft.server;
 
 public class PathfinderGoalHorseTrap extends PathfinderGoal {
 
-    private final EntityHorseSkeleton a;
+    private final EntityHorseSkeleton a; public EntityHorseSkeleton getHorse() { return a; } // Purpur - OBFHELPER
+    private EntityHuman player; // Purpur
 
     public PathfinderGoalHorseTrap(EntityHorseSkeleton entityhorseskeleton) {
         this.a = entityhorseskeleton;
@@ -10,13 +11,13 @@ public class PathfinderGoalHorseTrap extends PathfinderGoal {
 
     @Override
     public boolean a() {
-        return this.a.world.isPlayerNearby(this.a.locX(), this.a.locY(), this.a.locZ(), 10.0D);
+        return (player = getHorse().world.findNearbyPlayer(getHorse(), 10.0D)) != null; // Purpur
     }
 
     @Override
     public void e() {
         WorldServer worldserver = (WorldServer) this.a.world;
-        if (!new com.destroystokyo.paper.event.entity.SkeletonHorseTrapEvent((org.bukkit.entity.SkeletonHorse) this.a.getBukkitEntity()).callEvent()) return; // Paper
+        if (!new com.destroystokyo.paper.event.entity.SkeletonHorseTrapEvent((org.bukkit.entity.SkeletonHorse) this.a.getBukkitEntity(), (org.bukkit.entity.Player) player.getBukkitEntity()).callEvent()) return; // Paper // Purpur
         DifficultyDamageScaler difficultydamagescaler = worldserver.getDamageScaler(this.a.getChunkCoordinates());
 
         this.a.t(false);
