From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Sat, 6 Jul 2019 17:00:04 -0500
Subject: [PATCH] Dont send useless entity packets


diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 5cb6bb896e8e71ee1b6b1ee3fff2fdb40cacfe76..aea72b0db10eed151db18490c02f291c3cded92a 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -178,6 +178,7 @@ public class EntityTrackerEntry {
                         this.o = 0;
                         packet1 = new PacketPlayOutEntityTeleport(this.tracker);
                     }
+                    if (net.pl3x.purpur.PurpurConfig.dontSendUselessEntityPackets && isUselessPacket(packet1)) packet1 = null; // Purpur
                 }
 
                 if ((this.e || this.tracker.impulse || this.tracker instanceof EntityLiving && ((EntityLiving) this.tracker).isGliding()) && this.tickCounter > 0) {
@@ -264,6 +265,22 @@ public class EntityTrackerEntry {
 
     }
 
+    // Purpur start
+    private boolean isUselessPacket(Packet<?> possibleUselessPacket) {
+        if (possibleUselessPacket instanceof PacketPlayOutEntity) {
+            PacketPlayOutEntity packet = (PacketPlayOutEntity) possibleUselessPacket;
+            if (possibleUselessPacket instanceof PacketPlayOutEntity.PacketPlayOutRelEntityMove) {
+                return packet.getX() == 0 && packet.getY() == 0 && packet.getZ() == 0;
+            } else if (possibleUselessPacket instanceof PacketPlayOutEntity.PacketPlayOutRelEntityMoveLook) {
+                return packet.getX() == 0 && packet.getY() == 0 && packet.getZ() == 0 && packet.getYaw() == 0 && packet.getPitch() == 0;
+            } else if (possibleUselessPacket instanceof PacketPlayOutEntity.PacketPlayOutEntityLook) {
+                return packet.getYaw() == 0 && packet.getPitch() == 0;
+            }
+        }
+        return false;
+    }
+    // Purpur end
+
     public void a(EntityPlayer entityplayer) {
         this.tracker.c(entityplayer);
         entityplayer.c(this.tracker);
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutEntity.java b/src/main/java/net/minecraft/server/PacketPlayOutEntity.java
index 6d9d52e4fc2cf55b21bfcb6cb1bcc6db9fb56773..0010448e3aaafc53d0f8d60f36983cb381e7afc9 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutEntity.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutEntity.java
@@ -5,11 +5,11 @@ import java.io.IOException;
 public class PacketPlayOutEntity implements Packet<PacketListenerPlayOut> {
 
     protected int a;
-    protected short b;
-    protected short c;
-    protected short d;
-    protected byte e;
-    protected byte f;
+    protected short b; public short getX() { return b; } // Purpur - OBFHELPER
+    protected short c; public short getY() { return c; } // Purpur - OBFHELPER
+    protected short d; public short getZ() { return d; } // Purpur - OBFHELPER
+    protected byte e; public byte getYaw() { return e; } // Purpur - OBFHELPER
+    protected byte f; public byte getPitch() { return f; } // Purpur - OBFHELPER
     protected boolean g;
     protected boolean h;
     protected boolean i;
diff --git a/src/main/java/net/pl3x/purpur/PurpurConfig.java b/src/main/java/net/pl3x/purpur/PurpurConfig.java
index 3d0aebd0bf1bd570f19ee296cca9f5eb8e693af1..6a622c7166dcedc7454ed2183e946a14c712104d 100644
--- a/src/main/java/net/pl3x/purpur/PurpurConfig.java
+++ b/src/main/java/net/pl3x/purpur/PurpurConfig.java
@@ -178,6 +178,11 @@ public class PurpurConfig {
         enderChestPermissionRows = getBoolean("settings.blocks.ender_chest.use-permissions-for-rows", enderChestPermissionRows);
     }
 
+    public static boolean dontSendUselessEntityPackets = false;
+    private static void dontSendUselessEntityPackets() {
+        dontSendUselessEntityPackets = getBoolean("settings.dont-send-useless-entity-packets", dontSendUselessEntityPackets);
+    }
+
     public static boolean fixItemPositionDesync = false;
     private static void fixItemPositionDesync() {
         fixItemPositionDesync = getBoolean("settings.fix-item-position-desync", fixItemPositionDesync);
