From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Thu, 2 Jan 2020 01:23:22 -0600
Subject: [PATCH] MC-147659 - Fix non black cats spawning in swamp huts


diff --git a/src/main/java/net/minecraft/server/MobSpawnerCat.java b/src/main/java/net/minecraft/server/MobSpawnerCat.java
index 6d0ebd8afe690b3781917773419abe2856009769..c7a78700746384fb41e24e577022b3d3c9c4c9f6 100644
--- a/src/main/java/net/minecraft/server/MobSpawnerCat.java
+++ b/src/main/java/net/minecraft/server/MobSpawnerCat.java
@@ -80,8 +80,9 @@ public class MobSpawnerCat implements MobSpawner {
         if (entitycat == null) {
             return 0;
         } else {
+            entitycat.setPositionRotation(blockposition, 0.0F, 0.0F); // Purpur
             entitycat.prepare(worldserver, worldserver.getDamageScaler(blockposition), EnumMobSpawn.NATURAL, (GroupDataEntity) null, (NBTTagCompound) null);
-            entitycat.setPositionRotation(blockposition, 0.0F, 0.0F);
+            //entitycat.setPositionRotation(blockposition, 0.0F, 0.0F); // Purpur - move up - fixes non black cat types spawning inside swamp huts
             worldserver.addAllEntities(entitycat);
             return 1;
         }
