From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Fri, 21 Jun 2019 14:37:10 -0500
Subject: [PATCH] Allow soil to moisten from water directly under it


diff --git a/src/main/java/net/minecraft/world/level/block/FarmBlock.java b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
index d089887030ac7c7a79abca97134ba9291e244059..802ba244903de02424e106a6dc7d5a7ed0ef912c 100644
--- a/src/main/java/net/minecraft/world/level/block/FarmBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
@@ -158,7 +158,7 @@ public class FarmBlock extends Block {
             }
         }
 
-        return false;
+        return ((ServerLevel) world).purpurConfig().blocks.farmland.moistenFromBelow && world.getFluidState(pos.relative(Direction.DOWN)).is(FluidTags.WATER); // Purpur;
     }
 
     @Override
diff --git a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
index 7c8106bc4eacc9b01597fee19b6b12b94bffab49..f83ea1f98d0bd1c9788ee47b873567dfee53a238 100644
--- a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
+++ b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
@@ -1307,6 +1307,11 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
         public class Sign extends ConfigurationPart {
             public boolean rightClickEdit = false;
         }
+
+        public Farmland farmland;
+        public class Farmland extends ConfigurationPart {
+            public boolean moistenFromBelow = false;
+        }
     }
 
     public ArmorStand armorstand;
