From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Fri, 3 May 2019 23:53:16 -0500
Subject: [PATCH] Fix cow rotation when shearing mooshroom


diff --git a/src/main/java/net/minecraft/server/EntityMushroomCow.java b/src/main/java/net/minecraft/server/EntityMushroomCow.java
index 1aacbae04..38df17bd2 100644
--- a/src/main/java/net/minecraft/server/EntityMushroomCow.java
+++ b/src/main/java/net/minecraft/server/EntityMushroomCow.java
@@ -135,7 +135,13 @@ public class EntityMushroomCow extends EntityCow implements IShearable {
 
             entitycow.setPositionRotation(this.locX(), this.locY(), this.locZ(), this.yaw, this.pitch);
             entitycow.setHealth(this.getHealth());
-            entitycow.aA = this.aA;
+            // Purpur start - correctly copy rotation
+            entitycow.copyPositionRotation(this);
+            entitycow.setRenderYawOffset(this.getRenderYawOffset());
+            entitycow.setHeadRotation(this.getHeadRotation());
+            entitycow.lastYaw = this.lastYaw;
+            entitycow.lastPitch = this.lastPitch;
+            // Purpur end
             if (this.hasCustomName()) {
                 entitycow.setCustomName(this.getCustomName());
                 entitycow.setCustomNameVisible(this.getCustomNameVisible());
