From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Thu, 19 Mar 2020 19:39:34 -0500
Subject: [PATCH] Add option to allow loyalty on tridents to work in the void


diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
index 1b31b32d42eeb54680b902cd7e82d10ba7daa5d0..9bdc46f43efe846e1525008fb891a611c3cc75c5 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
@@ -60,7 +60,7 @@ public class ThrownTrident extends AbstractArrow {
         Entity entity = this.getOwner();
         byte b0 = (Byte) this.entityData.get(ThrownTrident.ID_LOYALTY);
 
-        if (b0 > 0 && (this.dealtDamage || this.isNoPhysics()) && entity != null) {
+        if (b0 > 0 && (this.dealtDamage || this.isNoPhysics() || (level.purpurConfig().gameplayMechanics.tridentLoyaltyVoidReturnHeight < 0.0D && getY() < level.purpurConfig().gameplayMechanics.tridentLoyaltyVoidReturnHeight)) && entity != null) { // Purpur
             if (!this.isAcceptibleReturnOwner()) {
                 if (!this.level.isClientSide && this.pickup == AbstractArrow.Pickup.ALLOWED) {
                     this.spawnAtLocation(this.getPickupItem(), 0.1F);
diff --git a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
index bbbdcd1ec5bf6ead16445ef0eb6bce8b9281da35..b3fb59fc7bc801fcfb715793a0c68cc5d9393b91 100644
--- a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
+++ b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
@@ -1269,6 +1269,7 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
     public class GameplayMechanics extends ConfigurationPart {
         public boolean disableCrammingDeathDrops = false;
         public boolean milkCuresBadOmen = true;
+        public double tridentLoyaltyVoidReturnHeight = 0.0D;
 
         public Player player;
         public class Player extends ConfigurationPart {
