From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Sat, 4 Apr 2020 17:16:30 -0500
Subject: [PATCH] Debug stick should not update neighbors


diff --git a/src/main/java/net/minecraft/server/ItemStack.java b/src/main/java/net/minecraft/server/ItemStack.java
index 3adb29f004d2fee36f3ee9b21ee5417e84b64837..e7759f24819ea7c62d3ae52ed1cc5c00f994a836 100644
--- a/src/main/java/net/minecraft/server/ItemStack.java
+++ b/src/main/java/net/minecraft/server/ItemStack.java
@@ -229,6 +229,11 @@ public final class ItemStack {
                 if (this.getItem() == Items.BONE_MEAL) {
                     world.captureTreeGeneration = true;
                 }
+                // Purpur start
+                else if (this.getItem() == Items.DEBUG_STICK) {
+                    world.captureBlockStates = false;
+                }
+                // Purpur end
             }
             Item item = this.getItem();
             EnumInteractionResult enuminteractionresult = item.a(itemactioncontext);
