From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Thu, 2 Jan 2020 11:31:36 -0600
Subject: [PATCH] Make lava flow speed configurable


diff --git a/src/main/java/net/minecraft/world/level/material/LavaFluid.java b/src/main/java/net/minecraft/world/level/material/LavaFluid.java
index 9ea5d189ecf2dbbf4619f11fe621ba0dc4d80c38..0fc587406c4a5368ca41135da62869be62ee708a 100644
--- a/src/main/java/net/minecraft/world/level/material/LavaFluid.java
+++ b/src/main/java/net/minecraft/world/level/material/LavaFluid.java
@@ -180,7 +180,7 @@ public abstract class LavaFluid extends FlowingFluid {
 
     @Override
     public int getTickDelay(LevelReader world) {
-        return world.dimensionType().ultraWarm() ? 10 : 30;
+        return world.dimensionType().ultraWarm() ? world.getWorldBorder().world.purpurConfig().blocks.lava.speed.nether : world.getWorldBorder().world.purpurConfig().blocks.lava.speed.notNether; // Purpur
     }
 
     @Override
diff --git a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
index d49162d3614fd168b1442575ac4bdd84ce67cced..7c214b5872c2fd3101fa6c90d6f7f7c807afd9b1 100644
--- a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
+++ b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
@@ -1354,6 +1354,12 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
         public class Lava extends ConfigurationPart {
             public boolean allowInfinite = false;
             public int infiniteSourcesRequired = 2;
+
+            public Speed speed;
+            public class Speed extends ConfigurationPart {
+                public int nether = 10;
+                public int notNether = 30;
+            }
         }
     }
 
