From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Thu, 26 Dec 2019 22:08:37 -0600
Subject: [PATCH] Add player death exp control options


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 74f14215280005a4d6fb35a77e2f1c4858250f2f..1393c7a3b66abbc0be831705ac2d623c89413797 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1978,9 +1978,18 @@ public abstract class Player extends LivingEntity {
     @Override
     public int getExperienceReward() {
         if (!this.level.getGameRules().getBoolean(GameRules.RULE_KEEPINVENTORY) && !this.isSpectator()) {
-            int i = this.experienceLevel * 7;
-
-            return i > 100 ? 100 : i;
+            // Purpur start
+            int toDrop;
+            try {
+                scriptEngine.eval("expLevel = " + experienceLevel);
+                scriptEngine.eval("expTotal = " + totalExperience);
+                scriptEngine.eval("exp = " + experienceProgress);
+                toDrop = (int) Math.round((Double) scriptEngine.eval(level.purpurConfig().gameplayMechanics.player.death.expDropEquation));
+            } catch (Exception ignore) {
+                toDrop = experienceLevel * 7;
+            }
+            return Math.min(toDrop, level.purpurConfig().gameplayMechanics.player.death.maxExpDropped);
+            // Purpur end
         } else {
             return 0;
         }
diff --git a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
index 40ade185f61d6b3af1113d15bdb230882d129845..61484b533a67801fc0c3a020b49c23d2ec3c1527 100644
--- a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
+++ b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
@@ -1296,6 +1296,12 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
                 public boolean whileAcceptingResourcePack = false;
                 public int afterSpawnTicks = 60;
             }
+
+            public Death death;
+            public class Death extends ConfigurationPart {
+                public String expDropEquation = "expLevel * 7";
+                public int maxExpDropped = 100;
+            }
         }
 
         public SilkTouchSpawner silkTouchSpawner;
