From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Sat, 23 Nov 2019 17:55:42 -0600
Subject: [PATCH] Implement infinite lava


diff --git a/src/main/java/net/minecraft/server/FluidTypeFlowing.java b/src/main/java/net/minecraft/server/FluidTypeFlowing.java
index d72a88e9275eb00eed35b6467538a46e5cee32d5..35d55bc15b179a8200082ca23fad04bbe550068d 100644
--- a/src/main/java/net/minecraft/server/FluidTypeFlowing.java
+++ b/src/main/java/net/minecraft/server/FluidTypeFlowing.java
@@ -195,7 +195,7 @@ public abstract class FluidTypeFlowing extends FluidType {
             }
         }
 
-        if (this.f() && j >= 2) {
+        if (infinite(iworldreader) && j >= getRequiredSources(iworldreader)) { // Purpur
             IBlockData iblockdata2 = iworldreader.getType(blockposition.down());
             Fluid fluid1 = iblockdata2.getFluid();
 
@@ -266,6 +266,17 @@ public abstract class FluidTypeFlowing extends FluidType {
         return (Fluid) this.e().h().set(FluidTypeFlowing.FALLING, flag);
     }
 
+    // Purpur start
+    protected boolean infinite(IWorldReader iworldreader) {
+        return infinite();
+    }
+
+    protected int getRequiredSources(IWorldReader iworldreader) {
+        return 2;
+    }
+    // Purpur end
+
+    protected boolean infinite() { return f(); } // Purpur - OBFHELPER
     protected abstract boolean f();
 
     protected void a(GeneratorAccess generatoraccess, BlockPosition blockposition, IBlockData iblockdata, EnumDirection enumdirection, Fluid fluid) {
diff --git a/src/main/java/net/minecraft/server/FluidTypeLava.java b/src/main/java/net/minecraft/server/FluidTypeLava.java
index 29930e801cdcb97bec2fb113ec478fe9c4a63b63..ffab2391925f577420ee52f3aa05041afa61464e 100644
--- a/src/main/java/net/minecraft/server/FluidTypeLava.java
+++ b/src/main/java/net/minecraft/server/FluidTypeLava.java
@@ -147,6 +147,18 @@ public abstract class FluidTypeLava extends FluidTypeFlowing {
         generatoraccess.triggerEffect(1501, blockposition, 0);
     }
 
+    // Purpur start
+    @Override
+    protected boolean infinite(IWorldReader iworldreader) {
+        return iworldreader.getWorldBorder().world.purpurConfig.lavaInfinite;
+    }
+
+    @Override
+    protected int getRequiredSources(IWorldReader iworldreader) {
+        return iworldreader.getWorldBorder().world.purpurConfig.lavaInfiniteRequiredSources;
+    }
+    // Purpur end
+
     @Override
     protected boolean f() {
         return false;
diff --git a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
index db6fb004be99341dba16c184244ed1a04c386fd4..d1f86ed2bbdec50e17b04814ea103bc0fb9ac0b1 100644
--- a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
+++ b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
@@ -158,6 +158,13 @@ public class PurpurWorldConfig {
         farmlandGetsMoistFromBelow = getBoolean("blocks.farmland.gets-moist-from-below", farmlandGetsMoistFromBelow);
     }
 
+    public boolean lavaInfinite = false;
+    public int lavaInfiniteRequiredSources = 2;
+    private void lavaSettings() {
+        lavaInfinite = getBoolean("blocks.lava.infinite-source", lavaInfinite);
+        lavaInfiniteRequiredSources = getInt("blocks.lava.infinite-required-sources", lavaInfiniteRequiredSources);
+    }
+
     public boolean signAllowColors = false;
     public boolean signRightClickEdit = false;
     private void signSettings() {
