From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Wed, 1 Jan 2020 20:12:39 -0600
Subject: [PATCH] Add obfhelpers for plugin use


diff --git a/src/main/java/net/minecraft/server/ItemStack.java b/src/main/java/net/minecraft/server/ItemStack.java
index e7759f248..2c908a604 100644
--- a/src/main/java/net/minecraft/server/ItemStack.java
+++ b/src/main/java/net/minecraft/server/ItemStack.java
@@ -644,6 +644,7 @@ public final class ItemStack {
         return this.tag;
     }
 
+    public NBTTagCompound getOrCreateSubTag(String s) { return a(s); } // Purpur - OBFHELPER
     public NBTTagCompound a(String s) {
         if (this.tag != null && this.tag.hasKeyOfType(s, 10)) {
             return this.tag.getCompound(s);
