From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: draycia <lonelyyordle@gmail.com>
Date: Tue, 14 Apr 2020 00:35:12 -0700
Subject: [PATCH] Redstone deactivates spawners


diff --git a/src/main/java/net/minecraft/world/level/BaseSpawner.java b/src/main/java/net/minecraft/world/level/BaseSpawner.java
index ee619590aa49323059947fbaee9e88d61df99789..2ccc06732e7b98fa71dacf9280ae83df2f0bce3c 100644
--- a/src/main/java/net/minecraft/world/level/BaseSpawner.java
+++ b/src/main/java/net/minecraft/world/level/BaseSpawner.java
@@ -52,6 +52,7 @@ public abstract class BaseSpawner {
     }
 
     public boolean isNearPlayer(Level world, BlockPos pos) {
+        if (world.purpurConfig().blocks.spawner.deactivateWithRedstone && world.hasNeighborSignal(pos)) return false; // Purpur
         return world.isAffectsSpawningPlayerNearby((double) pos.getX() + 0.5D, (double) pos.getY() + 0.5D, (double) pos.getZ() + 0.5D, (double) this.requiredPlayerRange); // Paper
     }
 
diff --git a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
index 45a948f21e5cd4006cddf765df8ece2710cc42e7..c515614f6633d7bd1f35ea39127cd93dc1ac9f5c 100644
--- a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
+++ b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
@@ -1483,6 +1483,11 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
             public boolean explosionFire = true;
             public Explosion.BlockInteraction explosionEffect = Explosion.BlockInteraction.DESTROY;
         }
+
+        public Spawner spawner;
+        public class Spawner extends ConfigurationPart {
+            public boolean deactivateWithRedstone = false;
+        }
     }
 
     public ArmorStand armorstand;
