From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Fri, 24 Jul 2020 19:38:21 -0500
Subject: [PATCH] Add vindicator johnny spawn chance


diff --git a/src/main/java/net/minecraft/world/entity/monster/Vindicator.java b/src/main/java/net/minecraft/world/entity/monster/Vindicator.java
index 4e2b4d1720513e0ad4496f614d20acbd561c60e5..631f2e8e5758332ebd56629d636ae55e465c4cc6 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Vindicator.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Vindicator.java
@@ -154,6 +154,12 @@ public class Vindicator extends AbstractIllager {
         RandomSource randomSource = world.getRandom();
         this.populateDefaultEquipmentSlots(randomSource, difficulty);
         this.populateDefaultEquipmentEnchantments(randomSource, difficulty);
+        // Purpur start
+        Level level = world.getMinecraftWorld();
+        if (level.purpurConfig().mobs.vindicator.johnnySpawnChance > 0D && random.nextDouble() <= level.purpurConfig().mobs.vindicator.johnnySpawnChance) {
+            setCustomName(Component.translatable("Johnny"));
+        }
+        // Purpur end
         return spawnGroupData;
     }
 
diff --git a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
index 9d7c87cf39eb8cabea2fcca8fbf325c507648f9a..6276b26d18b368cdba498d411f5f99b267fdd731 100644
--- a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
+++ b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
@@ -1143,6 +1143,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Vindicator vindicator;
         public class Vindicator extends ConfigurationPart {
+            public double johnnySpawnChance = 0.0D;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
