From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Mon, 10 Aug 2020 21:46:22 -0500
Subject: [PATCH] Short enderman height


diff --git a/src/main/java/net/minecraft/world/entity/EntityType.java b/src/main/java/net/minecraft/world/entity/EntityType.java
index 615394b8a0dada80b77c1c5d515f79edaca53c2e..12ff6a31d38db585f36e492ff07dc00e5da8cd6e 100644
--- a/src/main/java/net/minecraft/world/entity/EntityType.java
+++ b/src/main/java/net/minecraft/world/entity/EntityType.java
@@ -291,7 +291,8 @@ public class EntityType<T extends Entity> implements EntityTypeTest<Entity, T> {
     private Component description;
     @Nullable
     private ResourceLocation lootTable;
-    private final EntityDimensions dimensions;
+    private EntityDimensions dimensions; // Purpur - remove final
+    public void setDimensions(EntityDimensions dimensions) { this.dimensions = dimensions; } // Purpur
 
     private static <T extends Entity> EntityType<T> register(String id, EntityType.Builder type) { // CraftBukkit - decompile error
         return (EntityType) Registry.register(Registry.ENTITY_TYPE, id, (EntityType<T>) type.build(id)); // CraftBukkit - decompile error
diff --git a/src/main/java/net/minecraft/world/entity/monster/EnderMan.java b/src/main/java/net/minecraft/world/entity/monster/EnderMan.java
index 7fc2a82e46d0974c829388d8caa9cc7a9d5a7cb8..ffeb69ed581078215ffeb0a9efbac86394be8dfd 100644
--- a/src/main/java/net/minecraft/world/entity/monster/EnderMan.java
+++ b/src/main/java/net/minecraft/world/entity/monster/EnderMan.java
@@ -407,6 +407,7 @@ public class EnderMan extends Monster implements NeutralMob {
         if (this.isInvulnerableTo(source)) {
             return false;
         } else if (getRider() != null && this.isControllable()) { return super.hurt(source, amount); // Purpur - no teleporting on damage
+        } else if (org.purpurmc.purpur.configuration.PurpurConfiguration.get().entity.enderman.shortHeight && source == DamageSource.IN_WALL) { return false; // Purpur - no suffocation damage if short height
         } else if (source instanceof IndirectEntityDamageSource) {
             Entity entity = source.getDirectEntity();
             boolean flag;
diff --git a/src/main/java/org/purpurmc/purpur/configuration/PurpurConfiguration.java b/src/main/java/org/purpurmc/purpur/configuration/PurpurConfiguration.java
index dd7115df35e4bd9a5c45a263de7e9838241264a0..9bd22087bed022a392d3f0f3685c51246038ad41 100644
--- a/src/main/java/org/purpurmc/purpur/configuration/PurpurConfiguration.java
+++ b/src/main/java/org/purpurmc/purpur/configuration/PurpurConfiguration.java
@@ -3,6 +3,8 @@ package org.purpurmc.purpur.configuration;
 import com.mojang.logging.LogUtils;
 import io.papermc.paper.configuration.Configuration;
 import io.papermc.paper.configuration.ConfigurationPart;
+import net.minecraft.world.entity.EntityDimensions;
+import net.minecraft.world.entity.EntityType;
 import org.bukkit.event.inventory.InventoryType;
 import org.slf4j.Logger;
 import org.spongepowered.configurate.objectmapping.meta.Setting;
@@ -77,6 +79,21 @@ public class PurpurConfiguration extends ConfigurationPart {
         }
     }
 
+    public Entity entity;
+    public class Entity extends ConfigurationPart {
+        public Enderman enderman;
+        public class Enderman extends ConfigurationPart.Post {
+            public boolean shortHeight = false;
+
+            @Override
+            public void postProcess() {
+                if (shortHeight) {
+                    EntityType.ENDERMAN.setDimensions(EntityDimensions.scalable(0.6F, 1.9F));
+                }
+            }
+        }
+    }
+
     public TPS tps;
     public class TPS extends ConfigurationPart {
         public double laggingThreshold = 19.0D;
