From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Sat, 15 Aug 2020 06:51:46 -0500
Subject: [PATCH] Use configured height for nether surface builders


diff --git a/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherCappedSurfaceBuilder.java b/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherCappedSurfaceBuilder.java
index 7c3386b37aae7d10629e0a256102967f69b68a7e..0cfb34557eb52988681868ad772ca3b113ef5dc2 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherCappedSurfaceBuilder.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherCappedSurfaceBuilder.java
@@ -40,7 +40,7 @@ public abstract class NetherCappedSurfaceBuilder extends SurfaceBuilder<SurfaceB
             return entry.getValue().getValue((double)x, (double)seaLevel, (double)z);
         })).get().getKey();
         BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
-        BlockState blockState3 = chunk.getBlockState(mutableBlockPos.set(k, 128, m));
+        BlockState blockState3 = chunk.getBlockState(mutableBlockPos.set(k, height, m)); // Purpur - use configured height
 
         for(int p = height; p >= i; --p) { // Paper - fix MC-187716 - use configured height
             mutableBlockPos.set(k, p, m);
