From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Sat, 4 Jul 2020 11:23:46 -0500
Subject: [PATCH] PaperPR - Fix loottables erasing on viewing


diff --git a/src/main/java/net/minecraft/server/EntityMinecartContainer.java b/src/main/java/net/minecraft/server/EntityMinecartContainer.java
index fca0fb871e..72411faa14 100644
--- a/src/main/java/net/minecraft/server/EntityMinecartContainer.java
+++ b/src/main/java/net/minecraft/server/EntityMinecartContainer.java
@@ -223,7 +223,7 @@ public abstract class EntityMinecartContainer extends EntityMinecartAbstract imp
                 CriterionTriggers.N.a((EntityPlayer) entityhuman, this.lootTable);
             }
 
-            this.lootTable = null;
+            //this.lootTable = null; // Purpur
             this.lootableData.processRefill(entityhuman); // Paper
             LootTableInfo.Builder loottableinfo_builder = (new LootTableInfo.Builder((WorldServer) this.world)).set(LootContextParameters.POSITION, this.getChunkCoordinates()).a(this.lootTableSeed);
 
diff --git a/src/main/java/net/minecraft/server/TileEntityLootable.java b/src/main/java/net/minecraft/server/TileEntityLootable.java
index d017904561..7fd2b0e838 100644
--- a/src/main/java/net/minecraft/server/TileEntityLootable.java
+++ b/src/main/java/net/minecraft/server/TileEntityLootable.java
@@ -56,7 +56,7 @@ public abstract class TileEntityLootable extends TileEntityContainer {
                 CriterionTriggers.N.a((EntityPlayer) entityhuman, this.lootTable);
             }
 
-            this.lootTable = null;
+            //this.lootTable = null; // Purpur
             this.lootableData.processRefill(entityhuman); // Paper
             LootTableInfo.Builder loottableinfo_builder = (new LootTableInfo.Builder((WorldServer) this.world)).set(LootContextParameters.POSITION, new BlockPosition(this.position)).a(this.lootTableSeed);
 
