From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Sat, 4 Jul 2020 19:19:13 -0500
Subject: [PATCH] Fix armorstand entity lookups paper.yml setting not being
 used


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index d4abe2d9c..ac54b1df1 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -974,13 +974,6 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
         }
         MinecraftServer.getServer().executeMidTickTasks(); // Tuinity - execute chunk tasks mid tick
     }
-    // Paper start - Prevent armor stands from doing entity lookups
-    @Override
-    public boolean getCubes(@Nullable Entity entity, AxisAlignedBB axisAlignedBB) {
-        if (entity instanceof EntityArmorStand && !entity.world.paperConfig.armorStandEntityLookups) return false;
-        return GeneratorAccess.super.getCubes(entity, axisAlignedBB);
-    }
-    // Paper end
 
     public Explosion explode(@Nullable Entity entity, double d0, double d1, double d2, float f, Explosion.Effect explosion_effect) {
         return this.createExplosion(entity, (DamageSource) null, (ExplosionDamageCalculator) null, d0, d1, d2, f, false, explosion_effect);
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index e6ad5a6ab..7c9a3714e 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -688,6 +688,7 @@ public class WorldServer extends World implements GeneratorAccessSeed {
 
     @Override
     public boolean getCubes(@Nullable Entity entity, AxisAlignedBB axisalignedbb) {
+        if (entity instanceof EntityArmorStand && !entity.world.paperConfig.armorStandEntityLookups) return false; // Paper // Purpur - moved from World
         return !this.hasAnyCollisions(entity, axisalignedbb);
     }
 
