From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Sun, 4 Oct 2020 12:00:42 -0500
Subject: [PATCH] Flying squids! Oh my!


diff --git a/src/main/java/net/minecraft/world/entity/GlowSquid.java b/src/main/java/net/minecraft/world/entity/GlowSquid.java
index 4d802c909a38ba764334890b1212b16dd60b7d93..a780758d31083cce2d3dc052144a4029b445a407 100644
--- a/src/main/java/net/minecraft/world/entity/GlowSquid.java
+++ b/src/main/java/net/minecraft/world/entity/GlowSquid.java
@@ -43,6 +43,11 @@ public class GlowSquid extends Squid {
     public void initAttributes() {
         this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.glowSquid.attributes.maxHealth);
     }
+
+    @Override
+    public boolean canFly() {
+        return this.level.purpurConfig().mobs.glowSquid.canFly;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/Squid.java b/src/main/java/net/minecraft/world/entity/animal/Squid.java
index c7b7436a75c58e01d3f24c08cf4c153afb2827be..0b8062f1b22a864593ee88236c7a63f909f58572 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Squid.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Squid.java
@@ -86,6 +86,15 @@ public class Squid extends WaterAnimal {
         // Stops squids from floating just over the water
         return super.getAxisForFluidCheck().offsetY(level.purpurConfig().mobs.squid.offsetWaterCheck);
     }
+
+    public boolean canFly() {
+        return this.level.purpurConfig().mobs.squid.canFly;
+    }
+
+    @Override
+    public boolean isInWater() {
+        return this.wasTouchingWater || canFly();
+    }
     // Purpur end
 
     @Override
@@ -160,6 +169,7 @@ public class Squid extends WaterAnimal {
         }
 
         if (this.isInWaterOrBubble()) {
+            if (canFly()) setNoGravity(!wasTouchingWater); // Purpur
             if (this.tentacleMovement < 3.1415927F) {
                 float f = this.tentacleMovement / 3.1415927F;
 
@@ -319,7 +329,7 @@ public class Squid extends WaterAnimal {
 
             if (i > 100) {
                 this.squid.setMovementVector(0.0F, 0.0F, 0.0F);
-            } else if (this.squid.getRandom().nextInt(reducedTickDelay(50)) == 0 || !this.squid.wasTouchingWater || !this.squid.hasMovementVector()) {
+            } else if (this.squid.getRandom().nextInt(reducedTickDelay(50)) == 0 || !this.squid.isInWater() || !this.squid.hasMovementVector()) { // Purpur
                 float f = this.squid.getRandom().nextFloat() * 6.2831855F;
                 float f1 = Mth.cos(f) * 0.2F;
                 float f2 = -0.1F + this.squid.getRandom().nextFloat() * 0.2F;
diff --git a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
index 0550c63a2da8b62632e3fe6d6e0291b05cdded1a..606fbfb0dfa1df17769f6d42ac35a72810a73fc3 100644
--- a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
+++ b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
@@ -408,6 +408,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public GlowSquid glowSquid;
         public class GlowSquid extends ConfigurationPart {
+            public boolean canFly = false;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
@@ -999,6 +1001,7 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
         public class Squid extends ConfigurationPart {
             public boolean immuneToEAR = true;
             public double offsetWaterCheck = 0.0D;
+            public boolean canFly = false;
 
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
