From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Fri, 10 Jul 2020 09:40:30 -0500
Subject: [PATCH] ALlow color codes in books


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index c518d4346..194256e51 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -940,12 +940,13 @@ public class PlayerConnection implements PacketListenerPlayIn {
                             itemstack2.setTag(nbttagcompound.clone());
                         }
 
+                        boolean hasPerm = getPlayer().hasPermission("purpur.book.color.edit") || getPlayer().hasPermission("purpur.book.color.sign");
                         itemstack2.a("author", (NBTBase) NBTTagString.a(this.player.getDisplayName().getString()));
-                        itemstack2.a("title", (NBTBase) NBTTagString.a(itemstack.getTag().getString("title")));
+                        itemstack2.a("title", (NBTBase) NBTTagString.a(color(itemstack.getTag().getString("title"), hasPerm))); // Purpur - sign book
                         NBTTagList nbttaglist = itemstack.getTag().getList("pages", 8);
 
                         for (int i = 0; i < nbttaglist.size(); ++i) {
-                            String s = nbttaglist.getString(i);
+                            String s = color(nbttaglist.getString(i), hasPerm);// Purpur - sign book
                             ChatComponentText chatcomponenttext = new ChatComponentText(s);
 
                             s = IChatBaseComponent.ChatSerializer.a((IChatBaseComponent) chatcomponenttext);
@@ -957,7 +958,14 @@ public class PlayerConnection implements PacketListenerPlayIn {
                     } else {
                         // Paper start - dont mutate players current item, set it from the event
                         ItemStack newBook = itemstack1.cloneItemStack();
-                        newBook.getOrCreateTagAndSet("pages", (NBTBase)itemstack.getTag().getList("pages", 8));
+                        // Purpur start - edit book
+                        boolean hasPerm = getPlayer().hasPermission("purpur.book.color.edit");
+                        NBTTagList nbttaglist = itemstack.getTag().getList("pages", 8);
+                        for (int i = 0; i < nbttaglist.size(); ++i) {
+                            nbttaglist.set(i, NBTTagString.create(color(nbttaglist.getString(i), hasPerm, false)));
+                        }
+                        newBook.getOrCreateTagAndSet("pages", nbttaglist);
+                        // Purpur end - edit book
                         this.player.setSlot(enumitemslot, CraftEventFactory.handleEditBookEvent(player, enumitemslot, itemstack1, newBook));
                         // Paper end
                     }
@@ -967,6 +975,16 @@ public class PlayerConnection implements PacketListenerPlayIn {
         }
     }
 
+    // Purpur start
+    private String color(String str, boolean hasPerm) {
+        return color(str, hasPerm, true);
+    }
+
+    private String color(String str, boolean hasPerm, boolean parseHex) {
+        return hasPerm ? org.bukkit.ChatColor.color(str, parseHex) : str;
+    }
+    // Purpur end
+
     @Override
     public void a(PacketPlayInEntityNBTQuery packetplayinentitynbtquery) {
         PlayerConnectionUtils.ensureMainThread(packetplayinentitynbtquery, this, this.player.getWorldServer());
