From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Tue, 13 Oct 2020 20:04:33 -0500
Subject: [PATCH] Allow infinite and mending enchantments together


diff --git a/src/main/java/net/minecraft/world/item/enchantment/ArrowInfiniteEnchantment.java b/src/main/java/net/minecraft/world/item/enchantment/ArrowInfiniteEnchantment.java
index 3aece8245060dd1ba269c08d226c84247a6f0a83..e4a375dfd22a57ee20cc4639038f81e625999b93 100644
--- a/src/main/java/net/minecraft/world/item/enchantment/ArrowInfiniteEnchantment.java
+++ b/src/main/java/net/minecraft/world/item/enchantment/ArrowInfiniteEnchantment.java
@@ -1,6 +1,7 @@
 package net.minecraft.world.item.enchantment;
 
 import net.minecraft.world.entity.EquipmentSlot;
+import org.purpurmc.purpur.configuration.PurpurConfiguration;
 
 public class ArrowInfiniteEnchantment extends Enchantment {
     public ArrowInfiniteEnchantment(Enchantment.Rarity weight, EquipmentSlot... slotTypes) {
@@ -24,6 +25,6 @@ public class ArrowInfiniteEnchantment extends Enchantment {
 
     @Override
     public boolean checkCompatibility(Enchantment other) {
-        return other instanceof MendingEnchantment ? false : super.checkCompatibility(other);
+        return other instanceof MendingEnchantment ? PurpurConfiguration.get().enchantments.allowInfinityAndMending : super.checkCompatibility(other);
     }
 }
diff --git a/src/main/java/org/purpurmc/purpur/configuration/PurpurConfiguration.java b/src/main/java/org/purpurmc/purpur/configuration/PurpurConfiguration.java
index 89898663375c818f1f2c405d213b1d34cd627fb9..4bca3748aa23d30afca0cd1dc767fcfd81be5573 100644
--- a/src/main/java/org/purpurmc/purpur/configuration/PurpurConfiguration.java
+++ b/src/main/java/org/purpurmc/purpur/configuration/PurpurConfiguration.java
@@ -99,6 +99,11 @@ public class PurpurConfiguration extends ConfigurationPart {
         }
     }
 
+    public Enchantments enchantments;
+    public class Enchantments extends ConfigurationPart {
+        public boolean allowInfinityAndMending = false;
+    }
+
     public TPS tps;
     public class TPS extends ConfigurationPart {
         public double laggingThreshold = 19.0D;
