From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Fri, 10 Jul 2020 13:12:33 -0500
Subject: [PATCH] Fix SPIGOT-5824 Bukkit world-container is not used


diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 21d4c965d..220a084be 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -95,7 +95,20 @@ public class Main {
             }
             // Purpur end - fix SPIGOT-5885
 
-            File file = (File) optionset.valueOf("universe"); // CraftBukkit
+            // Purpur start - fix SPIGOT-5824
+            File file;
+            if (optionset.has("universe")) {
+                file = (File) optionset.valueOf("universe"); // CraftBukkit
+            } else {
+                try {
+                    file = new File(org.bukkit.configuration.file.YamlConfiguration
+                            .loadConfiguration((File) optionset.valueOf("bukkit-settings"))
+                            .getString("settings.world-container", "."));
+                } catch (Exception ignore) {
+                    file = new File(".");
+                }
+            }
+            // Purpur end - fix SPIGOT-5824
             YggdrasilAuthenticationService yggdrasilauthenticationservice = new com.destroystokyo.paper.profile.PaperAuthenticationService(Proxy.NO_PROXY, UUID.randomUUID().toString()); // Paper
             MinecraftSessionService minecraftsessionservice = yggdrasilauthenticationservice.createMinecraftSessionService();
             GameProfileRepository gameprofilerepository = yggdrasilauthenticationservice.createProfileRepository();
