From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date: Sun, 15 Nov 2020 02:18:15 -0800
Subject: [PATCH] Make entity breeding times configurable


diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java b/src/main/java/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
index b3adff55870c673d753f05dee4540bef0d142c7b..312e57772eb5138bcf41cacb72e7760ba452fcd3 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
@@ -123,8 +123,10 @@ public class VillagerMakeLove extends Behavior<Villager> {
                 return Optional.empty();
             }
             // CraftBukkit end
-            parent.setAge(6000);
-            partner.setAge(6000);
+            // Purpur start
+            parent.setAge(world.purpurConfig().mobs.villager.breedingDelay);
+            partner.setAge(world.purpurConfig().mobs.villager.breedingDelay);
+            // Purpur end
             world.addFreshEntityWithPassengers(entityvillager2, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING); // CraftBukkit - added SpawnReason
             world.broadcastEntityEvent(entityvillager2, (byte) 12);
             return Optional.of(entityvillager2);
diff --git a/src/main/java/net/minecraft/world/entity/animal/Animal.java b/src/main/java/net/minecraft/world/entity/animal/Animal.java
index 7ebbd2b9519ff67f93cf24d69c020f1b0fbd364e..99c11af5247bffff812e900a6a1e29254db0214b 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Animal.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Animal.java
@@ -39,6 +39,7 @@ public abstract class Animal extends AgeableMob {
     @Nullable
     public UUID loveCause;
     public ItemStack breedItem; // CraftBukkit - Add breedItem variable
+    public abstract int getPurpurBreedTime(); // Purpur
 
     protected Animal(EntityType<? extends Animal> type, Level world) {
         super(type, world);
@@ -261,8 +262,10 @@ public abstract class Animal extends AgeableMob {
                 CriteriaTriggers.BRED_ANIMALS.trigger(entityplayer, this, other, entityageable);
             }
 
-            this.setAge(6000);
-            other.setAge(6000);
+            // Purpur start
+            this.setAge(this.getPurpurBreedTime());
+            other.setAge(other.getPurpurBreedTime());
+            // Purpur end
             this.resetLove();
             other.resetLove();
             world.addFreshEntityWithPassengers(entityageable, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING); // CraftBukkit - added SpawnReason
diff --git a/src/main/java/net/minecraft/world/entity/animal/Bee.java b/src/main/java/net/minecraft/world/entity/animal/Bee.java
index 6654f883e578561060e2841052d9bbaf4b62f394..59dd1c8bbda413b6425b3b9d1fd54bc4f3d836b5 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Bee.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Bee.java
@@ -219,6 +219,11 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.bee.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.bee.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/Cat.java b/src/main/java/net/minecraft/world/entity/animal/Cat.java
index 188a65bb749c236f97ff097f4348c07245453491..a6a61f4a5b3effd34c243a7abd85b7ee4aaed8f0 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Cat.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Cat.java
@@ -123,6 +123,11 @@ public class Cat extends TamableAnimal {
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.cat.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.cat.breedingDelay;
+    }
     // Purpur end
 
     public ResourceLocation getResourceLocation() {
diff --git a/src/main/java/net/minecraft/world/entity/animal/Chicken.java b/src/main/java/net/minecraft/world/entity/animal/Chicken.java
index 3113cc7d0d808afe72c20a485979b0521e0982b2..e45d9721f1c579f0d78fb021b448f9457e630c7f 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Chicken.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Chicken.java
@@ -76,6 +76,11 @@ public class Chicken extends Animal {
             this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(2.0D);
         }
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.chicken.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/Cow.java b/src/main/java/net/minecraft/world/entity/animal/Cow.java
index c0520377bde5f11c585eb6a6257dd9d7041e0b46..35bf300e90ce31c2266130182f45d8764a58d126 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Cow.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Cow.java
@@ -62,6 +62,11 @@ public class Cow extends Animal {
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.cow.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.cow.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/Fox.java b/src/main/java/net/minecraft/world/entity/animal/Fox.java
index fc56ba4725f140b3f5b62050e42fa82057bbc450..c66e4df7164738b0545feaf3807cfd4507882e13 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Fox.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Fox.java
@@ -183,6 +183,11 @@ public class Fox extends Animal {
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.fox.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.fox.breedingDelay;
+    }
     // Purpur end
 
     @Override
@@ -969,8 +974,10 @@ public class Fox extends Animal {
                     CriteriaTriggers.BRED_ANIMALS.trigger(entityplayer2, this.animal, this.partner, entityfox);
                 }
 
-                this.animal.setAge(6000);
-                this.partner.setAge(6000);
+                // Purpur start
+                this.animal.setAge(this.animal.getPurpurBreedTime());
+                this.partner.setAge(this.partner.getPurpurBreedTime());
+                // Purpur end
                 this.animal.resetLove();
                 this.partner.resetLove();
                 worldserver.addFreshEntityWithPassengers(entityfox, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING); // CraftBukkit - added SpawnReason
diff --git a/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java b/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java
index 20c3f76fbf9d08e0d993117ab8fc412425b40d63..1878154c01d14bdd0d5d5f0b6fd648720510ca75 100644
--- a/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java
+++ b/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java
@@ -84,6 +84,11 @@ public class MushroomCow extends Cow implements Shearable {
     public void initAttributes() {
         this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.mooshroom.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.mooshroom.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/Ocelot.java b/src/main/java/net/minecraft/world/entity/animal/Ocelot.java
index ec435c37a2a353b53102e7520c0a5ee92caccc62..80b2772615d7faa9b563c4aee93965883cc54443 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Ocelot.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Ocelot.java
@@ -88,6 +88,11 @@ public class Ocelot extends Animal {
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.ocelot.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.ocelot.breedingDelay;
+    }
     // Purpur end
 
     public boolean isTrusting() {
diff --git a/src/main/java/net/minecraft/world/entity/animal/Panda.java b/src/main/java/net/minecraft/world/entity/animal/Panda.java
index c97b3da80ad51e1937e4615435db1559823ff542..e3162f649a25a73c0106bc56f85c5a86bc1f88d8 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Panda.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Panda.java
@@ -137,6 +137,11 @@ public class Panda extends Animal {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.panda.attributes.maxHealth);
         setAttributes();
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.panda.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/Parrot.java b/src/main/java/net/minecraft/world/entity/animal/Parrot.java
index c24efae0060aabb1187d82968089a362a55358bf..9328617b7b9ee8fd4d53b187ab66c8bd0385cb09 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Parrot.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Parrot.java
@@ -190,6 +190,11 @@ public class Parrot extends ShoulderRidingEntity implements FlyingAnimal {
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.parrot.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return 6000;
+    }
     // Purpur end
 
     @Nullable
diff --git a/src/main/java/net/minecraft/world/entity/animal/Pig.java b/src/main/java/net/minecraft/world/entity/animal/Pig.java
index 75ecd7241390ca4d5502a918515df601f66b0850..9522cfff5b9ee474674e640fa7a96bbfba9f0654 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Pig.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Pig.java
@@ -84,6 +84,11 @@ public class Pig extends Animal implements ItemSteerable, Saddleable {
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.pig.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.pig.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/PolarBear.java b/src/main/java/net/minecraft/world/entity/animal/PolarBear.java
index ba2968c8734bb25fe54b2612f619de338cc19b2a..4e8b33ddc5e43b5922f871cc6f623075e0157bc8 100644
--- a/src/main/java/net/minecraft/world/entity/animal/PolarBear.java
+++ b/src/main/java/net/minecraft/world/entity/animal/PolarBear.java
@@ -118,6 +118,11 @@ public class PolarBear extends Animal implements NeutralMob {
             return this.isInLove() && bear.isInLove();
         }
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.polarBear.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/Rabbit.java b/src/main/java/net/minecraft/world/entity/animal/Rabbit.java
index 097496e4f66ec9e649edb237403e665d7161a268..8159df70cde6c031aa49befc51fbe467a26486cf 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Rabbit.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Rabbit.java
@@ -142,6 +142,11 @@ public class Rabbit extends Animal {
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.rabbit.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.rabbit.breedingDelay;
+    }
     // Purpur end
 
     // CraftBukkit start - code from constructor
diff --git a/src/main/java/net/minecraft/world/entity/animal/Sheep.java b/src/main/java/net/minecraft/world/entity/animal/Sheep.java
index d577bba65287bb0cd20a36f3bbb846fa561e2259..92742c6dcaff9399686056a14430e10f6c28a6c9 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Sheep.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Sheep.java
@@ -136,6 +136,11 @@ public class Sheep extends Animal implements Shearable {
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.sheep.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.sheep.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/Turtle.java b/src/main/java/net/minecraft/world/entity/animal/Turtle.java
index 35772b9d8faa45ec01c0d4de3eec5ded6c7a46ff..afffb7312d82751fb1c5b6fa4d6561c1b2d98252 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Turtle.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Turtle.java
@@ -102,6 +102,11 @@ public class Turtle extends Animal {
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.turtle.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.turtle.breedingDelay;
+    }
     // Purpur end
 
     public void setHomePos(BlockPos pos) {
diff --git a/src/main/java/net/minecraft/world/entity/animal/Wolf.java b/src/main/java/net/minecraft/world/entity/animal/Wolf.java
index ea5d0f850d2b26fc25e5593abc4240911aecfef9..dec5b186bd8c7beeac32d2b1e26397a687996266 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Wolf.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Wolf.java
@@ -127,6 +127,11 @@ public class Wolf extends TamableAnimal implements NeutralMob {
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.wolf.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.wolf.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/axolotl/Axolotl.java b/src/main/java/net/minecraft/world/entity/animal/axolotl/Axolotl.java
index f465080d3df4bd1b1ab90c5930d5ec88b13471ad..1dee81c41040fd607fa97351f721785ac7a5dc10 100644
--- a/src/main/java/net/minecraft/world/entity/animal/axolotl/Axolotl.java
+++ b/src/main/java/net/minecraft/world/entity/animal/axolotl/Axolotl.java
@@ -122,6 +122,11 @@ public class Axolotl extends Animal implements LerpingModel, Bucketable {
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.axolotl.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.axolotl.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java b/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java
index 32f0c117bdac8aa319e3a8844cdd90e1d8324fe8..80a837d4908dc13618a3f27358d46fd5d6e695b6 100644
--- a/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java
+++ b/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java
@@ -109,6 +109,11 @@ public class Goat extends Animal {
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.goat.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.goat.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/Donkey.java b/src/main/java/net/minecraft/world/entity/animal/horse/Donkey.java
index e9e930be31694d0cad7610475e6522b94143b00e..2a389c07fbeeea674934a6dca316db2ac59750f9 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/Donkey.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/Donkey.java
@@ -35,6 +35,11 @@ public class Donkey extends AbstractChestedHorse {
     public double generateRandomSpeed(net.minecraft.util.RandomSource random) {
         return generateRandomSpeed(this.level.purpurConfig().mobs.donkey.attributes.movementSpeedMin, this.level.purpurConfig().mobs.donkey.attributes.movementSpeedMax);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.donkey.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/Horse.java b/src/main/java/net/minecraft/world/entity/animal/horse/Horse.java
index 2ac10933dc74d382ee897ac1446c1e2f63e92d10..9191be8920ffa70f54f9c67d418d538b89231eeb 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/Horse.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/Horse.java
@@ -60,6 +60,11 @@ public class Horse extends AbstractHorse {
     public double generateRandomSpeed(RandomSource random) {
         return generateRandomSpeed(this.level.purpurConfig().mobs.horse.attributes.movementSpeedMin, this.level.purpurConfig().mobs.horse.attributes.movementSpeedMax);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.horse.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java b/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
index fd412606114908fb6e62889c927583331c26aba3..d12344b6f95562d7ecbe120411f54c7078f9bd35 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
@@ -131,6 +131,11 @@ public class Llama extends AbstractChestedHorse implements RangedAttackMob {
     public double generateRandomSpeed(RandomSource random) {
         return generateRandomSpeed(this.level.purpurConfig().mobs.llama.attributes.movementSpeedMin, this.level.purpurConfig().mobs.llama.attributes.movementSpeedMax);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.llama.breedingDelay;
+    }
     // Purpur end
 
     public boolean isTraderLlama() {
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/Mule.java b/src/main/java/net/minecraft/world/entity/animal/horse/Mule.java
index 662ea10d4b5b4e90877256e497c4934eedf53963..e6d73dfcc4eb6f4a9b8c0696cf1d0e92f9e107e3 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/Mule.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/Mule.java
@@ -34,6 +34,11 @@ public class Mule extends AbstractChestedHorse {
     public double generateRandomSpeed(net.minecraft.util.RandomSource random) {
         return generateRandomSpeed(this.level.purpurConfig().mobs.mule.attributes.movementSpeedMin, this.level.purpurConfig().mobs.mule.attributes.movementSpeedMax);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.mule.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/SkeletonHorse.java b/src/main/java/net/minecraft/world/entity/animal/horse/SkeletonHorse.java
index e15a75c8fae6f5f3d7de33f7bd7a0df1862dea94..62bda7704a1e035a4055f51aa5ebd3add38855b8 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/SkeletonHorse.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/SkeletonHorse.java
@@ -50,6 +50,11 @@ public class SkeletonHorse extends AbstractHorse {
     public double generateRandomSpeed(RandomSource random) {
         return generateRandomSpeed(this.level.purpurConfig().mobs.skeletonHorse.attributes.movementSpeedMin, this.level.purpurConfig().mobs.skeletonHorse.attributes.movementSpeedMax);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return 6000;
+    }
     // Purpur end
 
     public static AttributeSupplier.Builder createAttributes() {
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/TraderLlama.java b/src/main/java/net/minecraft/world/entity/animal/horse/TraderLlama.java
index ec69508421463fede63afc1185365d067c0fa3fa..4c734a6f3be1c703b32124f3e34c97b6085c759e 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/TraderLlama.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/TraderLlama.java
@@ -62,6 +62,11 @@ public class TraderLlama extends Llama {
     public double generateRandomSpeed(net.minecraft.util.RandomSource random) {
         return generateRandomSpeed(this.level.purpurConfig().mobs.traderLlama.attributes.movementSpeedMin, this.level.purpurConfig().mobs.traderLlama.attributes.movementSpeedMax);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.traderLlama.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/ZombieHorse.java b/src/main/java/net/minecraft/world/entity/animal/horse/ZombieHorse.java
index da36107aab9219ce9c8f40c3a80e6198e6b9f5a9..d6c8b902d960a349b46da352a175b3b02ee75a94 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/ZombieHorse.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/ZombieHorse.java
@@ -48,6 +48,11 @@ public class ZombieHorse extends AbstractHorse {
     public double generateRandomSpeed(RandomSource random) {
         return generateRandomSpeed(this.level.purpurConfig().mobs.zombieHorse.attributes.movementSpeedMin, this.level.purpurConfig().mobs.zombieHorse.attributes.movementSpeedMax);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return 6000;
+    }
     // Purpur end
 
     public static AttributeSupplier.Builder createAttributes() {
diff --git a/src/main/java/net/minecraft/world/entity/monster/Strider.java b/src/main/java/net/minecraft/world/entity/monster/Strider.java
index c65d8d7dcf4e4fea3c82d7c373e264db17b4a983..8a04d40aa1c2dd2ff5927c1d5ae851bea85b632c 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Strider.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Strider.java
@@ -116,6 +116,11 @@ public class Strider extends Animal implements ItemSteerable, Saddleable {
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.strider.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.strider.breedingDelay;
+    }
     // Purpur end
 
     public static boolean checkStriderSpawnRules(EntityType<Strider> type, LevelAccessor world, MobSpawnType spawnReason, BlockPos pos, RandomSource random) {
diff --git a/src/main/java/net/minecraft/world/entity/monster/hoglin/Hoglin.java b/src/main/java/net/minecraft/world/entity/monster/hoglin/Hoglin.java
index e837aa96800c68aff867d585eecabe52a27aaca1..fc2fbd9cdd15056001804fbb93c4b5b006583fa4 100644
--- a/src/main/java/net/minecraft/world/entity/monster/hoglin/Hoglin.java
+++ b/src/main/java/net/minecraft/world/entity/monster/hoglin/Hoglin.java
@@ -87,6 +87,11 @@ public class Hoglin extends Animal implements Enemy, HoglinBase {
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig().mobs.hoglin.attributes.maxHealth);
     }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level.purpurConfig().mobs.hoglin.breedingDelay;
+    }
     // Purpur end
 
     @Override
diff --git a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
index 49ae08b28ea1760d5e8f8e3a0196d89a853cb80e..f6868b42aa6f51caf0a90068ce5c4e77964d02ae 100644
--- a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
+++ b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
@@ -38,6 +38,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
     public class Mobs extends ConfigurationPart {
         public Axolotl axolotl;
         public class Axolotl extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
@@ -52,6 +54,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Bat bat;
         public class Bat extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
@@ -75,6 +79,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Bee bee;
         public class Bee extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
@@ -107,6 +113,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Cat cat;
         public class Cat extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
@@ -145,6 +153,7 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
         public Chicken chicken;
         public class Chicken extends ConfigurationPart {
             public boolean retaliate = false;
+            public int breedingDelay = 6000;
 
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
@@ -176,6 +185,7 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
         public Cow cow;
         public class Cow extends ConfigurationPart {
             public int mushroomsForMooshroom = 0;
+            public int breedingDelay = 6000;
 
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
@@ -233,6 +243,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Donkey donkey;
         public class Donkey extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean inWater = false;
@@ -354,6 +366,7 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
         public Fox fox;
         public class Fox extends ConfigurationPart {
             public boolean changeTypeWithTulips = false;
+            public int breedingDelay = 6000;
 
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
@@ -424,6 +437,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Goat goat;
         public class Goat extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
@@ -453,6 +468,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Hoglin hoglin;
         public class Hoglin extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
@@ -468,6 +485,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Horse horse;
         public class Horse extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean inWater = false;
@@ -543,6 +562,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Llama llama;
         public class Llama extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
@@ -588,6 +609,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Mooshroom mooshroom;
         public class Mooshroom extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
@@ -603,6 +626,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Mule mule;
         public class Mule extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean inWater = false;
@@ -621,6 +646,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Ocelot ocelot;
         public class Ocelot extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
@@ -636,6 +663,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Panda panda;
         public class Panda extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
@@ -728,6 +757,7 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
         public Pig pig;
         public class Pig extends ConfigurationPart {
             public boolean giveBackSaddle = false;
+            public int breedingDelay = 6000;
 
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
@@ -792,6 +822,7 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
             @Setting("breedable-item")
             public String breedableItemString = "";
             public transient Item breedableItem = null;
+            public int breedingDelay = 6000;
 
             @Override
             public void postProcess() {
@@ -830,6 +861,7 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
         public class Rabbit extends ConfigurationPart {
             public double spawnToastChance = 0.0D;
             public double spawnKillerChance = 0.0D;
+            public int breedingDelay = 6000;
 
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
@@ -875,6 +907,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Sheep sheep;
         public class Sheep extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
@@ -1047,6 +1081,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Strider strider;
         public class Strider extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
@@ -1062,6 +1098,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public TraderLlama traderLlama;
         public class TraderLlama extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
@@ -1096,6 +1134,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Turtle turtle;
         public class Turtle extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
@@ -1130,6 +1170,7 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
             public boolean followEmeraldBlock = false;
             public boolean canBeLeashed = false;
             public boolean canBreed = true;
+            public int breedingDelay = 6000;
 
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
@@ -1236,6 +1277,8 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
 
         public Wolf wolf;
         public class Wolf extends ConfigurationPart {
+            public int breedingDelay = 6000;
+
             public Ridable ridable;
             public class Ridable extends ConfigurationPart {
                 public boolean enabled = false;
