From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Tue, 24 Nov 2020 05:32:02 -0600
Subject: [PATCH] Add configurable snowball damage


diff --git a/src/main/java/net/minecraft/world/entity/projectile/Snowball.java b/src/main/java/net/minecraft/world/entity/projectile/Snowball.java
index 516f84a915027f63bc42f3ea0e6d8010a73091f5..33e8a11fc92b7c771ebda046752047a473c67c04 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/Snowball.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/Snowball.java
@@ -54,7 +54,7 @@ public class Snowball extends ThrowableItemProjectile {
     protected void onHitEntity(EntityHitResult entityHitResult) {
         super.onHitEntity(entityHitResult);
         Entity entity = entityHitResult.getEntity();
-        int i = entity instanceof Blaze ? 3 : 0;
+        int i = entity.level.purpurConfig().gameplayMechanics.projectileDamage.snowball >= 0 ? entity.level.purpurConfig().gameplayMechanics.projectileDamage.snowball : entity instanceof Blaze ? 3 : 0; // Purpur
         entity.hurt(DamageSource.thrown(this, this.getOwner()), (float)i);
     }
 
diff --git a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
index 1860b15a064e6c990d2cfef2a8254a5f1015f5a1..4fb7c961be48df25714eb72479a89b2ec18fb72e 100644
--- a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
+++ b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
@@ -1536,6 +1536,11 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
             public int daytimeTicks = 12000;
             public int nighttimeTicks = 12000;
         }
+
+        public ProjectileDamage projectileDamage;
+        public class ProjectileDamage extends ConfigurationPart {
+            public int snowball = -1;
+        }
     }
 
     public Blocks blocks;
