From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ben Kerllenevich <ben@omega24.dev>
Date: Mon, 30 Nov 2020 11:40:11 -0500
Subject: [PATCH] Changeable Mob Left Handed Chance


diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index c3af02075b2a4c362edb91337c81b1284be4f17e..9011b034e32f1a81604f79866167d80d97ea4612 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -1233,7 +1233,7 @@ public abstract class Mob extends LivingEntity {
         RandomSource randomsource = world.getRandom();
 
         this.getAttribute(Attributes.FOLLOW_RANGE).addPermanentModifier(new AttributeModifier("Random spawn bonus", randomsource.triangle(0.0D, 0.11485000000000001D), AttributeModifier.Operation.MULTIPLY_BASE));
-        if (randomsource.nextFloat() < 0.05F) {
+        if (randomsource.nextFloat() < world.getLevel().purpurConfig().gameplayMechanics.entityLeftHandedChance) { // Purpur
             this.setLeftHanded(true);
         } else {
             this.setLeftHanded(false);
diff --git a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
index 4fb7c961be48df25714eb72479a89b2ec18fb72e..08c6e3721ba82c18eedaac746fdd46e88c19945f 100644
--- a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
+++ b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
@@ -1411,6 +1411,7 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
         public int raidCooldownSeconds = 0;
         public int animalBreedingCooldown = 0;
         public boolean persistentDroppableEntityDisplayNames = false;
+        public float entityLeftHandedChance = 0.05f;
 
         public Player player;
         public class Player extends ConfigurationPart {
