From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Sun, 19 Jul 2020 13:32:00 -0500
Subject: [PATCH] Fix scoreboard vanilla colors not working in chat


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index cb2358872..bc4a4ee2c 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1930,7 +1930,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
                 if (this.player.getWorld().paperConfig.useVanillaScoreboardColoring) {
                     IChatBaseComponent nameFromTeam = ScoreboardTeam.a(this.player.getScoreboardTeam(), ((CraftPlayer) player).getHandle().getDisplayName());
                     // Explicitly add a RESET here, vanilla uses components for this now...
-                    displayName = CraftChatMessage.fromComponent(nameFromTeam) + org.bukkit.ChatColor.RESET;
+                    displayName = new net.md_5.bungee.api.chat.TextComponent(net.md_5.bungee.chat.ComponentSerializer.parse(IChatBaseComponent.ChatSerializer.componentToJson(nameFromTeam))).toLegacyText() + org.bukkit.ChatColor.RESET; // Purpur
                 }
 
                 s = String.format(event.getFormat(), displayName, event.getMessage());
