From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Thu, 24 Dec 2020 11:00:15 -0600
Subject: [PATCH] Full netherite armor grants fire resistance


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 1393c7a3b66abbc0be831705ac2d623c89413797..7561eb610746534fc18e33ad19c368db346efd40 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -367,6 +367,16 @@ public abstract class Player extends LivingEntity {
             this.addEffect(new MobEffectInstance(MobEffects.WATER_BREATHING, 200, 0, false, false, true), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.TURTLE_HELMET); // CraftBukkit
         }
 
+        // Purpur start
+        if (this.level.purpurConfig().gameplayMechanics.player.netheriteFireResistance.duration > 0 && this.level.getGameTime() % 20 == 0) {
+            if (itemstack.is(Items.NETHERITE_HELMET)
+                    && this.getItemBySlot(EquipmentSlot.CHEST).is(Items.NETHERITE_CHESTPLATE)
+                    && this.getItemBySlot(EquipmentSlot.LEGS).is(Items.NETHERITE_LEGGINGS)
+                    && this.getItemBySlot(EquipmentSlot.FEET).is(Items.NETHERITE_BOOTS)) {
+                this.addEffect(new MobEffectInstance(MobEffects.FIRE_RESISTANCE, this.level.purpurConfig().gameplayMechanics.player.netheriteFireResistance.duration, this.level.purpurConfig().gameplayMechanics.player.netheriteFireResistance.amplifier, this.level.purpurConfig().gameplayMechanics.player.netheriteFireResistance.ambient, this.level.purpurConfig().gameplayMechanics.player.netheriteFireResistance.showParticals, this.level.purpurConfig().gameplayMechanics.player.netheriteFireResistance.showIcon), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.NETHERITE_ARMOR);
+            }
+        }
+        // Purpur end
     }
 
     protected ItemCooldowns createItemCooldowns() {
diff --git a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
index f77cd1ba5cbe70c2d6613a433cb78640413c7b96..35857a071a49a55b451f0c008248edaeba70b438 100644
--- a/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
+++ b/src/main/java/org/purpurmc/purpur/configuration/PurpurWorldConfiguration.java
@@ -1457,6 +1457,15 @@ public class PurpurWorldConfiguration extends ConfigurationPart {
                 public String expDropEquation = "expLevel * 7";
                 public int maxExpDropped = 100;
             }
+
+            public NetheriteFireResistance netheriteFireResistance;
+            public class NetheriteFireResistance extends ConfigurationPart {
+                public int duration = 0;
+                public int amplifier = 0;
+                public boolean ambient = false;
+                public boolean showParticals = false;
+                public boolean showIcon = true;
+            }
         }
 
         public SilkTouchSpawner silkTouchSpawner;
