From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Sat, 15 Aug 2020 09:19:56 -0500
Subject: [PATCH] Fix incorrect return for WorldServer#addAllEntitiesSafely


diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 5e4ccb0663..3c159c6fa7 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -1712,7 +1712,10 @@ public class WorldServer extends World implements GeneratorAccessSeed {
         if (entity.co().anyMatch(this::isUUIDTaken)) {
             return false;
         } else {
-            return this.addAllEntities(entity, reason); // CraftBukkit
+            // Purpur start - this MUST return true if UUID is not taken regardless if actually spawned or not
+            this.addAllEntities(entity, reason); // CraftBukkit
+            return true;
+            // Purpur end
         }
     }
 
