From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Mon, 30 Nov 2020 06:03:06 -0600
Subject: [PATCH] Left handed API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
index 6549d7c40d6a0ca307fdcb6fd3ca01d2ab732b59..c128bb7fbea43e9b4245870d78f920dfb23e23f3 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
@@ -132,4 +132,16 @@ public abstract class CraftMob extends CraftLivingEntity implements Mob {
         return getHandle().getMaxHeadXRot();
     }
     // Paper end
+
+    // Purpur start
+    @Override
+    public boolean isLeftHanded() {
+        return getHandle().isLeftHanded();
+    }
+
+    @Override
+    public void setLeftHanded(boolean leftHanded) {
+        getHandle().setLeftHanded(leftHanded);
+    }
+    // Purpur end
 }
