From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date: Wed, 2 Dec 2020 03:07:58 -0800
Subject: [PATCH] PaperPR - Config option for Piglins guarding chests


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index b07368ed5b8fcacd617988d09d5d27904315cf27..11576c89a4d7a6a29b1d17a6126fcba8154a6a4b 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -49,6 +49,11 @@ public class PaperWorldConfig {
         zombiesTargetTurtleEggs = getBoolean("zombies-target-turtle-eggs", zombiesTargetTurtleEggs);
     }
 
+    public boolean piglinsGuardChests = true;
+    private void piglinsGuardChests() {
+        piglinsGuardChests = getBoolean("piglins-guard-chests", piglinsGuardChests);
+    }
+
     public boolean useEigencraftRedstone = false;
     private void useEigencraftRedstone() {
         useEigencraftRedstone = this.getBoolean("use-faster-eigencraft-redstone", false);
diff --git a/src/main/java/net/minecraft/world/entity/monster/piglin/PiglinAi.java b/src/main/java/net/minecraft/world/entity/monster/piglin/PiglinAi.java
index 01d307e0022656509bc8fd02d6bac78a6af3a7be..2366735f571113a543fd15607b5ff74c09cd0d24 100644
--- a/src/main/java/net/minecraft/world/entity/monster/piglin/PiglinAi.java
+++ b/src/main/java/net/minecraft/world/entity/monster/piglin/PiglinAi.java
@@ -468,6 +468,7 @@ public class PiglinAi {
     }
 
     public static void angerNearbyPiglins(Player player, boolean blockOpen) {
+        if (!player.level.paperConfig.piglinsGuardChests) return; // Paper
         List<AbstractPiglin> list = (List) player.level.getEntitiesOfClass(Piglin.class, player.getBoundingBox().inflate(16.0D)); // CraftBukkit - decompile error
 
         list.stream().filter(PiglinAi::isIdle).filter((entitypiglin) -> {
