From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: jmp <jasonpenilla2@me.com>
Date: Tue, 17 Nov 2020 13:12:09 -0800
Subject: [PATCH] Set name visible when using a Name Tag on an Armor Stand


diff --git a/src/main/java/net/minecraft/server/ItemNameTag.java b/src/main/java/net/minecraft/server/ItemNameTag.java
index 01163ce38602f9345f00ee0535b4e73be7c6d735..a7efce97318fcf95d98f33ad4ac2da69a1ba0df0 100644
--- a/src/main/java/net/minecraft/server/ItemNameTag.java
+++ b/src/main/java/net/minecraft/server/ItemNameTag.java
@@ -11,6 +11,11 @@ public class ItemNameTag extends Item {
         if (itemstack.hasName() && !(entityliving instanceof EntityHuman)) {
             if (!entityhuman.world.isClientSide && entityliving.isAlive()) {
                 entityliving.setCustomName(itemstack.getName());
+                // Purpur start
+                if (entityhuman.world.purpurConfig.armorstandFixNametags && entityliving instanceof EntityArmorStand) {
+                    entityliving.setCustomNameVisible(true);
+                }
+                // Purpur end
                 if (entityliving instanceof EntityInsentient) {
                     ((EntityInsentient) entityliving).setPersistent();
                 }
diff --git a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
index af3f910e639a17dbcade8609a3c8c42b0d138cbb..a77debc7a8923693733d748ce9abd3ae2d057fb9 100644
--- a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
+++ b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
@@ -84,9 +84,11 @@ public class PurpurWorldConfig {
     }
 
     public boolean armorstandSetNameVisible = false;
+    public boolean armorstandFixNametags = false;
     public float armorstandStepHeight = 0.0F;
     private void armorstandSettings() {
         armorstandSetNameVisible = getBoolean("gameplay-mechanics.armorstand.set-name-visible-when-placing-with-custom-name", armorstandSetNameVisible);
+        armorstandFixNametags = getBoolean("gameplay-mechanics.armorstand.fix-nametags", armorstandFixNametags);
         armorstandStepHeight = (float) getDouble("gameplay-mechanics.armorstand.step-height", armorstandStepHeight);
     }
 
