From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: James Lyne <jim+github@not-null.co.uk>
Date: Mon, 7 Dec 2020 17:52:36 +0000
Subject: [PATCH] Spread out and optimise player list ticks


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 3501c42e11a5b1f690bcbf26bcea4ebbfbba1cb1..48b4e6dce27448bf4c83d5339cd35c24c5485df5 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -73,7 +73,7 @@ public abstract class PlayerList {
     private int viewDistance;
     private EnumGamemode u;
     private boolean v;
-    private int w;
+    private int w; private int getTick() { return this.w; } private int setTick(int i) { return this.w = i; } // Purpur - OBFHELPER
 
     // CraftBukkit start
     private CraftServer cserver;
@@ -933,22 +933,23 @@ public abstract class PlayerList {
     }
 
     public void tick() {
-        if (++this.w > 600) {
-            // CraftBukkit start
-            for (int i = 0; i < this.players.size(); ++i) {
-                final EntityPlayer target = (EntityPlayer) this.players.get(i);
-
-                target.playerConnection.sendPacket(new PacketPlayOutPlayerInfo(PacketPlayOutPlayerInfo.EnumPlayerInfoAction.UPDATE_LATENCY, Iterables.filter(this.players, new Predicate<EntityPlayer>() {
-                    @Override
-                    public boolean apply(EntityPlayer input) {
-                        return target.getBukkitEntity().canSee(input.getBukkitEntity());
-                    }
-                })));
+        // Purpur start
+        int tick = getTick();
+        if (tick < this.players.size()) {
+            final org.bukkit.craftbukkit.entity.CraftPlayer target = this.players.get(tick).getBukkitEntity();
+            final java.util.List<EntityPlayer> list = new java.util.ArrayList<>();
+            for (EntityPlayer entityplayer : this.players) {
+                if (target.canSee(entityplayer.getUniqueID())) {
+                    list.add(entityplayer);
+                }
             }
-            // CraftBukkit end
-            this.w = 0;
+            target.getHandle().playerConnection.sendPacket(new PacketPlayOutPlayerInfo(PacketPlayOutPlayerInfo.EnumPlayerInfoAction.UPDATE_LATENCY, list));
         }
-
+        if (++tick > 600) {
+            tick = 0;
+        }
+        setTick(tick);
+        // Purpur end
     }
 
     public void sendAll(Packet<?> packet) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 2b25e58ce0d5b6d54c65e70cffd46e1085b02275..86d10c342a741e85e434b65ae172bbdf5284d23b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1435,7 +1435,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     @Override
     public boolean canSee(Player player) {
-        return !hiddenPlayers.containsKey(player.getUniqueId());
+        // Purpur start
+        return canSee(player.getUniqueId());
+    }
+
+    public boolean canSee(UUID uuid) {
+        return !hiddenPlayers.containsKey(uuid);
+        // Purpur end
     }
 
     @Override
