From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Wed, 6 Jan 2021 02:19:29 -0600
Subject: [PATCH] Fix rotating UP/DOWN CW and CCW


diff --git a/src/main/java/net/minecraft/core/Direction.java b/src/main/java/net/minecraft/core/Direction.java
index 593d6251c75ec337175d08b85000239ba7da1af2..c040c9ce886f9970a70867344fd0e0b226d3ff9f 100644
--- a/src/main/java/net/minecraft/core/Direction.java
+++ b/src/main/java/net/minecraft/core/Direction.java
@@ -235,6 +235,12 @@ public enum Direction implements StringRepresentable {
             return NORTH;
         case EAST:
             return SOUTH;
+        // Purpur start
+        case UP:
+            return UP;
+        case DOWN:
+            return DOWN;
+        // Purpur end
         default:
             throw new IllegalStateException("Unable to get Y-rotated facing of " + this);
         }
@@ -314,6 +320,12 @@ public enum Direction implements StringRepresentable {
             return SOUTH;
         case EAST:
             return NORTH;
+        // Purpur start
+        case UP:
+            return UP;
+        case DOWN:
+            return DOWN;
+        // Purpur end
         default:
             throw new IllegalStateException("Unable to get CCW facing of " + this);
         }
