From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Thu, 7 Jan 2021 07:51:25 -0600
Subject: [PATCH] Add MC-4 fix back


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index d262196f88f32ecaca476e4a1d549bd456516f77..f9516fb2072dff1426d249ee25449e1c7c9cf625 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -4073,7 +4073,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
         // Paper end
         // Paper start - fix MC-4
         if (this instanceof ItemEntity) {
-            if (false && com.destroystokyo.paper.PaperConfig.fixEntityPositionDesync) { // Tuinity - revert
+            if (com.destroystokyo.paper.PaperConfig.fixEntityPositionDesync) { // Purpur
                 // encode/decode from PacketPlayOutEntity
                 x = Mth.lfloor(x * 4096.0D) * (1 / 4096.0D);
                 y = Mth.lfloor(y * 4096.0D) * (1 / 4096.0D);
