From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Wed, 6 Jan 2021 02:19:29 -0600
Subject: [PATCH] Fix rotating UP/DOWN CW and CCW


diff --git a/src/main/java/net/minecraft/server/EnumDirection.java b/src/main/java/net/minecraft/server/EnumDirection.java
index c187f04ef657e9012d2fcdc2fd5b4cca69d78c3f..a660033af081cef69e4646c9ed9eaa66cf90217f 100644
--- a/src/main/java/net/minecraft/server/EnumDirection.java
+++ b/src/main/java/net/minecraft/server/EnumDirection.java
@@ -113,6 +113,12 @@ public enum EnumDirection implements INamable {
                 return EnumDirection.NORTH;
             case EAST:
                 return EnumDirection.SOUTH;
+            // Purpur start
+            case UP:
+                return EnumDirection.UP;
+            case DOWN:
+                return EnumDirection.DOWN;
+            // Purpur end
             default:
                 throw new IllegalStateException("Unable to get Y-rotated facing of " + this);
         }
@@ -129,6 +135,12 @@ public enum EnumDirection implements INamable {
                 return EnumDirection.SOUTH;
             case EAST:
                 return EnumDirection.NORTH;
+            // Purpur start
+            case UP:
+                return EnumDirection.UP;
+            case DOWN:
+                return EnumDirection.DOWN;
+            // Purpur end
             default:
                 throw new IllegalStateException("Unable to get CCW facing of " + this);
         }
