From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Mon, 25 Jan 2021 10:22:28 -0600
Subject: [PATCH] Fix PlayerEditBookEvent not saving new book


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 75d955948a407d94e6f3a88f86afa8b1d6ba33cb..ca190409b6718607183df2e86f0b4be98c9b6c33 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1036,7 +1036,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
             list.stream().map(s -> s = color(s, hasPerm, false)).map(NBTTagString::a).forEach(nbttaglist::add); // Purpur - edit book
             ItemStack old = itemstack.cloneItemStack(); // CraftBukkit
             itemstack.a("pages", (NBTBase) nbttaglist);
-            CraftEventFactory.handleEditBookEvent(player, i, old, itemstack); // CraftBukkit
+            this.player.inventory.setItem(i, CraftEventFactory.handleEditBookEvent(player, i, old, itemstack)); // CraftBukkit // Purpur
         }
     }
 
