From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ben Kerllenevich <me@notom3ga.me>
Date: Sat, 13 Feb 2021 09:28:56 -0500
Subject: [PATCH] End crystal explosion options


diff --git a/src/main/java/net/minecraft/server/EntityEnderCrystal.java b/src/main/java/net/minecraft/server/EntityEnderCrystal.java
index 1942fae27051af79b6eb1d790a219da100bf889e..1200a76047c212f13d36b6af08a46f179ddb67c1 100644
--- a/src/main/java/net/minecraft/server/EntityEnderCrystal.java
+++ b/src/main/java/net/minecraft/server/EntityEnderCrystal.java
@@ -143,15 +143,24 @@ public class EntityEnderCrystal extends Entity {
                 // CraftBukkit end
                 this.die();
                 if (!damagesource.isExplosion()) {
+                    // Purpur start
+                    if (isShowingBottom() ? world.purpurConfig.baseCrystalExplode : world.purpurConfig.baselessCrystalExplode) {
                     // CraftBukkit start
-                    ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), 6.0F, false);
+                    ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(),
+                            (float) (isShowingBottom() ? world.purpurConfig.baseCrystalExplosionPower : world.purpurConfig.baselessCrystalExplosionPower),
+                            isShowingBottom() ? world.purpurConfig.baseCrystalExplosionFire : world.purpurConfig.baselessCrystalExplosionFire);
                     this.world.getServer().getPluginManager().callEvent(event);
                     if (event.isCancelled()) {
                         this.dead = false;
                         return false;
                     }
-                    this.world.createExplosion(this, this.locX(), this.locY(), this.locZ(), event.getRadius(), event.getFire(), Explosion.Effect.DESTROY);
+                    this.world.createExplosion(this, this.locX(), this.locY(), this.locZ(), event.getRadius(), event.getFire(),
+                            isShowingBottom() ? world.purpurConfig.baseCrystalExplosionEffect : world.purpurConfig.baselessCrystalExplosionEffect);
                     // CraftBukkit end
+                    } else {
+                        this.dead = false;
+                    }
+                    // Purpur End
                 }
 
                 this.a(damagesource);
diff --git a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
index ad6f5bf3a2f0734ffeba044025a52fd15e9970fb..a9e6f4241496a946468d1292b784bda4f0e174ca 100644
--- a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
+++ b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
@@ -1907,4 +1907,33 @@ public class PurpurWorldConfig {
     private void imposeTeleportRestrictionsOnGateways() {
         imposeTeleportRestrictionsOnGateways = getBoolean("gameplay-mechanics.impose-teleport-restrictions-on-gateways", imposeTeleportRestrictionsOnGateways);
     }
+
+    public boolean baselessCrystalExplode = true;
+    public double baselessCrystalExplosionPower = 6.0D;
+    public boolean baselessCrystalExplosionFire = false;
+    public Explosion.Effect baselessCrystalExplosionEffect = Explosion.Effect.DESTROY;
+    public boolean baseCrystalExplode = true;
+    public double baseCrystalExplosionPower = 6.0D;
+    public boolean baseCrystalExplosionFire = false;
+    public Explosion.Effect baseCrystalExplosionEffect = Explosion.Effect.DESTROY;
+    private void crystalSettings() {
+        baselessCrystalExplode = getBoolean("blocks.end-crystal.baseless.explode", baselessCrystalExplode);
+        baselessCrystalExplosionPower = getDouble("blocks.end-crystal.baseless.explosion-power", baselessCrystalExplosionPower);
+        baselessCrystalExplosionFire = getBoolean("blocks.end-crystal.baseless.explosion-fire", baselessCrystalExplosionFire);
+        try {
+            baselessCrystalExplosionEffect = Explosion.Effect.valueOf(getString("blocks.end-crystal.baseless.explosion-effect", baselessCrystalExplosionEffect.name()));
+        } catch (IllegalArgumentException e) {
+            log(Level.SEVERE, "Unknown value for `blocks.end-crystal.baseless.explosion-effect`! Using default of `DESTROY`");
+            baselessCrystalExplosionEffect = Explosion.Effect.DESTROY;
+        }
+        baseCrystalExplode = getBoolean("blocks.end-crystal.base.explode", baseCrystalExplode);
+        baseCrystalExplosionPower = getDouble("blocks.end-crystal.base.explosion-power", baseCrystalExplosionPower);
+        baseCrystalExplosionFire = getBoolean("blocks.end-crystal.base.explosion-fire", baseCrystalExplosionFire);
+        try {
+            baseCrystalExplosionEffect = Explosion.Effect.valueOf(getString("blocks.end-crystal.base.explosion-effect", baseCrystalExplosionEffect.name()));
+        } catch (IllegalArgumentException e) {
+            log(Level.SEVERE, "Unknown value for `blocks.end-crystal.base.explosion-effect`! Using default of `DESTROY`");
+            baseCrystalExplosionEffect = Explosion.Effect.DESTROY;
+        }
+    }
 }
