From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ben Kerllenevich <me@notom3ga.me>
Date: Thu, 18 Mar 2021 07:23:27 -0400
Subject: [PATCH] Add config for villager trading


diff --git a/src/main/java/net/minecraft/server/EntityVillager.java b/src/main/java/net/minecraft/server/EntityVillager.java
index 4f5ad3836d39ccb88923a38f2fd556cb88ec241a..75328bfb735c1066b64393c33c3da82673c74d92 100644
--- a/src/main/java/net/minecraft/server/EntityVillager.java
+++ b/src/main/java/net/minecraft/server/EntityVillager.java
@@ -299,7 +299,7 @@ public class EntityVillager extends EntityVillagerAbstract implements Reputation
                     return tryRide(entityhuman, enumhand); // Purpur
                 } else {
                     if (world.purpurConfig.villagerRidable && itemstack.isEmpty()) return tryRide(entityhuman, enumhand); // Purpur
-                    if (!this.world.isClientSide && !this.trades.isEmpty()) {
+                    if (!this.world.isClientSide && this.world.purpurConfig.allowVillagerTrading && !this.trades.isEmpty()) { // Purpur
                         this.h(entityhuman);
                     }
 
diff --git a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
index 0f385273214c469e0f0d8a57b6c9bad9a79d221b..b7e357fffc3e4de6ff0e83115a066b7a0f8b4907 100644
--- a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
+++ b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
@@ -1705,6 +1705,7 @@ public class PurpurWorldConfig {
     public boolean villagerClericsFarmWarts = false;
     public boolean villagerClericFarmersThrowWarts = true;
     public double villagerMaxHealth = 20.0D;
+    public boolean allowVillagerTrading = true;
     private void villagerSettings() {
         villagerRidable = getBoolean("mobs.villager.ridable", villagerRidable);
         villagerRidableInWater = getBoolean("mobs.villager.ridable-in-water", villagerRidableInWater);
@@ -1732,6 +1733,7 @@ public class PurpurWorldConfig {
             set("mobs.villager.attributes.max_health", oldValue);
         }
         villagerMaxHealth = getDouble("mobs.villager.attributes.max_health", villagerMaxHealth);
+        allowVillagerTrading = getBoolean("mobs.villager.allow-trading", allowVillagerTrading);
     }
 
     public boolean villagerTraderRidable = false;
