From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Mon, 10 May 2021 02:08:53 -0500
Subject: [PATCH] Fix SPIGOT-6278


diff --git a/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java b/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
index e51d4254170a68900e95e697120d24b2b8b0ade4..cdb797f23b4460a24a090d3983debcc523913943 100644
--- a/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -217,7 +217,7 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
 
     @Override
     public boolean removeWhenFarAway(double distanceSquared) {
-        return !this.isPersistenceRequired();
+        return true || !this.isPersistenceRequired(); // Purpur - fix SPIGOT-6278
     }
 
     @Override
