From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ben Kerllenevich <me@notom3ga.me>
Date: Thu, 29 Apr 2021 14:06:29 -0400
Subject: [PATCH] Configurable damage settings for magma blocks


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 706a1e2fadb0880277093be5de63f4dc0792fb72..f498bba37f3a40060a0847500b70e7cb55c745cd 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -894,9 +894,15 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, ne
             }
             // CraftBukkit end
 
+            // Purpur start
+            if (block instanceof net.minecraft.world.level.block.BlockMagma && this.onGround) {
+                block.stepOn(this.world, blockposition, this);
+            } else {
             if (this.onGround && !this.bv()) {
                 block.stepOn(this.world, blockposition, this);
             }
+            }
+            // Purpur end
 
             if (this.playStepSound() && !this.isPassenger()) {
                 double d0 = vec3d1.x;
diff --git a/src/main/java/net/minecraft/world/item/enchantment/EnchantmentManager.java b/src/main/java/net/minecraft/world/item/enchantment/EnchantmentManager.java
index b2d28c2bf0a9e93d38583e2d734c12fed4f63d5d..b889e5426781da215b5c26477015d04457142213 100644
--- a/src/main/java/net/minecraft/world/item/enchantment/EnchantmentManager.java
+++ b/src/main/java/net/minecraft/world/item/enchantment/EnchantmentManager.java
@@ -241,7 +241,7 @@ public class EnchantmentManager {
         return a(Enchantments.WATER_WORKER, entityliving) > 0;
     }
 
-    public static boolean i(EntityLiving entityliving) {
+    public static boolean hasFrostWalker(EntityLiving entityLiving) { return i(entityLiving);} public static boolean i(EntityLiving entityliving) { // Purpur - OBFHELPER
         return a(Enchantments.FROST_WALKER, entityliving) > 0;
     }
 
diff --git a/src/main/java/net/minecraft/world/level/block/BlockMagma.java b/src/main/java/net/minecraft/world/level/block/BlockMagma.java
index 4559085fa4452d3a9f59ed967ccb69a7823718e5..f1c53428c8f9eb0922fc16f152c6566d1a32cac0 100644
--- a/src/main/java/net/minecraft/world/level/block/BlockMagma.java
+++ b/src/main/java/net/minecraft/world/level/block/BlockMagma.java
@@ -27,7 +27,7 @@ public class BlockMagma extends Block {
 
     @Override
     public void stepOn(World world, BlockPosition blockposition, Entity entity) {
-        if (!entity.isFireProof() && entity instanceof EntityLiving && !EnchantmentManager.i((EntityLiving) entity)) {
+        if (!entity.isFireProof() && entity instanceof EntityLiving && (world.purpurConfig.magmaBlockDamageWhenSneaking || !entity.isSneaking()) && (world.purpurConfig.magmaBlockDamageWithFrostWalker || !EnchantmentManager.hasFrostWalker((EntityLiving) entity))) { // Purpur
             org.bukkit.craftbukkit.event.CraftEventFactory.blockDamage = world.getWorld().getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ()); // CraftBukkit
             entity.damageEntity(DamageSource.HOT_FLOOR, 1.0F);
             org.bukkit.craftbukkit.event.CraftEventFactory.blockDamage = null; // CraftBukkit
diff --git a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
index f3cde1e3d06da82b7a54898df5d331f279b2a954..78e39c9424c296bf07f89d63ce42b02b96c2cd35 100644
--- a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
+++ b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
@@ -2124,4 +2124,11 @@ public class PurpurWorldConfig {
             baseCrystalExplosionEffect = Explosion.Effect.DESTROY;
         }
     }
+
+    public boolean magmaBlockDamageWhenSneaking = false;
+    public boolean magmaBlockDamageWithFrostWalker = false;
+    private void magmaBlockSettings() {
+        magmaBlockDamageWhenSneaking = getBoolean("blocks.magma-block.damage-when-sneaking", magmaBlockDamageWhenSneaking);
+        magmaBlockDamageWithFrostWalker = getBoolean("blocks.magma-block.damage-with-frost-walker", magmaBlockDamageWithFrostWalker);
+    }
 }
