From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Thu, 1 Jul 2021 19:25:05 -0500
Subject: [PATCH] Structure seed options


diff --git a/src/main/java/net/minecraft/world/level/chunk/ChunkGenerator.java b/src/main/java/net/minecraft/world/level/chunk/ChunkGenerator.java
index f8a2aa31e38f64f88a82d5a388b58f1962d5fda0..9d22417b254dca722ea287a7d64b76efe1b851f8 100644
--- a/src/main/java/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/src/main/java/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -580,6 +580,23 @@ public abstract class ChunkGenerator implements BiomeManager.NoiseBiomeSource {
                     case "village":
                         seed = conf.villageSeed;
                         break;
+                    // Purpur start
+                    case "stronghold":
+                        if (net.pl3x.purpur.PurpurConfig.seedStructureStronghold != -1) {
+                            seed = net.pl3x.purpur.PurpurConfig.seedStructureStronghold;
+                        }
+                        break;
+                    case "mineshaft":
+                        if (net.pl3x.purpur.PurpurConfig.seedStructureMineshaft != -1) {
+                            seed = net.pl3x.purpur.PurpurConfig.seedStructureMineshaft;
+                        }
+                        break;
+                    case "buried_treasure":
+                        if (net.pl3x.purpur.PurpurConfig.seedStructureBuriedTreasure != -1) {
+                            seed = net.pl3x.purpur.PurpurConfig.seedStructureBuriedTreasure;
+                        }
+                        break;
+                    // Purpur end
                 }
 
                 updated.put(entry.getKey(), new StructureFeatureConfiguration(feature.spacing(), feature.separation(), seed));
diff --git a/src/main/java/org/purpurmc/purpur/PurpurConfig.java b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
index c419af801e31bec057b04131ca5634aad19b343e..51283ec0bd1903c0eb3c04c7ed90785795c39311 100644
--- a/src/main/java/org/purpurmc/purpur/PurpurConfig.java
+++ b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
@@ -194,6 +194,17 @@ public class PurpurConfig {
         deathMessageOnlyBroadcastToAffectedPlayer = getBoolean("settings.broadcasts.death.only-broadcast-to-affected-player", deathMessageOnlyBroadcastToAffectedPlayer);
     }
 
+    public static int seedStructureBuriedTreasure = -1;
+    public static int seedStructureMineshaft = -1;
+    public static int seedStructureStronghold = -1;
+    private static void seedSettings() {
+        seedStructureBuriedTreasure = getInt("settings.seed.structure.buried_treasure", seedStructureBuriedTreasure);
+        seedStructureMineshaft = getInt("settings.seed.structure.mineshaft", seedStructureMineshaft);
+        seedStructureStronghold = getInt("settings.seed.structure.stronghold", seedStructureStronghold);
+        // hide these from timings report
+        if (!TimingsManager.hiddenConfigs.contains("settings.seed")) TimingsManager.hiddenConfigs.add("settings.seed");
+    }
+
     public static String serverModName = "Purpur";
     private static void serverModName() {
         serverModName = getString("settings.server-mod-name", serverModName);
