From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Fri, 25 Jun 2021 22:03:45 -0500
Subject: [PATCH] Fix MC-229490


diff --git a/src/main/java/net/minecraft/world/entity/animal/Chicken.java b/src/main/java/net/minecraft/world/entity/animal/Chicken.java
index a907520c2cd9eacf2fe80d87004afc5f30b6a630..49e26bf3cb8e7de1071a4893a7ea1ceba9912d6a 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Chicken.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Chicken.java
@@ -112,6 +112,7 @@ public class Chicken extends Animal {
         // CraftBukkit start
         if (this.isChickenJockey()) {
             this.persistenceRequired = !this.removeWhenFarAway(0);
+            if (this.level.purpurConfig.chickenFixMC229490) this.persist = false; // Purpur - fix MC-229490
         }
         // CraftBukkit end
         super.aiStep();
diff --git a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
index 6eb1b895517436a552796d416c3785556fe57c05..2354c740741597271273a1980b3e3ac984d524d3 100644
--- a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
+++ b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
@@ -874,6 +874,7 @@ public class PurpurWorldConfig {
     public double chickenMaxHealth = 4.0D;
     public boolean chickenRetaliate = false;
     public int chickenBreedingTicks = 6000;
+    public boolean chickenFixMC229490 = true;
     private void chickenSettings() {
         chickenRidable = getBoolean("mobs.chicken.ridable", chickenRidable);
         chickenRidableInWater = getBoolean("mobs.chicken.ridable-in-water", chickenRidableInWater);
@@ -885,6 +886,7 @@ public class PurpurWorldConfig {
         chickenMaxHealth = getDouble("mobs.chicken.attributes.max_health", chickenMaxHealth);
         chickenRetaliate = getBoolean("mobs.chicken.retaliate", chickenRetaliate);
         chickenBreedingTicks = getInt("mobs.chicken.breeding-delay-ticks", chickenBreedingTicks);
+        chickenFixMC229490 = getBoolean("mobs.chicken.fix-mc-229490", chickenFixMC229490);
     }
 
     public boolean codRidable = false;
