From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DoctaEnkoda <bierquejason@gmail.com>
Date: Thu, 27 May 2021 06:46:30 +0200
Subject: [PATCH] Bee can work when rains or night


diff --git a/src/main/java/net/minecraft/world/entity/animal/EntityBee.java b/src/main/java/net/minecraft/world/entity/animal/EntityBee.java
index e425e093c233a21e5ef457e3a94defe8b74261d2..e1c4ca1f2f6613d78b340d86939f788fc7feceb6 100644
--- a/src/main/java/net/minecraft/world/entity/animal/EntityBee.java
+++ b/src/main/java/net/minecraft/world/entity/animal/EntityBee.java
@@ -356,7 +356,7 @@ public class EntityBee extends EntityAnimal implements IEntityAngerable, EntityB
 
     private boolean fd() {
         if (this.cannotEnterHiveTicks <= 0 && !this.bC.k() && !this.hasStung() && this.getGoalTarget() == null) {
-            boolean flag = this.canPollinate() || this.world.isRaining() || this.world.isNight() || this.hasNectar();
+            boolean flag = this.canPollinate() || (this.world.isRaining() && !this.world.purpurConfig.beeCanWorkRain) || (this.world.isNight() && !this.world.purpurConfig.beeCanWorkNight) || this.hasNectar(); // Purpur
 
             return flag && !this.ff();
         } else {
diff --git a/src/main/java/net/minecraft/world/level/block/entity/TileEntityBeehive.java b/src/main/java/net/minecraft/world/level/block/entity/TileEntityBeehive.java
index 58bf1008ab19340bce5111e006a8de0e7f39e0e5..2e53ca467e3422bfdc6f81fc4db17655993be442 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/TileEntityBeehive.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/TileEntityBeehive.java
@@ -181,7 +181,7 @@ public class TileEntityBeehive extends TileEntity implements ITickable {
     }
 
     private boolean releaseBee(IBlockData iblockdata, TileEntityBeehive.HiveBee tileentitybeehive_hivebee, @Nullable List<Entity> list, TileEntityBeehive.ReleaseStatus tileentitybeehive_releasestatus, boolean force) {
-        if (!force && (this.world.isNight() || this.world.isRaining()) && tileentitybeehive_releasestatus != TileEntityBeehive.ReleaseStatus.EMERGENCY) {
+        if (!force && ((this.world.isNight() && !this.world.purpurConfig.beeCanWorkNight) || (this.world.isRaining() && !this.world.purpurConfig.beeCanWorkRain)) && tileentitybeehive_releasestatus != TileEntityBeehive.ReleaseStatus.EMERGENCY) { // Purpur
             // CraftBukkit end
             return false;
         } else {
diff --git a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
index cda80cf8d2556defd1453b473aca76bf7c014794..5321c87308cf1c79e4c8b8c20356743b5c64f568 100644
--- a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
+++ b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
@@ -721,6 +721,8 @@ public class PurpurWorldConfig {
     public double beeMaxY = 256D;
     public int beeBreedingTicks = 6000;
     public double beeMaxHealth = 10.0D;
+    public boolean beeCanWorkNight = false;
+    public boolean beeCanWorkRain = false;
     private void beeSettings() {
         beeRidable = getBoolean("mobs.bee.ridable", beeRidable);
         beeRidableInWater = getBoolean("mobs.bee.ridable-in-water", beeRidableInWater);
@@ -732,6 +734,8 @@ public class PurpurWorldConfig {
             set("mobs.bee.attributes.max_health", oldValue);
         }
         beeMaxHealth = getDouble("mobs.bee.attributes.max_health", beeMaxHealth);
+        beeCanWorkNight = getBoolean("mobs.bee.can-work-when-night", beeCanWorkNight);
+        beeCanWorkRain = getBoolean("mobs.bee.can-work-when-rain", beeCanWorkRain);
     }
 
     public boolean blazeRidable = false;
