From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DoctaEnkoda <bierquejason@gmail.com>
Date: Wed, 26 May 2021 02:48:38 +0200
Subject: [PATCH] Allows to display the death message only per player


diff --git a/src/main/java/net/minecraft/server/level/EntityPlayer.java b/src/main/java/net/minecraft/server/level/EntityPlayer.java
index 96520df171fadbc90c130555bd0c7139579f73d8..95e937f634281a2b709dd339bb099f9e246b9e4e 100644
--- a/src/main/java/net/minecraft/server/level/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/level/EntityPlayer.java
@@ -1041,12 +1041,26 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
 
             if (scoreboardteambase != null && scoreboardteambase.getDeathMessageVisibility() != ScoreboardTeamBase.EnumNameTagVisibility.ALWAYS) {
                 if (scoreboardteambase.getDeathMessageVisibility() == ScoreboardTeamBase.EnumNameTagVisibility.HIDE_FOR_OTHER_TEAMS) {
-                    this.server.getPlayerList().a((EntityHuman) this, ichatbasecomponent);
+                    // Purpur Start - DeathMessage By Player
+                    if (net.pl3x.purpur.PurpurConfig.deathMessageOnlyBroadcastToAffectedPlayer) {
+                        this.sendMessage(ichatbasecomponent);
+                    } else {
+                        this.server.getPlayerList().a((EntityHuman) this, ichatbasecomponent);
+                    }
                 } else if (scoreboardteambase.getDeathMessageVisibility() == ScoreboardTeamBase.EnumNameTagVisibility.HIDE_FOR_OWN_TEAM) {
-                    this.server.getPlayerList().b(this, ichatbasecomponent);
+                    if (net.pl3x.purpur.PurpurConfig.deathMessageOnlyBroadcastToAffectedPlayer) {
+                        this.sendMessage(ichatbasecomponent);
+                    } else {
+                        this.server.getPlayerList().b(this, ichatbasecomponent);
+                    }
                 }
             } else {
-                this.server.getPlayerList().sendMessage(ichatbasecomponent, ChatMessageType.SYSTEM, SystemUtils.b);
+                if (net.pl3x.purpur.PurpurConfig.deathMessageOnlyBroadcastToAffectedPlayer) {
+                    this.sendMessage(ichatbasecomponent);
+                } else {
+                    this.server.getPlayerList().sendMessage(ichatbasecomponent, ChatMessageType.SYSTEM, SystemUtils.getNullUUID());
+                }
+                // Purpur End
             }
         } else {
             this.playerConnection.sendPacket(new PacketPlayOutCombatEvent(this.getCombatTracker(), PacketPlayOutCombatEvent.EnumCombatEventType.ENTITY_DIED));
diff --git a/src/main/java/net/pl3x/purpur/PurpurConfig.java b/src/main/java/net/pl3x/purpur/PurpurConfig.java
index ce9cbc3964ef626e7fe3baf59225114ca93b0b53..c53e9cc22161cf2a35d2d7ba57fa08f244129b8a 100644
--- a/src/main/java/net/pl3x/purpur/PurpurConfig.java
+++ b/src/main/java/net/pl3x/purpur/PurpurConfig.java
@@ -298,4 +298,9 @@ public class PurpurConfig {
     private static void advancementSettings() {
         advancementOnlyBroadcastToAffectedPlayer  = getBoolean("settings.advancement.only-broadcast-to-affected-player", advancementOnlyBroadcastToAffectedPlayer );
     }
+
+    public static boolean deathMessageOnlyBroadcastToAffectedPlayer = false;
+    private static void deathSettings() {
+        deathMessageOnlyBroadcastToAffectedPlayer = getBoolean("settings.death.only-broadcast-to-affected-player", deathMessageOnlyBroadcastToAffectedPlayer);
+    }
 }
