From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Encode42 <me@encode42.dev>
Date: Mon, 28 Jun 2021 15:32:57 -0400
Subject: [PATCH] Config to allow armor hotswapping

Not working on swap:
- Equip sound

diff --git a/src/main/java/net/minecraft/world/item/ArmorItem.java b/src/main/java/net/minecraft/world/item/ArmorItem.java
index ad095c6bc7e00d286c83e37f84d8ed5d4a2c6881..aac5b2e26f132c1c7a9e7cce8616aba78a0501c5 100644
--- a/src/main/java/net/minecraft/world/item/ArmorItem.java
+++ b/src/main/java/net/minecraft/world/item/ArmorItem.java
@@ -139,14 +139,17 @@ public class ArmorItem extends Item implements Wearable {
         EquipmentSlot enumitemslot = Mob.getEquipmentSlotForItem(itemstack);
         ItemStack itemstack1 = user.getItemBySlot(enumitemslot);
 
-        if (itemstack1.isEmpty()) {
+        // Purpur start
+        boolean shouldSwap = world.purpurConfig.armorSwapping && !itemstack1.isEmpty();
+        if (shouldSwap || itemstack1.isEmpty()) {
             user.setItemSlot(enumitemslot, itemstack.copy());
             if (!world.isClientSide()) {
                 user.awardStat(Stats.ITEM_USED.get(this));
             }
 
             itemstack.setCount(0);
-            return InteractionResultHolder.sidedSuccess(itemstack, world.isClientSide());
+            return InteractionResultHolder.sidedSuccess(shouldSwap ? itemstack1 : itemstack, true);
+            // Purpur end
         } else {
             return InteractionResultHolder.fail(itemstack);
         }
diff --git a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
index 273ef02ba9f3bc5aea17ef082cf87005005d0927..6316a199deee3168b577883d2bea82b841ae651a 100644
--- a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
+++ b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
@@ -2270,4 +2270,9 @@ public class PurpurWorldConfig {
         }
         randomPopulatorSeed = getBoolean("seed.random-populator-seed", randomPopulatorSeed);
     }
+
+    public boolean armorSwapping = false;
+    private void armorSettings() {
+        armorSwapping = getBoolean("armor.hot-swapping", armorSwapping);
+    }
 }
