From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ben Kerllenevich <me@notom3ga.me>
Date: Fri, 28 May 2021 05:51:52 -0400
Subject: [PATCH] Allow disabling offline mode message on proxy


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 7773f15f4eabb73b16ce971971f0c6c7625998ed..4bd6412a2302fcc2f6bfc6757ba492266c2ae740 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -282,7 +282,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
         server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.STARTUP);
         // CraftBukkit end
 
-        if (!this.getOnlineMode()) {
+        if (!this.getOnlineMode() && !(com.destroystokyo.paper.PaperConfig.isProxyOnlineMode() && net.pl3x.purpur.PurpurConfig.loggerSuppressOfflineModeWithProxy)) { // Purpur
             DedicatedServer.LOGGER.warn("**** SERVER IS RUNNING IN OFFLINE/INSECURE MODE!");
             DedicatedServer.LOGGER.warn("The server will make no attempt to authenticate usernames. Beware.");
             // Spigot start
diff --git a/src/main/java/net/pl3x/purpur/PurpurConfig.java b/src/main/java/net/pl3x/purpur/PurpurConfig.java
index ce9cbc3964ef626e7fe3baf59225114ca93b0b53..2be292df2d6faefb746bf96ebd1167b3d6bffd67 100644
--- a/src/main/java/net/pl3x/purpur/PurpurConfig.java
+++ b/src/main/java/net/pl3x/purpur/PurpurConfig.java
@@ -284,9 +284,11 @@ public class PurpurConfig {
 
     public static boolean loggerSuppressInitLegacyMaterialError = false;
     public static boolean loggerSuppressIgnoredAdvancementWarnings = false;
+    public static boolean loggerSuppressOfflineModeWithProxy = false;
     private static void loggerSettings() {
         loggerSuppressInitLegacyMaterialError = getBoolean("settings.logger.suppress-init-legacy-material-errors", loggerSuppressInitLegacyMaterialError);
         loggerSuppressIgnoredAdvancementWarnings = getBoolean("settings.logger.suppress-ignored-advancement-warnings", loggerSuppressIgnoredAdvancementWarnings);
+        loggerSuppressOfflineModeWithProxy = getBoolean("settings.logger.suppress-offline-mode-warning-with-proxy", loggerSuppressOfflineModeWithProxy);
     }
 
     public static boolean tpsCatchup = true;
