From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TreyRuffy <TreyRuffy@users.noreply.github.com>
Date: Wed, 9 Jun 2021 16:31:14 -0600
Subject: [PATCH] Flying Fall Damage API


diff --git a/src/main/java/net/minecraft/world/entity/player/EntityHuman.java b/src/main/java/net/minecraft/world/entity/player/EntityHuman.java
index 383b4cacc8c9f4fd5d2a83f683e5174ea5025f61..be7c38c54311d6ea99a4ce2eec63863b649de3b2 100644
--- a/src/main/java/net/minecraft/world/entity/player/EntityHuman.java
+++ b/src/main/java/net/minecraft/world/entity/player/EntityHuman.java
@@ -169,7 +169,10 @@ public abstract class EntityHuman extends EntityLiving {
     // Paper start
     public boolean affectsSpawning = true;
     // Paper end
-    public boolean isCritical = false; // Purpur
+    // Purpur start
+    public boolean flyingFallDamage = false;
+    public boolean isCritical = false;
+    // Purpur end
 
     // CraftBukkit start
     public boolean fauxSleeping;
@@ -1651,7 +1654,7 @@ public abstract class EntityHuman extends EntityLiving {
 
     @Override
     public boolean b(float f, float f1) {
-        if (this.abilities.canFly) {
+        if (this.abilities.canFly && !flyingFallDamage) { // Purpur - allow fall damage with can fly attribute
             return false;
         } else {
             if (f >= 2.0F) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index a047ab4830639578f21c6400141adde8c14b8564..ae147879392d828f9692f78d1f8a9a2e7307d8ff 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2542,5 +2542,15 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     public void setSpawnInvulnerableTicks(int invulnerableTicks) {
         getHandle().invulnerableTicks = invulnerableTicks;
     }
+
+    @Override
+    public void setFlyingFallDamage(boolean flyingFallDamage) {
+        getHandle().flyingFallDamage = flyingFallDamage;
+    }
+
+    @Override
+    public boolean hasFlyingFallDamage() {
+        return getHandle().flyingFallDamage;
+    }
     // Purpur end
 }
