From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Your Name <you@example.com>
Date: Tue, 21 Dec 2021 18:12:26 +0100
Subject: [PATCH] Custom IP forwarding warning message


diff --git a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
index 150050fb343ef6119204b7d5220207765a6937bc..f85f66786614f4f65f5933f0a7a0bd374823927c 100644
--- a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
@@ -130,7 +130,7 @@ public class ServerHandshakePacketListenerImpl implements ServerHandshakePacketL
                             connection.spoofedUUID = com.mojang.util.UUIDTypeAdapter.fromString( split[2] );
                         } else
                         {
-                            TranslatableComponent chatmessage = new TranslatableComponent("If you wish to use IP forwarding, please enable it in your BungeeCord config as well!");
+                            TranslatableComponent chatmessage = new TranslatableComponent("§e§lDragon§6§lCraft §e» §cWejdź przez IP: §eDCRFT.PL");
                             this.connection.send(new ClientboundLoginDisconnectPacket(chatmessage));
                             this.connection.disconnect(chatmessage);
                             return;
