From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Oharass <90105244+Oharass@users.noreply.github.com>
Date: Sat, 4 Sep 2021 16:26:31 +0000
Subject: [PATCH] configurable damage for aggressive cows


diff --git a/src/main/java/net/minecraft/world/entity/animal/Cow.java b/src/main/java/net/minecraft/world/entity/animal/Cow.java
index 6855c409f1c793ec436f447838008853b9b249d2..0f0522a7ee308e0527d4a85d41378b1ae014ec5d 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Cow.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Cow.java
@@ -57,7 +57,7 @@ public class Cow extends Animal {
     @Override
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level.purpurConfig.cowMaxHealth);
-        this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(2.0D); // Purpur
+        this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(level.purpurConfig.cowNaturallyAggressiveToPlayersDamage); // Purpur
     }
 
     @Override
diff --git a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
index 99304124b8055a6d15375922b2b47c2bfd45383c..4620810bc5025557e30e7c5fcab0a1553f306fd2 100644
--- a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
+++ b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
@@ -1152,6 +1152,7 @@ public class PurpurWorldConfig {
     public int cowFeedMushrooms = 0;
     public int cowBreedingTicks = 6000;
     public double cowNaturallyAggressiveToPlayersChance = 0.0D;
+    public double cowNaturallyAggressiveToPlayersDamage = 2.0D;
     private void cowSettings() {
         cowRidable = getBoolean("mobs.cow.ridable", cowRidable);
         cowRidableInWater = getBoolean("mobs.cow.ridable-in-water", cowRidableInWater);
@@ -1164,6 +1165,7 @@ public class PurpurWorldConfig {
         cowFeedMushrooms = getInt("mobs.cow.feed-mushrooms-for-mooshroom", cowFeedMushrooms);
         cowBreedingTicks = getInt("mobs.cow.breeding-delay-ticks", cowBreedingTicks);
         cowNaturallyAggressiveToPlayersChance = getDouble("mobs.cow.naturally-aggressive-to-players-chance", cowNaturallyAggressiveToPlayersChance);
+	  cowNaturallyAggressiveToPlayersDamage = getDouble("mobs.cow.naturally-aggressive-to-players-damage", cowNaturallyAggressiveToPlayersDamage);
     }
 
     public boolean creeperRidable = false;
