From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Encode42 <me@encode42.dev>
Date: Wed, 5 Jan 2022 14:32:05 -0500
Subject: [PATCH] Temporary build fix


diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 6e57000706fa51cbde64a6c63d0da39e04b51826..638e59af11d6a0609a866ecf198530038e987804 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -210,7 +210,7 @@ public abstract class Mob extends LivingEntity {
         return this.lookControl;
     }
 
-    int tickCount = 0; // Pufferfish - throttle inactive goal selector ticking
+    public int tickCount = 0; // Pufferfish - throttle inactive goal selector ticking // Purpur
     // Paper start
     @Override
     public void inactiveTick() {
