From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MiniDigger <admin@minidigger.me>
Date: Sat, 25 Sep 2021 21:34:34 -0500
Subject: [PATCH] add player packet event

Co-authored by: oharass <oharass@bk.ru>


diff --git a/src/main/java/net/minecraft/network/protocol/PacketUtils.java b/src/main/java/net/minecraft/network/protocol/PacketUtils.java
index acc12307f61e1e055896b68fe16654c9c4a606a0..3311833f31b0bf911faf32e7b2fd6fc63325656e 100644
--- a/src/main/java/net/minecraft/network/protocol/PacketUtils.java
+++ b/src/main/java/net/minecraft/network/protocol/PacketUtils.java
@@ -53,6 +53,7 @@ public class PacketUtils {
                 if (MinecraftServer.getServer().hasStopped() || (listener instanceof ServerGamePacketListenerImpl && ((ServerGamePacketListenerImpl) listener).processedDisconnect)) return; // CraftBukkit, MC-142590
                 if (listener.getConnection().isConnected()) {
                     try (Timing ignored = timing.startTiming()) { // Paper - timings
+                    if (!createEvent(packet, listener)) return; // Purpur
                     packet.handle(listener);
                     } // Paper - timings
                     // Paper start
@@ -87,5 +88,16 @@ public class PacketUtils {
             throw RunningOnDifferentThreadException.RUNNING_ON_DIFFERENT_THREAD;
             // CraftBukkit end
         }
+        if (!createEvent(packet, listener)) throw RunningOnDifferentThreadException.RUNNING_ON_DIFFERENT_THREAD; // Purpur
     }
+    // Purpur start
+    private static <T extends PacketListener> boolean createEvent(Packet<T> packet, T listener) {
+        if (listener instanceof ServerGamePacketListenerImpl connection) {
+            Integer id = net.minecraft.network.ConnectionProtocol.PLAY.getPacketId(PacketFlow.SERVERBOUND, packet);
+            if (id == null) return false;
+            return new net.pl3x.purpur.event.player.PlayerPacketSendEvent(connection.getCraftPlayer(), id, packet).callEvent();
+        }
+        return true;
+    }
+    // Purpur end
 }
