From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ben Kerllenevich <ben@omega24.dev>
Date: Wed, 13 Jul 2022 16:27:43 -0400
Subject: [PATCH] Send client custom name of BE


diff --git a/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
index 59d616137088af46d4494171fe96ba0129082496..efc1bdcaf141a2c2152173439d707f50feaaf91a 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -206,10 +206,24 @@ public abstract class BlockEntity {
 
     @Nullable
     public Packet<ClientGamePacketListener> getUpdatePacket() {
+        // Purpur start
+        if (this instanceof net.minecraft.world.Nameable nameable && nameable.hasCustomName()) {
+            CompoundTag nbt = this.saveWithoutMetadata();
+            nbt.remove("Items");
+            return net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket.create(this, $ -> nbt);
+        }
+        // Purpur end
         return null;
     }
 
     public CompoundTag getUpdateTag() {
+        // Purpur start
+        if (this instanceof net.minecraft.world.Nameable nameable && nameable.hasCustomName()) {
+            CompoundTag nbt = this.saveWithoutMetadata();
+            nbt.remove("Items");
+            return nbt;
+        }
+        // Purpur end
         return new CompoundTag();
     }
 
