From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Sat, 11 Jun 2022 08:18:40 -0500
Subject: [PATCH] Add log suppression for sent expired chat


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 05bda808496f38d10416c3702b1f97fe18247769..99c8b4957194868c36f9ea214c7f77a23ce8002a 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2248,7 +2248,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
             }); // Paper - push to main
             return false;
         } else {
-            if (this.isChatExpired(instant)) {
+            if (!org.purpurmc.purpur.PurpurConfig.loggerSuppressSentExpiredChat && this.isChatExpired(instant)) { // Purpur
                 ServerGamePacketListenerImpl.LOGGER.warn("{} sent expired chat: '{}'. Is the client/server system time unsynchronized?", this.player.getName().getString(), s);
             }
 
diff --git a/src/main/java/org/purpurmc/purpur/PurpurConfig.java b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
index 2340a0ce689a5b83b3b1ed3726e6e14422c7db5e..4bebe841f1a031121569b20f4a39ef59b2e63e9e 100644
--- a/src/main/java/org/purpurmc/purpur/PurpurConfig.java
+++ b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
@@ -431,11 +431,13 @@ public class PurpurConfig {
     public static boolean loggerSuppressIgnoredAdvancementWarnings = false;
     public static boolean loggerSuppressUnrecognizedRecipeErrors = false;
     public static boolean loggerSuppressSetBlockFarChunk = false;
+    public static boolean loggerSuppressSentExpiredChat = false;
     private static void loggerSettings() {
         loggerSuppressInitLegacyMaterialError = getBoolean("settings.logger.suppress-init-legacy-material-errors", loggerSuppressInitLegacyMaterialError);
         loggerSuppressIgnoredAdvancementWarnings = getBoolean("settings.logger.suppress-ignored-advancement-warnings", loggerSuppressIgnoredAdvancementWarnings);
         loggerSuppressUnrecognizedRecipeErrors = getBoolean("settings.logger.suppress-unrecognized-recipe-errors", loggerSuppressUnrecognizedRecipeErrors);
         loggerSuppressSetBlockFarChunk = getBoolean("settings.logger.suppress-setblock-in-far-chunk-errors", loggerSuppressSetBlockFarChunk);
+        loggerSuppressSentExpiredChat = getBoolean("settings.logger.suppress-sent-expired-chat", loggerSuppressSentExpiredChat);
     }
 
     public static boolean tpsCatchup = true;
