From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Sun, 3 Jul 2022 04:13:57 -0500
Subject: [PATCH] Option to disable kick for out of order chat


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index a7d1bd88ba6b36c2a80edad2e38471d72cba0451..91dadd78069d5e2a850d1aa6dd59149c3a527c2a 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2268,7 +2268,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
         do {
             instant1 = (Instant) this.lastChatTimeStamp.get();
             if (timestamp.isBefore(instant1)) {
-                return false;
+                return !org.purpurmc.purpur.PurpurConfig.kickForOutOfOrderChat; // Purpur
             }
         } while (!this.lastChatTimeStamp.compareAndSet(instant1, timestamp));
 
diff --git a/src/main/java/org/purpurmc/purpur/PurpurConfig.java b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
index 22b744c0b047e1fa071e08f7a030199b4d7544f0..3bed3b7d7f35833b4d2d58ffb6900c707deb7c9b 100644
--- a/src/main/java/org/purpurmc/purpur/PurpurConfig.java
+++ b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
@@ -433,9 +433,11 @@ public class PurpurConfig {
 
     public static boolean useUPnP = false;
     public static boolean maxJoinsPerSecond = false;
+    public static boolean kickForOutOfOrderChat = true;
     private static void networkSettings() {
         useUPnP = getBoolean("settings.network.upnp-port-forwarding", useUPnP);
         maxJoinsPerSecond = getBoolean("settings.network.max-joins-per-second", maxJoinsPerSecond);
+        kickForOutOfOrderChat = getBoolean("settings.network.kick-for-out-of-order-chat", kickForOutOfOrderChat);
     }
 
     public static java.util.regex.Pattern usernameValidCharactersPattern;
