From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MelnCat <melncatuwu@gmail.com>
Date: Fri, 23 Sep 2022 18:41:05 -0700
Subject: [PATCH] Add death screen API


diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerCombatKillPacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerCombatKillPacket.java
index 53b75f5737a910ffc5448cd9a85eae57f9c1488f..ea95873dd034779e56a8b924cd27f9375be05daf 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerCombatKillPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerCombatKillPacket.java
@@ -9,6 +9,7 @@ public class ClientboundPlayerCombatKillPacket implements Packet<ClientGamePacke
     private final int playerId;
     private final int killerId;
     private final Component message;
+    public net.kyori.adventure.text.Component adventure$message; // Purpur
 
     public ClientboundPlayerCombatKillPacket(CombatTracker damageTracker, Component message) {
         this(damageTracker.getMob().getId(), damageTracker.getKillerId(), message);
@@ -30,6 +31,12 @@ public class ClientboundPlayerCombatKillPacket implements Packet<ClientGamePacke
     public void write(FriendlyByteBuf buf) {
         buf.writeVarInt(this.playerId);
         buf.writeInt(this.killerId);
+        // Purpur start
+        if (this.adventure$message != null) {
+            buf.writeComponent(this.adventure$message);
+            return;
+        }
+        // Purpur end
         buf.writeComponent(this.message);
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index d27d31233a937f4e2e6e7486594f665b107f12c6..264660979ead8b523248f0828591531105b8d143 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3277,5 +3277,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         if (this.getHandle().connection == null) return;
         this.getHandle().connection.send(new net.minecraft.network.protocol.game.ClientboundCustomPayloadPacket(ClientboundCustomPayloadPacket.DEBUG_GAME_TEST_CLEAR, new FriendlyByteBuf(io.netty.buffer.Unpooled.buffer())));
     }
+
+    @Override
+    public void sendDeathScreen(net.kyori.adventure.text.Component message) {
+        sendDeathScreen(message, null);
+    }
+
+    @Override
+    public void sendDeathScreen(net.kyori.adventure.text.Component message, org.bukkit.entity.Entity killer) {
+        if (this.getHandle().connection == null) return;
+        net.minecraft.network.protocol.game.ClientboundPlayerCombatKillPacket packet = new net.minecraft.network.protocol.game.ClientboundPlayerCombatKillPacket(getEntityId(), killer == null ? -1 : killer.getEntityId(), null);
+        packet.adventure$message = message;
+        this.getHandle().connection.send(packet);
+    }
     // Purpur end
 }
