From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: granny <granny@purpurmc.org>
Date: Fri, 26 Aug 2022 21:45:00 -0700
Subject: [PATCH] add Player#showCreditScreen


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index ef8c5b407072c16ec57cc92e14e5b2b2cec2c09c..b7fab4e46c5125359edb66990f31f94715601eb8 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2607,6 +2607,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         this.getHandle().connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.DEMO_EVENT, ClientboundGameEventPacket.DEMO_PARAM_INTRO));
     }
 
+    @Override
+    public void showCreditScreen() {
+        if (this.getHandle().connection == null) return;
+
+        this.getHandle().connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.WIN_GAME, 1F));
+    }
+
     @Override
     public boolean isAllowingServerListings() {
         return this.getHandle().allowsListing();
