From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: granny <granny@purpurmc.org>
Date: Fri, 26 Aug 2022 21:45:00 -0700
Subject: [PATCH] add Player#showCreditScreen


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 61cc3f41a8ef8d3cfe15fe6e92fa9b4977a26c24..1e9e73762bea7c80ec732e9b7977b6908917c044 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2606,6 +2606,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         this.getHandle().connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.DEMO_EVENT, ClientboundGameEventPacket.DEMO_PARAM_INTRO));
     }
 
+    @Override
+    public void showCreditScreen() {
+        if (this.getHandle().connection == null) return;
+
+        this.getHandle().connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.WIN_GAME, 1F));
+    }
+
     @Override
     public boolean isAllowingServerListings() {
         return this.getHandle().allowsListing();
