From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Mon, 11 Jul 2022 20:44:19 -0500
Subject: [PATCH] Chat Preview API


diff --git a/src/main/java/net/minecraft/network/chat/ChatDecorator.java b/src/main/java/net/minecraft/network/chat/ChatDecorator.java
index b9cdd89a3871d934a0449ed70766c2e9d6369ab9..cf415137057a309db0961dcc993cecf33203151e 100644
--- a/src/main/java/net/minecraft/network/chat/ChatDecorator.java
+++ b/src/main/java/net/minecraft/network/chat/ChatDecorator.java
@@ -7,6 +7,18 @@ import net.minecraft.server.level.ServerPlayer;
 @FunctionalInterface
 public interface ChatDecorator {
     ChatDecorator PLAIN = (sender, message) -> {
+        // Purpur start
+        if (sender != null) {
+            net.kyori.adventure.text.Component adventureComponent = io.papermc.paper.adventure.PaperAdventure.asAdventure(message);
+            org.purpurmc.purpur.event.player.PlayerPreviewChatEvent previewEvent = new org.purpurmc.purpur.event.player.PlayerPreviewChatEvent(sender.getBukkitEntity(), "", adventureComponent);
+            if (!previewEvent.callEvent()) {
+                return CompletableFuture.completedFuture(null);
+            }
+            if (!previewEvent.getQuery().equals(adventureComponent)) {
+                return CompletableFuture.completedFuture(io.papermc.paper.adventure.PaperAdventure.asVanilla(previewEvent.getQuery()));
+            }
+        }
+        // Purpur end
         return CompletableFuture.completedFuture(message);
     };
 
