From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Mon, 11 Jul 2022 20:44:19 -0500
Subject: [PATCH] Chat Preview API


diff --git a/src/main/java/net/minecraft/network/chat/ChatDecorator.java b/src/main/java/net/minecraft/network/chat/ChatDecorator.java
index d07bb2d1f365cb7c92a1252b83210fa38cbeeb1d..d307fd9656c2067eb70c5031dd250dc8aabfd663 100644
--- a/src/main/java/net/minecraft/network/chat/ChatDecorator.java
+++ b/src/main/java/net/minecraft/network/chat/ChatDecorator.java
@@ -8,6 +8,18 @@ import net.minecraft.server.network.FilteredText;
 @FunctionalInterface
 public interface ChatDecorator {
     ChatDecorator PLAIN = (sender, message) -> {
+        // Purpur start
+        if (sender != null) {
+            net.kyori.adventure.text.Component adventureComponent = io.papermc.paper.adventure.PaperAdventure.asAdventure(message);
+            org.purpurmc.purpur.event.player.PlayerPreviewChatEvent previewEvent = new org.purpurmc.purpur.event.player.PlayerPreviewChatEvent(sender.getBukkitEntity(), "", adventureComponent);
+            if (!previewEvent.callEvent()) {
+                return CompletableFuture.completedFuture(null);
+            }
+            if (!previewEvent.getQuery().equals(adventureComponent)) {
+                return CompletableFuture.completedFuture(io.papermc.paper.adventure.PaperAdventure.asVanilla(previewEvent.getQuery()));
+            }
+        }
+        // Purpur end
         return CompletableFuture.completedFuture(message);
     };
 
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 3a4c6c91222ad6c60d1263ca7f912b1e3a85c164..ac527a253e2c238099cb3755fe5ab4855289d6f5 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2512,7 +2512,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
 
     @Override
     public void handleChatPreview(ServerboundChatPreviewPacket packet) {
-        if (false && this.server.previewsChat()) { // CraftBukkit - preview NYI
+        if (this.server.previewsChat()) { // CraftBukkit - preview NYI // Purpur - lets implement it already.. sheesh
             this.chatPreviewThrottler.schedule(() -> {
                 int i = packet.queryId();
                 String s = packet.query();
