From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Tue, 20 Sep 2022 17:56:21 -0500
Subject: [PATCH] Allay respect item NBT


diff --git a/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java b/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java
index c03eca3f9e5ca87e7607aeca86ff62638d294ff9..a3d34e100647853f887d6ee8a5cf87a3f0d051a5 100644
--- a/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java
@@ -412,9 +412,31 @@ public class Allay extends PathfinderMob implements InventoryCarrier {
 
     @Override
     public boolean wantsToPickUp(ItemStack stack) {
-        ItemStack itemstack1 = this.getItemInHand(InteractionHand.MAIN_HAND);
-
-        return !itemstack1.isEmpty() && this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) && this.inventory.canAddItem(stack) && this.allayConsidersItemEqual(itemstack1, stack);
+        // Purpur start
+        if (!this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+            return false;
+        }
+        ItemStack itemStack = this.getItemInHand(InteractionHand.MAIN_HAND);
+        if (itemStack.isEmpty()) {
+            return false;
+        }
+        if (!allayConsidersItemEqual(itemStack, stack)) {
+            return false;
+        }
+        if (!this.inventory.canAddItem(stack)) {
+            return false;
+        }
+        for (String tag : this.level.purpurConfig.allayRespectNBT) {
+            if (stack.hasTag() && itemStack.hasTag()) {
+                Tag tag1 = stack.getTag().get(tag);
+                Tag tag2 = itemStack.getTag().get(tag);
+                if (!Objects.equals(tag1, tag2)) {
+                    return false;
+                }
+            }
+        }
+        return true;
+        // Purpur end
     }
 
     private boolean allayConsidersItemEqual(ItemStack stack, ItemStack stack2) {
diff --git a/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java b/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
index 7e20ab899271be23ed90b7d3c3b8f79d07076716..6fcbdb4b6567196f99fddaa70b9b1f393edf4ae6 100644
--- a/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
+++ b/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
@@ -1061,10 +1061,13 @@ public class PurpurWorldConfig {
     public boolean allayRidable = false;
     public boolean allayRidableInWater = false;
     public boolean allayControllable = true;
+    public List<String> allayRespectNBT = new ArrayList<>();
     private void allaySettings() {
         allayRidable = getBoolean("mobs.allay.ridable", allayRidable);
         allayRidableInWater = getBoolean("mobs.allay.ridable-in-water", allayRidableInWater);
         allayControllable = getBoolean("mobs.allay.controllable", allayControllable);
+        allayRespectNBT.clear();
+        getList("mobs.allay.respect-nbt", new ArrayList<>()).forEach(key -> allayRespectNBT.add(key.toString()));
     }
 
     public boolean axolotlRidable = false;
