From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: granny <granny@purpurmc.org>
Date: Fri, 26 Aug 2022 21:45:00 -0700
Subject: [PATCH] add Player#showCreditScreen


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 897597dbd38b7433c9e414ed71269d48cf2d2c41..a432feb58be5947688502e7b4e3e81cf09eb3f47 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2654,6 +2654,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         this.getHandle().connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.DEMO_EVENT, ClientboundGameEventPacket.DEMO_PARAM_INTRO));
     }
 
+    @Override
+    public void showCreditScreen() {
+        if (this.getHandle().connection == null) return;
+
+        this.getHandle().connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.WIN_GAME, 1F));
+    }
+
     @Override
     public boolean isAllowingServerListings() {
         return this.getHandle().allowsListing();
