From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: rafaelflromao <12960698+rafaelflromao@users.noreply.github.com>
Date: Mon, 8 May 2023 20:43:29 +0100
Subject: [PATCH] Add mending multiplier


diff --git a/src/main/java/net/minecraft/world/entity/ExperienceOrb.java b/src/main/java/net/minecraft/world/entity/ExperienceOrb.java
index a3ef8e6a698df4c78e0494da48063ddd29caf62f..1b8f5680c4122c7b6b19e2bc1ed6295caa024622 100644
--- a/src/main/java/net/minecraft/world/entity/ExperienceOrb.java
+++ b/src/main/java/net/minecraft/world/entity/ExperienceOrb.java
@@ -359,13 +359,15 @@ public class ExperienceOrb extends Entity {
         }
     }
 
+    // Purpur start
     public int durabilityToXp(int repairAmount) {
-        return repairAmount / 2;
+        return (int) (repairAmount / (2 * level().purpurConfig.mendingMultiplier));
     }
 
     public int xpToDurability(int experienceAmount) {
-        return experienceAmount * 2;
+        return (int) ((experienceAmount * 2) * level().purpurConfig.mendingMultiplier);
     }
+    // Purpur end
 
     public int getValue() {
         return this.value;
diff --git a/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java b/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
index db47c443a765a87d2cabde2d3956909661d65237..93f350d5059e9385f4ba344cd24be8ca4d3198e9 100644
--- a/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
+++ b/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
@@ -119,6 +119,7 @@ public class PurpurWorldConfig {
     }
 
     public boolean useBetterMending = false;
+    public double mendingMultiplier = 1.0;
     public boolean alwaysTameInCreative = false;
     public boolean boatEjectPlayersOnLand = false;
     public boolean boatsDoFallDamage = false;
@@ -146,6 +147,7 @@ public class PurpurWorldConfig {
     public int mobLastHurtByPlayerTime = 100;
     private void miscGameplayMechanicsSettings() {
         useBetterMending = getBoolean("gameplay-mechanics.use-better-mending", useBetterMending);
+        mendingMultiplier = getDouble("gameplay-mechanics.mending-multiplier", mendingMultiplier);
         alwaysTameInCreative = getBoolean("gameplay-mechanics.always-tame-in-creative", alwaysTameInCreative);
         boatEjectPlayersOnLand = getBoolean("gameplay-mechanics.boat.eject-players-on-land", boatEjectPlayersOnLand);
         boatsDoFallDamage = getBoolean("gameplay-mechanics.boat.do-fall-damage", boatsDoFallDamage);
