From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Wed, 28 Sep 2022 13:51:41 -0500
Subject: [PATCH] Use simple class names in Paper loggers


diff --git a/src/main/java/com/mojang/logging/LogUtils.java b/src/main/java/com/mojang/logging/LogUtils.java
index 46cab7a8c7b87ab01b26074b04f5a02b3907cfc4..f7449f6ec04356175a0349001a6daa27ab5873f1 100644
--- a/src/main/java/com/mojang/logging/LogUtils.java
+++ b/src/main/java/com/mojang/logging/LogUtils.java
@@ -59,6 +59,12 @@ public class LogUtils {
      * Caller sensitive, DO NOT WRAP
      */
     public static Logger getLogger() {
-        return LoggerFactory.getLogger(STACK_WALKER.getCallerClass());
+        // Purpur start
+        Class<?> clazz = STACK_WALKER.getCallerClass();
+        if (clazz.getPackageName().startsWith("io.papermc.paper")) {
+            return LoggerFactory.getLogger(clazz.getSimpleName());
+        }
+        return LoggerFactory.getLogger(clazz);
+        // Purpur end
     }
 }
