From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Fri, 30 Sep 2022 16:56:07 -0500
Subject: [PATCH] Fix MC-123848

Fix for sounds firing twice

diff --git a/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java b/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
index 2a3ec299283c2cc039877eade6e2ddee8f3812ea..420c13e00ca3e22a101d3867f99fdf2dd30776a8 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -288,6 +288,13 @@ public class ItemFrame extends HangingEntity {
         }
     }
 
+    // Purpur start
+    @Nullable
+    public net.minecraft.world.entity.item.ItemEntity spawnAtLocation(ItemStack stack) {
+        return this.spawnAtLocation(stack, getDirection().equals(Direction.DOWN) ? -0.6F : 0.0F);
+    }
+    // Purpur end
+
     private void removeFramedMap(ItemStack itemstack) {
         this.getFramedMapId().ifPresent((i) -> {
             MapItemSavedData worldmap = MapItem.getSavedData(i, this.level);
diff --git a/src/main/java/net/minecraft/world/item/AxeItem.java b/src/main/java/net/minecraft/world/item/AxeItem.java
index 22c8ef3d8d20c14d3b032d054827e074a4422ad6..3b71f1e73caf14531e082b7ff541f3ab068828f9 100644
--- a/src/main/java/net/minecraft/world/item/AxeItem.java
+++ b/src/main/java/net/minecraft/world/item/AxeItem.java
@@ -34,6 +34,7 @@ public class AxeItem extends DiggerItem {
         Player player = context.getPlayer();
         BlockState blockState = level.getBlockState(blockPos);
         // Purpur start
+        Block clickedBlock = level.getBlockState(blockPos).getBlock();
         Optional<org.purpurmc.purpur.tool.Actionable> optional = Optional.ofNullable(level.purpurConfig.axeStrippables.get(blockState.getBlock()));
         Optional<org.purpurmc.purpur.tool.Actionable> optional2 = Optional.ofNullable(level.purpurConfig.axeWeatherables.get(blockState.getBlock()));
         Optional<org.purpurmc.purpur.tool.Actionable> optional3 = Optional.ofNullable(level.purpurConfig.axeWaxables.get(blockState.getBlock()));
@@ -41,14 +42,20 @@ public class AxeItem extends DiggerItem {
         ItemStack itemStack = context.getItemInHand();
         Optional<org.purpurmc.purpur.tool.Actionable> optional4 = Optional.empty(); // Purpur
         if (optional.isPresent()) {
-            level.playSound(null, blockPos, SoundEvents.AXE_STRIP, SoundSource.BLOCKS, 1.0F, 1.0F); // Purpur - force sound
+            if(!STRIPPABLES.containsKey(clickedBlock)) {
+                level.playSound(null, blockPos, SoundEvents.AXE_STRIP, SoundSource.BLOCKS, 1.0F, 1.0F); // Purpur - force sound
+            }
             optional4 = optional;
         } else if (optional2.isPresent()) {
-            level.playSound(null, blockPos, SoundEvents.AXE_SCRAPE, SoundSource.BLOCKS, 1.0F, 1.0F); // Purpur - force sound
+            if(!HoneycombItem.WAXABLES.get().containsKey(clickedBlock)) {
+                level.playSound(null, blockPos, SoundEvents.AXE_SCRAPE, SoundSource.BLOCKS, 1.0F, 1.0F); // Purpur - force sound
+            }
             level.levelEvent(player, 3005, blockPos, 0);
             optional4 = optional2;
         } else if (optional3.isPresent()) {
-            level.playSound(null, blockPos, SoundEvents.AXE_WAX_OFF, SoundSource.BLOCKS, 1.0F, 1.0F); // Purpur - force sound
+            if(!HoneycombItem.WAX_OFF_BY_BLOCK.get().containsKey(clickedBlock)) {
+                level.playSound(null, blockPos, SoundEvents.AXE_WAX_OFF, SoundSource.BLOCKS, 1.0F, 1.0F); // Purpur - force sound
+            }
             level.levelEvent(player, 3004, blockPos, 0);
             optional4 = optional3;
         }
diff --git a/src/main/java/net/minecraft/world/item/HoeItem.java b/src/main/java/net/minecraft/world/item/HoeItem.java
index bf0e1f481b4c81f0663366bbe597b066f95a6200..326a97d7c4e365cf1ddd3c5ce29d5c0e56b58a6d 100644
--- a/src/main/java/net/minecraft/world/item/HoeItem.java
+++ b/src/main/java/net/minecraft/world/item/HoeItem.java
@@ -35,6 +35,7 @@ public class HoeItem extends DiggerItem {
         Level level = context.getLevel();
         BlockPos blockPos = context.getClickedPos();
         // Purpur start
+        Block clickedBlock = level.getBlockState(blockPos).getBlock();
         var tillable = level.purpurConfig.hoeTillables.get(level.getBlockState(blockPos).getBlock());
         if (tillable == null) { return InteractionResult.PASS; } else {
             Predicate<UseOnContext> predicate = tillable.condition().predicate();
@@ -49,7 +50,10 @@ public class HoeItem extends DiggerItem {
             // Purpur end
             if (predicate.test(context)) {
                 Player player = context.getPlayer();
-                level.playSound(null, blockPos, SoundEvents.HOE_TILL, SoundSource.BLOCKS, 1.0F, 1.0F); // Purpur - force sound
+                // Purpur start
+                if(!TILLABLES.containsKey(clickedBlock)) {
+                    level.playSound(null, blockPos, SoundEvents.HOE_TILL, SoundSource.BLOCKS, 1.0F, 1.0F); // Purpur - force sound
+                } // Purpur end
                 if (!level.isClientSide) {
                     consumer.accept(context);
                     if (player != null) {
