From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Sat, 1 Oct 2022 22:43:24 -0500
Subject: [PATCH] Fix MC-252892


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
index bd77301ef6cde21e1ee1966d2ffa418685f4a214..f946772601cf7ab74104611842fffed4ea9a3bb9 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -521,7 +521,7 @@ public class ServerPlayerGameMode {
         }
 
         if (player.getCooldowns().isOnCooldown(stack.getItem())) {
-            cancelledBlock = true;
+            cancelledBlock = stack.getItem() != net.minecraft.world.item.Items.GOAT_HORN; // Purpur
         }
 
         PlayerInteractEvent event = CraftEventFactory.callPlayerInteractEvent(player, Action.RIGHT_CLICK_BLOCK, blockposition, hitResult.getDirection(), stack, cancelledBlock, hand, hitResult.getLocation()); // Paper
