From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rhythmic <mc.ADHDMC@gmail.com>
Date: Thu, 6 Oct 2022 10:41:01 -0700
Subject: [PATCH] Milk keeps beneficial effects


diff --git a/src/main/java/net/minecraft/world/item/MilkBucketItem.java b/src/main/java/net/minecraft/world/item/MilkBucketItem.java
index 56dc04d8875971ee9a5d077a695509af74fe2473..5aaf1405871c2e12f63199e8db8e671ff4dfdd68 100644
--- a/src/main/java/net/minecraft/world/item/MilkBucketItem.java
+++ b/src/main/java/net/minecraft/world/item/MilkBucketItem.java
@@ -5,12 +5,16 @@ import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.stats.Stats;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResultHolder;
+import net.minecraft.world.effect.MobEffect;
 import net.minecraft.world.effect.MobEffectInstance;
 import net.minecraft.world.effect.MobEffects;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.Level;
 
+import java.util.ArrayList;
+import java.util.Collection;
+
 public class MilkBucketItem extends Item {
 
     private static final int DRINK_DURATION = 32;
@@ -34,7 +38,22 @@ public class MilkBucketItem extends Item {
 
         if (!world.isClientSide) {
             MobEffectInstance badOmen = user.getEffect(MobEffects.BAD_OMEN);
-            user.removeAllEffects(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.MILK); // CraftBukkit
+            // Purpur start
+            Collection<MobEffectInstance> activeEffects = user.getActiveEffects();
+            if (!world.purpurConfig.milkClearsBeneficialEffects && !activeEffects.isEmpty()){
+                ArrayList<MobEffect> effectList = new ArrayList<>();
+                for (MobEffectInstance effect : activeEffects){
+                    if(!effect.getEffect().isBeneficial()){
+                        effectList.add(effect.getEffect());
+                    }
+                }
+                for (MobEffect effectType : effectList){
+                    user.removeEffect(effectType);
+                }
+            } else {
+                user.removeAllEffects(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.MILK); // CraftBukkit
+            }
+            // Purpur end
             if (!world.purpurConfig.milkCuresBadOmen && badOmen != null) user.addEffect(badOmen); // Purpur
         }
 
diff --git a/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java b/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
index a6356d19f0fc664be523f19c2ea1704191a7a73a..207691bf5833655efac7b5ceddbe53142d7a895a 100644
--- a/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
+++ b/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
@@ -131,6 +131,7 @@ public class PurpurWorldConfig {
     public boolean fireballsBypassMobGriefing = false;
     public boolean imposeTeleportRestrictionsOnGateways = false;
     public boolean milkCuresBadOmen = true;
+    public boolean milkClearsBeneficialEffects;
     public boolean noteBlockIgnoreAbove = false;
     public boolean persistentDroppableEntityDisplayNames = false;
     public boolean persistentTileEntityDisplayNames = false;
@@ -157,6 +158,7 @@ public class PurpurWorldConfig {
         fireballsBypassMobGriefing = getBoolean("gameplay-mechanics.fireballs-bypass-mob-griefing", fireballsBypassMobGriefing);
         imposeTeleportRestrictionsOnGateways = getBoolean("gameplay-mechanics.impose-teleport-restrictions-on-gateways", imposeTeleportRestrictionsOnGateways);
         milkCuresBadOmen = getBoolean("gameplay-mechanics.milk-cures-bad-omen", milkCuresBadOmen);
+        milkClearsBeneficialEffects = getBoolean("gameplay-mechanics.milk-clears-beneficial-effects", milkClearsBeneficialEffects);
         noteBlockIgnoreAbove = getBoolean("gameplay-mechanics.note-block-ignore-above", noteBlockIgnoreAbove);
         persistentTileEntityDisplayNames = getBoolean("gameplay-mechanics.persistent-tileentity-display-names-and-lore", persistentTileEntityDisplayNames);
         persistentDroppableEntityDisplayNames = getBoolean("gameplay-mechanics.persistent-droppable-entity-display-names", persistentDroppableEntityDisplayNames);
