From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aniket Joshi <arealronin@gmail.com>
Date: Mon, 10 Oct 2022 12:07:10 +0530
Subject: [PATCH] advancements rewards exp toggle


diff --git a/src/main/java/net/minecraft/advancements/AdvancementRewards.java b/src/main/java/net/minecraft/advancements/AdvancementRewards.java
index 04a5e1cbacf67d971ebdc3c266035ba95aa02c57..a9ca73a648688c0e431bf2d0486aa5f02031a6a0 100644
--- a/src/main/java/net/minecraft/advancements/AdvancementRewards.java
+++ b/src/main/java/net/minecraft/advancements/AdvancementRewards.java
@@ -42,7 +42,9 @@ public class AdvancementRewards {
     }
 
     public void grant(ServerPlayer player) {
-        player.giveExperiencePoints(this.experience);
+        if (org.purpurmc.purpur.PurpurConfig.advancementsRewardsExp) {
+            player.giveExperiencePoints(this.experience); //Purpur
+        }
         LootContext lootContext = (new LootContext.Builder(player.getLevel())).withParameter(LootContextParams.THIS_ENTITY, player).withParameter(LootContextParams.ORIGIN, player.position()).withRandom(player.getRandom()).create(LootContextParamSets.ADVANCEMENT_REWARD);
         boolean bl = false;
 
diff --git a/src/main/java/org/purpurmc/purpur/PurpurConfig.java b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
index dd6e3c74e89860d63813f5ca65de623928be9f96..a4440b21b9dab7b608709c213214b5fa2b27bb57 100644
--- a/src/main/java/org/purpurmc/purpur/PurpurConfig.java
+++ b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
@@ -645,4 +645,9 @@ public class PurpurConfig {
             block.fallDistanceMultiplier = fallDistanceMultiplier.floatValue();
         });
     }
+
+    public static boolean advancementsRewardsExp = true;
+    private static void advancementsRewardExp() {
+        advancementsRewardsExp = getBoolean("settings.advancements-rewards-xp", advancementsRewardsExp);
+    }
 }
