From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rhythmic <mc.ADHDMC@gmail.com>
Date: Thu, 13 Oct 2022 20:09:12 -0700
Subject: [PATCH] Add Additional Permissions


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
index 194e9e341b9d1a966d0d78af19120422d7b7eb0a..4816b5bd54ae6e080af4c19df43d57e852e76146 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -607,7 +607,7 @@ public class ServerPlayerGameMode {
 
     // Purpur start
     public boolean shiftClickMended(ItemStack itemstack) {
-        if (this.player.level.purpurConfig.shiftRightClickRepairsMendingPoints > 0 && this.player.isShiftKeyDown()) {
+        if (this.player.level.purpurConfig.shiftRightClickRepairsMendingPoints > 0 && this.player.isShiftKeyDown() && this.player.getBukkitEntity().hasPermission("purpur.mending_shift_click")) {
             int points = Math.min(this.player.totalExperience, this.player.level.purpurConfig.shiftRightClickRepairsMendingPoints);
             if (points > 0 && itemstack.isDamaged() && net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.MENDING, itemstack) > 0) {
                 this.player.giveExperiencePoints(-points);
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index ad4bd5fa9bcd085a1eb99c4da5fc95ebc126d028..1fc877194950ee754e9ffdbe3ff9b80bb316560f 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1574,8 +1574,7 @@ public abstract class LivingEntity extends Entity {
             }
 
             // Purpur start
-            if (level.purpurConfig.totemOfUndyingWorksInInventory && this instanceof ServerPlayer && (itemstack == null || itemstack.getItem() != Items.TOTEM_OF_UNDYING)) {
-                ServerPlayer player = (ServerPlayer) this;
+            if (level.purpurConfig.totemOfUndyingWorksInInventory && this instanceof ServerPlayer player && (itemstack == null || itemstack.getItem() != Items.TOTEM_OF_UNDYING) && player.getBukkitEntity().hasPermission("purpur.inventory_totem")) {
                 for (ItemStack item : player.getInventory().items) {
                     if (item.getItem() == Items.TOTEM_OF_UNDYING) {
                         itemstack1 = item;
diff --git a/src/main/java/net/minecraft/world/inventory/AnvilMenu.java b/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
index 4a3d216006037b308d71f7784e5fef4d6278bde7..df275727dac81a8a94ab73bdb206364539a33d07 100644
--- a/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
@@ -282,16 +282,23 @@ public class AnvilMenu extends ItemCombinerMenu {
                 b1 = 1;
                 i += b1;
                 // Purpur start
-                if (player != null && player.level.purpurConfig.anvilAllowColors && player.getBukkitEntity().hasPermission("purpur.anvil.color")) {
+                org.bukkit.craftbukkit.entity.CraftHumanEntity bukkitPlayer = player.getBukkitEntity();
+                if (player != null && player.level.purpurConfig.anvilAllowColors) {
                     final net.kyori.adventure.text.Component renameTextComponent;
-                    if (player.level.purpurConfig.anvilColorsUseMiniMessage && player.getBukkitEntity().hasPermission("purpur.anvil.minimessage")) {
+                    if (player.level.purpurConfig.anvilColorsUseMiniMessage && bukkitPlayer.hasPermission("purpur.anvil.minimessage")) {
                         renameTextComponent = net.kyori.adventure.text.minimessage.MiniMessage.miniMessage().deserialize(itemName);
-                    } else if (itemName.startsWith("&r") && player.getBukkitEntity().hasPermission("purpur.anvil.remove_italics")) {
+                        itemstack1.setHoverName(io.papermc.paper.adventure.PaperAdventure.asVanilla(renameTextComponent));
+                    } else if (itemName.startsWith("&r") && bukkitPlayer.hasPermission("purpur.anvil.remove_italics")) {
                         renameTextComponent = net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacyAmpersand().deserialize(itemName.substring(2)).decoration(net.kyori.adventure.text.format.TextDecoration.ITALIC, false);
-                    } else {
+                        itemstack1.setHoverName(io.papermc.paper.adventure.PaperAdventure.asVanilla(renameTextComponent));
+                    } else if (bukkitPlayer.hasPermission("purpur.anvil.color")) {
                         renameTextComponent = net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacyAmpersand().deserialize(itemName);
+                        itemstack1.setHoverName(io.papermc.paper.adventure.PaperAdventure.asVanilla(renameTextComponent));
+                    } else if (player.getBukkitEntity().hasPermission("purpur.anvil.format")){
+                        itemName = itemName.replaceAll("(?i)&([l-or])", "\u00a7$1");
+                        renameTextComponent = net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(itemName);
+                        itemstack1.setHoverName(io.papermc.paper.adventure.PaperAdventure.asVanilla(renameTextComponent));
                     }
-                    itemstack1.setHoverName(io.papermc.paper.adventure.PaperAdventure.asVanilla(renameTextComponent));
                 } else
                 // Purpur end
                 itemstack1.setHoverName(Component.literal(this.itemName));
