From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RhythmicSys <mc.adhdmc@gmail.com>
Date: Sat, 14 Jan 2023 18:51:19 -0800
Subject: [PATCH] Option to disable mining slowdown while flying


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index a667401600fc47efa9efa2964ea4099cddc7167a..a575ff6bd89b5d518242086c13c3a9477c7fcbcd 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -113,6 +113,7 @@ import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.scores.PlayerTeam;
 import net.minecraft.world.scores.Scoreboard;
+import org.purpurmc.purpur.PurpurConfig;
 import org.slf4j.Logger;
 import org.bukkit.craftbukkit.entity.CraftHumanEntity;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
@@ -875,7 +876,7 @@ public abstract class Player extends LivingEntity {
             f /= 5.0F;
         }
 
-        if (!this.onGround) {
+        if ((PurpurConfig.airMiningSlowdown && !this.onGround)||(!PurpurConfig.airMiningSlowdown && !this.getBukkitEntity().hasPermission("purpur.flight.mining-slowdown-bypass"))) {
             f /= 5.0F;
         }
 
diff --git a/src/main/java/org/purpurmc/purpur/PurpurConfig.java b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
index c7c0ed8dfe58c841faf684a1fe228eeda6cd57b7..342b00f91fce9eec0a97a8c0a881829132eb1953 100644
--- a/src/main/java/org/purpurmc/purpur/PurpurConfig.java
+++ b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
@@ -565,6 +565,12 @@ public class PurpurConfig {
         clampAttributes = getBoolean("settings.clamp-attributes", clampAttributes);
     }
 
+    public static boolean airMiningSlowdown = true;
+
+    private static void  flyingSettings() {
+        airMiningSlowdown = getBoolean("flight.keep-mining-slowdown", airMiningSlowdown);
+    }
+
     public static boolean limitArmor = true;
     private static void limitArmor() {
         limitArmor = getBoolean("settings.limit-armor", limitArmor);
