From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
index 630677ccc0276882169265e1f7aa63ed0fdd86a3..2e1a18d99b251130ca888eba2102a5d2cff9945d 100644
--- a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
@@ -242,8 +242,8 @@ public class OldMinecartBehavior extends MinecartBehavior {
             Vec3 vec3 = this.getDeltaMovement();
             double speedLength = vec3.horizontalDistance();
             if (speedLength > 0.01) {
-                double speed = 0.06;
-                this.setDeltaMovement(vec3.add(vec3.x / speedLength * 0.06, 0.0, vec3.z / speedLength * 0.06));
+                double speed = level.purpurConfig.poweredRailBoostModifier; // Purpur - Configurable powered rail boost modifier
+                this.setDeltaMovement(vec3.add(vec3.x / speedLength * level.purpurConfig.poweredRailBoostModifier, 0.0, vec3.z / speedLength * level.purpurConfig.poweredRailBoostModifier)); // Purpur - Configurable powered rail boost modifier
             } else {
                 Vec3 deltaMovement = this.getDeltaMovement();
                 double dx = deltaMovement.x;
