From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/material/FlowingFluid.java b/net/minecraft/world/level/material/FlowingFluid.java
index d7270de83fa13618d8f94de6eff63ffa37b823cc..c65d31a38755ecf8aca772bb2ab0fc84d8fc1493 100644
--- a/net/minecraft/world/level/material/FlowingFluid.java
+++ b/net/minecraft/world/level/material/FlowingFluid.java
@@ -235,7 +235,7 @@ public abstract class FlowingFluid extends Fluid {
             }
         }
 
-        if (neighbourSources >= 2 && this.canConvertToSource(level)) {
+        if (neighbourSources >= this.getRequiredSources(level) && this.canConvertToSource(level)) { // Purpur - Implement infinite liquids
             BlockState belowState = level.getBlockState(mutablePos.setWithOffset(pos, Direction.DOWN));
             FluidState belowFluid = belowState.getFluidState();
             if (belowState.isSolid() || this.isSourceBlockOfThisType(belowFluid)) {
@@ -322,6 +322,12 @@ public abstract class FlowingFluid extends Fluid {
 
     protected abstract boolean canConvertToSource(ServerLevel level);
 
+    // Purpur start - Implement infinite liquids
+    protected int getRequiredSources(Level level) {
+        return 2;
+    }
+    // Purpur end - Implement infinite liquids
+
     protected void spreadTo(final LevelAccessor level, final BlockPos pos, final BlockState state, final Direction direction, final FluidState target) {
         if (state.getBlock() instanceof LiquidBlockContainer container) {
             container.placeLiquid(level, pos, state, target);
