--- a/net/minecraft/world/level/material/FlowingFluid.java
+++ b/net/minecraft/world/level/material/FlowingFluid.java
@@ -193,7 +_,7 @@
             }
         }
 
-        if (neighbourSources >= 2 && this.canConvertToSource(level)) {
+        if (neighbourSources >= this.getRequiredSources(level) && this.canConvertToSource(level)) { // Purpur - Implement infinite liquids
             BlockState belowState = level.getBlockState(mutablePos.setWithOffset(pos, Direction.DOWN));
             FluidState belowFluid = belowState.getFluidState();
             if (belowState.isSolid() || this.isSourceBlockOfThisType(belowFluid)) {
@@ -279,6 +_,12 @@
     }
 
     protected abstract boolean canConvertToSource(ServerLevel level);
+
+    // Purpur start - Implement infinite liquids
+    protected int getRequiredSources(Level level) {
+        return 2;
+    }
+    // Purpur end - Implement infinite liquids
 
     protected void spreadTo(final LevelAccessor level, final BlockPos pos, final BlockState state, final Direction direction, final FluidState target) {
         if (state.getBlock() instanceof LiquidBlockContainer container) {
