From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/material/FlowingFluid.java b/net/minecraft/world/level/material/FlowingFluid.java
index e93c900385994d391242def7d022d06e6ef23861..91db225ebf22e801b4c831159f96c08bba490305 100644
--- a/net/minecraft/world/level/material/FlowingFluid.java
+++ b/net/minecraft/world/level/material/FlowingFluid.java
@@ -234,7 +234,7 @@ public abstract class FlowingFluid extends Fluid {
             }
         }
 
-        if (neighbourSources >= 2 && this.canConvertToSource(level)) {
+        if (neighbourSources >= this.getRequiredSources(level) && this.canConvertToSource(level)) { // Purpur - Implement infinite liquids
             BlockState belowState = level.getBlockState(mutablePos.setWithOffset(pos, Direction.DOWN));
             FluidState belowFluid = belowState.getFluidState();
             if (belowState.isSolid() || this.isSourceBlockOfThisType(belowFluid)) {
@@ -321,6 +321,12 @@ public abstract class FlowingFluid extends Fluid {
 
     protected abstract boolean canConvertToSource(ServerLevel level);
 
+    // Purpur start - Implement infinite liquids
+    protected int getRequiredSources(Level level) {
+        return 2;
+    }
+    // Purpur end - Implement infinite liquids
+
     protected void spreadTo(final LevelAccessor level, final BlockPos pos, final BlockState state, final Direction direction, final FluidState target) {
         if (state.getBlock() instanceof LiquidBlockContainer container) {
             container.placeLiquid(level, pos, state, target);
