--- a/net/minecraft/world/level/material/FlowingFluid.java
+++ b/net/minecraft/world/level/material/FlowingFluid.java
@@ -233,7 +_,7 @@
             }
         }
 
-        if (i1 >= 2 && this.canConvertToSource(level)) {
+        if (i1 >= this.getRequiredSources(level) && this.canConvertToSource(level)) { // Purpur - Implement infinite liquids
             BlockState blockState1 = level.getBlockState(mutableBlockPos.setWithOffset(pos, Direction.DOWN));
             FluidState fluidState1 = blockState1.getFluidState();
             if (blockState1.isSolid() || this.isSourceBlockOfThisType(fluidState1)) {
@@ -320,6 +_,12 @@
     }
 
     protected abstract boolean canConvertToSource(ServerLevel level);
+
+    // Purpur start - Implement infinite liquids
+    protected int getRequiredSources(Level level) {
+        return 2;
+    }
+    // Purpur end - Implement infinite liquids
 
     protected void spreadTo(LevelAccessor level, BlockPos pos, BlockState state, Direction direction, FluidState fluidState) {
         if (state.getBlock() instanceof LiquidBlockContainer liquidBlockContainer) {
