From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/material/LavaFluid.java b/net/minecraft/world/level/material/LavaFluid.java
index 76284429b4d9439de04a9a71acac9802ac8d14d2..a99bea7ac0610fbf9427568d57c554fd26f35f9d 100644
--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -191,7 +191,7 @@ public abstract class LavaFluid extends FlowingFluid {
 
     @Override
     public int getTickDelay(final LevelReader level) {
-        return isFastLava(level) ? 10 : 30;
+        return level.getWorldBorder().world != null ? (isFastLava(level) ? level.getWorldBorder().world.purpurConfig.lavaSpeedNether : level.getWorldBorder().world.purpurConfig.lavaSpeedNotNether) : (isFastLava(level) ? 10 : 30); // Purpur - Make lava flow speed configurable
     }
 
     @Override
@@ -213,6 +213,13 @@ public abstract class LavaFluid extends FlowingFluid {
         level.levelEvent(LevelEvent.LAVA_FIZZ, pos, 0);
     }
 
+    // Purpur start - Implement infinite liquids
+    @Override
+    protected int getRequiredSources(Level level) {
+        return level.purpurConfig.lavaInfiniteRequiredSources;
+    }
+    // Purpur end - Implement infinite liquids
+
     @Override
     protected boolean canConvertToSource(final ServerLevel level) {
         return level.getGameRules().get(GameRules.LAVA_SOURCE_CONVERSION);
