--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -199,6 +_,13 @@
         level.levelEvent(1501, pos, 0);
     }
 
+    // Purpur start - Implement infinite liquids
+    @Override
+    protected int getRequiredSources(Level level) {
+        return level.purpurConfig.lavaInfiniteRequiredSources;
+    }
+    // Purpur end - Implement infinite liquids
+
     @Override
     protected boolean canConvertToSource(ServerLevel level) {
         return level.getGameRules().getBoolean(GameRules.RULE_LAVA_SOURCE_CONVERSION);
