From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/material/WaterFluid.java b/net/minecraft/world/level/material/WaterFluid.java
index d0a62bad5592c3ef8bd2ba5bc9719fad66c4edf0..e52362ae20646f0311262853af0cdc690a603b52 100644
--- a/net/minecraft/world/level/material/WaterFluid.java
+++ b/net/minecraft/world/level/material/WaterFluid.java
@@ -77,6 +77,13 @@ public abstract class WaterFluid extends FlowingFluid {
         return level.getGameRules().get(GameRules.WATER_SOURCE_CONVERSION);
     }
 
+    // Purpur start - Implement infinite liquids
+    @Override
+    protected int getRequiredSources(Level level) {
+        return level.purpurConfig.waterInfiniteRequiredSources;
+    }
+    // Purpur end - Implement infinite liquids
+
     // Paper start - Add BlockBreakBlockEvent
     @Override
     protected void beforeDestroyingBlock(LevelAccessor level, BlockPos pos, BlockState state, BlockPos source) {
