--- a/net/minecraft/world/level/block/piston/PistonStructureResolver.java
+++ b/net/minecraft/world/level/block/piston/PistonStructureResolver.java
@@ -90,7 +_,7 @@
         }
 
         int blockCount = 1;
-        if (blockCount + this.toPush.size() > 12) {
+        if (blockCount + this.toPush.size() > this.level.purpurConfig.pistonBlockPushLimit) { // Purpur - Configurable piston push limit
             return false;
         }
 
@@ -105,7 +_,7 @@
                 break;
             }
 
-            if (++blockCount + this.toPush.size() > 12) {
+            if (++blockCount + this.toPush.size() > this.level.purpurConfig.pistonBlockPushLimit) { // Purpur - Configurable piston push limit
                 return false;
             }
         }
@@ -149,7 +_,7 @@
                 return true;
             }
 
-            if (this.toPush.size() >= 12) {
+            if (this.toPush.size() >= this.level.purpurConfig.pistonBlockPushLimit) { // Purpur - Configurable piston push limit
                 return false;
             }
 
