From 7dba6f0d46dbbd3d36b75b36e56ea6b5da391b16 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/world/level/block/piston/PistonStructureResolver.java b/net/minecraft/world/level/block/piston/PistonStructureResolver.java
index 4a2cd2241a65fe051aae0b87e33580784bd9c2cb..c47a763c75913dcaefc6709e2b7f44a8b87fcef1 100644
--- a/net/minecraft/world/level/block/piston/PistonStructureResolver.java
+++ b/net/minecraft/world/level/block/piston/PistonStructureResolver.java
@@ -90,7 +90,7 @@ public class PistonStructureResolver {
         }
 
         int blockCount = 1;
-        if (blockCount + this.toPush.size() > 12) {
+        if (blockCount + this.toPush.size() > this.level.purpurConfig.pistonBlockPushLimit) { // Purpur - Configurable piston push limit
             return false;
         }
 
@@ -105,7 +105,7 @@ public class PistonStructureResolver {
                 break;
             }
 
-            if (++blockCount + this.toPush.size() > 12) {
+            if (++blockCount + this.toPush.size() > this.level.purpurConfig.pistonBlockPushLimit) { // Purpur - Configurable piston push limit
                 return false;
             }
         }
@@ -149,7 +149,7 @@ public class PistonStructureResolver {
                 return true;
             }
 
-            if (this.toPush.size() >= 12) {
+            if (this.toPush.size() >= this.level.purpurConfig.pistonBlockPushLimit) { // Purpur - Configurable piston push limit
                 return false;
             }
 
