From 505663ae4ebe983e4f5d0c9694cb5d07e2efc170 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] purpur File Patches


diff --git a/net/minecraft/CrashReport.java b/net/minecraft/CrashReport.java
index 84247ddc9e063676c59a6e62da1eb919320bd18f..c239edddec50792b14b2c8163f18cb3dc0c05e63 100644
--- a/net/minecraft/CrashReport.java
+++ b/net/minecraft/CrashReport.java
@@ -30,6 +30,7 @@ public class CrashReport {
     private boolean trackingStackTrace = true;
     private StackTraceElement[] uncategorizedStackTrace = new StackTraceElement[0];
     private final SystemReport systemReport = new SystemReport();
+    private List<String> extraInfo = List.of("", "DO NOT REPORT THIS TO PAPER! REPORT TO PURPUR INSTEAD!", ""); // Purpur - Rebrand
 
     public CrashReport(final String title, final Throwable t) {
         this.title = title;
@@ -129,7 +130,7 @@ public class CrashReport {
     }
 
     public String getFriendlyReport(final ReportType reportType) {
-        return this.getFriendlyReport(reportType, List.of());
+        return this.getFriendlyReport(reportType, extraInfo); // Purpur - Rebrand
     }
 
     public @Nullable Path getSaveFile() {
@@ -159,7 +160,7 @@ public class CrashReport {
     }
 
     public boolean saveToFile(final Path file, final ReportType reportType) {
-        return this.saveToFile(file, reportType, List.of());
+        return this.saveToFile(file, reportType, extraInfo); // Purpur - Rebrand
     }
 
     public SystemReport getSystemReport() {
